{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Import relevant libraries\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib\n",
    "from matplotlib import rcParams\n",
    "rcParams['font.family'] = 'monospace'\n",
    "#rcParams['font.sans-serif'] = ['Tahoma']\n",
    "\n",
    "import math\n",
    "\n",
    "import datetime\n",
    "\n",
    "import numpy as np\n",
    "\n",
    "import networkx as nx\n",
    "\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Defines TodaysDate, which is used in output-file name\n",
    "def TodaysDate():\n",
    "        \n",
    "    Today = datetime.date.today()\n",
    "    TodaysDate = Today.strftime('%d%b%Y')\n",
    "    \n",
    "    return TodaysDate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Validate the DNACodingSequence based on (1) its Lenght AND (2) QualityScore of each nucleotide \n",
    "\n",
    "def DNACodingSequence(DNASequence,\n",
    "                      QualitySequence,\n",
    "                      StartSequence,\n",
    "                      StopSequence,\n",
    "                      cDNA_MinLength,\n",
    "                      cDNA_MaxLength,\n",
    "                      QualityScore):\n",
    "#utilises ONLY ONE StopSequence, returns ONLY ONE CodingSequence\n",
    "    \n",
    "    QualityScoreString = \"\"\"!\"#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~\"\"\"\n",
    "    ThresholdQualityString = QualityScoreString[QualityScore:]\n",
    "            \n",
    "    StartIndex = DNASequence.find(StartSequence) + len(StartSequence)\n",
    "    StopIndex = DNASequence.rfind(StopSequence)\n",
    "    CodingSequence =  DNASequence[StartIndex:StopIndex]\n",
    "    if cDNA_MinLength <= len(CodingSequence) and len(CodingSequence) <= cDNA_MaxLength and len(CodingSequence)%3 == 0:\n",
    "        for Character in QualitySequence[StartIndex:StopIndex]:\n",
    "            if Character not in ThresholdQualityString:\n",
    "                return None\n",
    "        return str(CodingSequence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Define Translation of RNASequence into Peptide (peptide can be encoded as DNA or as RNA)\n",
    "\n",
    "def Translation(CodingSequence):\n",
    "\n",
    "    # Set RNA TranslationCode\n",
    "    TranslationCode = {\n",
    "                    'AAA':'K','AAC':'N','AAG':'K','AAU':'N',\n",
    "                    'ACA':'T','ACC':'T','ACG':'T','ACU':'T',\n",
    "                    'AGA':'R','AGC':'S','AGG':'R','AGU':'S',\n",
    "                    'AUA':'I','AUC':'I','AUG':'M','AUU':'I',\n",
    "                    \n",
    "                    'CAA':'Q','CAC':'H','CAG':'Q','CAU':'H',\n",
    "                    'CCA':'P','CCC':'P','CCG':'P','CCU':'P',\n",
    "                    'CGA':'R','CGC':'R','CGG':'R','CGU':'R',\n",
    "                    'CUA':'L','CUC':'L','CUG':'L','CUU':'L',\n",
    "                    \n",
    "                    'GAA':'E','GAC':'D','GAG':'E','GAU':'D',\n",
    "                    'GCA':'A','GCC':'A','GCG':'A','GCU':'A',\n",
    "                    'GGA':'G','GGC':'G','GGG':'G','GGU':'G',\n",
    "                    'GUA':'V','GUC':'V','GUG':'V','GUU':'V',\n",
    "                    \n",
    "                    'UAA':'#','UAC':'Y','UAG':'*','UAU':'Y',\n",
    "                    'UCA':'S','UCC':'S','UCG':'S','UCU':'S',\n",
    "                    'UGA':'&','UGC':'C','UGG':'W','UGU':'C',\n",
    "                    'UUA':'L','UUC':'F','UUG':'L','UUU':'F'\n",
    "                        }\n",
    "    # Stop Codons are as follows:\n",
    "    # UAA (ochre) — #\n",
    "    # UAG (amber) — *\n",
    "    # UGA (opal) — &\n",
    "                    \n",
    "    TranscriptionCode = {'A':'A','C':'C','G':'G','T':'U','U':'T'}\n",
    "    \n",
    "    # Convert DNA to RNA\n",
    "    RNASequence = ''\n",
    "    for Nucleotide in CodingSequence:\n",
    "        RNASequence += TranscriptionCode.get(Nucleotide,'X')\n",
    "        \n",
    "    #RNASequence-Test Print\n",
    "    #print (RNASequence)\n",
    "        \n",
    "    Peptide = ''\n",
    "    while len(RNASequence) != 0:\n",
    "        Peptide += TranslationCode.get(RNASequence[0:3],'Do not fuck with me!')\n",
    "        RNASequence = RNASequence[3:]\n",
    "    return Peptide"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Define SingleSelectionRoundSummary function, which returns the Occurence of CodingSequences, grouped by Peptide\n",
    "# SingleSelectionRoundSummary = {PeptideY:    {CodingSequence_YZ:    Occurence_YZ}}\n",
    "\n",
    "def SingleSelectionRoundSummary(fastqFileLocation, StartSequence, StopSequence, cDNA_MinLength, cDNA_MaxLength, QualityScore):\n",
    "    \n",
    "    RawDataFile = open(fastqFileLocation, 'r')\n",
    "    Lines = RawDataFile.readlines()\n",
    "    RawDataFile.close\n",
    "\n",
    "    # Create an empty SingleSelectionRoundSummary dictionary to store the results from a single cycle of selection\n",
    "    # SingleSelectionRoundSummary = {PeptideY:    {CodingSequence_YZ:    Occurence_YZ}}\n",
    "    SingleSelectionRoundSummary = {}\n",
    "   \n",
    "    # Populate SingleSelectionRoundSummary with the results from a single round of selection\n",
    "    # Go through the NGS-data file (.fastq) line by line\n",
    "    for i,Line in enumerate(Lines):\n",
    "        # Check that the line contains a valid sequence\n",
    "        if StartSequence in Line and StopSequence in Line:\n",
    "            CodingSequence = DNACodingSequence(Line, Lines[i + 2], StartSequence, StopSequence, cDNA_MinLength, cDNA_MaxLength, QualityScore)                      \n",
    "            if CodingSequence != None:\n",
    "                # Translated DNA or RNA sequence into PeptideSequence\n",
    "                PeptideSequence = Translation(CodingSequence)\n",
    "                # Add sequence to a SingleSelectionRoundSummary\n",
    "                if PeptideSequence not in SingleSelectionRoundSummary:\n",
    "                    SingleSelectionRoundSummary[str(PeptideSequence)] = {str(CodingSequence) : 1}\n",
    "                else:\n",
    "                    if CodingSequence not in SingleSelectionRoundSummary[str(PeptideSequence)]:\n",
    "                        SingleSelectionRoundSummary[str(PeptideSequence)][str(CodingSequence)] = 1\n",
    "                    else:\n",
    "                        SingleSelectionRoundSummary[str(PeptideSequence)][str(CodingSequence)] += 1\n",
    "\n",
    "    return SingleSelectionRoundSummary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Define HammingDistance, which returns the number of mismatches between two Sequences\n",
    "def HammingDistance(Sequence1, Sequence2):\n",
    "    \n",
    "    if len(Sequence1) < len(Sequence2):\n",
    "        Sequence1 = Sequence1 + (len(Sequence2) - len(Sequence1)) * '%'\n",
    "    elif len(Sequence1) > len(Sequence2):\n",
    "        Sequence2 = Sequence2 + (len(Sequence1) - len(Sequence2)) * '%'\n",
    "    \n",
    "    HammingDistance = 0\n",
    "    for i in range(len(Sequence1)):\n",
    "        if Sequence1[i] == Sequence2[i]:\n",
    "            HammingDistance = HammingDistance\n",
    "        else:\n",
    "            HammingDistance = HammingDistance + 1\n",
    "            \n",
    "    return HammingDistance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Define HammingDistanceBasedFormating, which returns FormatedSequence, such that mismatches are capitalised, and the rest of the FormatedSequence is lowercase \n",
    "def HammingDistanceBasedFormating(Sequence1, Sequence2):\n",
    "    \n",
    "    if len(Sequence1) < len(Sequence2):\n",
    "        Sequence1 = Sequence1 + (len(Sequence2) - len(Sequence1)) * '-'\n",
    "    elif len(Sequence1) > len(Sequence2):\n",
    "        Sequence2 = Sequence2 + (len(Sequence1) - len(Sequence2)) * '-'\n",
    "    \n",
    "    HammingDistance = 0\n",
    "    FormatedSequence2 = ''\n",
    "    for i in range(len(Sequence1)):\n",
    "        if Sequence1[i] == Sequence2[i]:\n",
    "            FormatedSequence2 += Sequence2[i].lower()\n",
    "            HammingDistance = HammingDistance\n",
    "        else:\n",
    "            FormatedSequence2 += Sequence2[i]\n",
    "            HammingDistance = HammingDistance + 1            \n",
    "    return FormatedSequence2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Define CompleteSelectionSummary, which returns the Occurence of CodingSequences, grouped by Peptide, grouped by SelectionRound\n",
    "# SelectionSummary = {SelectionRound_X:    {PeptideXY:    {CodingDNA_XYZ:    Occurence_XYZ}}}\n",
    "def CompleteSelectionSummary(DataFolderLocation,\n",
    "                             StartSequence,\n",
    "                             StopSequence,\n",
    "                             cDNA_MinLength,\n",
    "                             cDNA_MaxLength,\n",
    "                             QualityScore):\n",
    "    # Create an empty SelectionSummary dictionary to store the results from all the rounds of selection\n",
    "    CompleteSelectionSummary = {}\n",
    "    \n",
    "    for file in os.listdir(DataFolderLocation):\n",
    "        \n",
    "        FileLocation = os.path.join(DataFolderLocation, file)\n",
    "        \n",
    "        #(1.A) Extract the round number from the file name (file name should have two digit number before full stop — '00.')    \n",
    "        if file.endswith('.fastq'): # this conditional is necessary; without it some shit appears in the beginning of the file list\n",
    "            RoundNumberFirstDigit = file[file.find('.')-2]\n",
    "            RoundNumberSecondDigit = file[file.find('.')-1]\n",
    "            if RoundNumberFirstDigit == '0':\n",
    "                RoundNumber = int(RoundNumberSecondDigit)\n",
    "                # RoundNumber_Test Print\n",
    "                #print RoundNumber\n",
    "            elif RoundNumberFirstDigit != '0':\n",
    "                RoundNumber = int(file[file.find('.')-2 : file.find('.')])\n",
    "                # RoundNumber_Test Print\n",
    "                #print RoundNumber\n",
    "            \n",
    "            #(1.B) Extract single round results\n",
    "            SelectionRoundSummary = SingleSelectionRoundSummary(FileLocation, StartSequence, StopSequence, cDNA_MinLength, cDNA_MaxLength, QualityScore)\n",
    "             \n",
    "            #(1.C) Populate ConcatenatedResultsList\n",
    "            CompleteSelectionSummary[RoundNumber] = SelectionRoundSummary\n",
    "            \n",
    "            # ConcatenatedResultsList-Test Print\n",
    "            #print ConcatenatedResultsList\n",
    "            \n",
    "    return CompleteSelectionSummary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Define PeptidesOccurences_BY_Round, which returns the Occurences of Peptides groupped by round\n",
    "# PeptidesOccurences_BY_Round = {Round_X:    {PeptideXY:    Occurence_XY}}\n",
    "def PeptidesOccurences_BY_Round(DataFolderLocation, cDNA_MinLength, cDNA_MaxLength, QualityScore):\n",
    "    SelectionSummary = CompleteSelectionSummary(DataFolderLocation, StartSequence, StopSequence, cDNA_MinLength, cDNA_MaxLength, QualityScore)\n",
    "    \n",
    "    PeptidesOccurences_BY_Round = {}\n",
    "    for Round in SelectionSummary:\n",
    "        PeptidesOccurences_IN_Round = {}\n",
    "        for Peptide in SelectionSummary[Round]:\n",
    "            PeptidesOccurences_IN_Round[Peptide] = sum(SelectionSummary[Round][Peptide].values())\n",
    "        PeptidesOccurences_BY_Round[Round] = PeptidesOccurences_IN_Round\n",
    "        \n",
    "    return PeptidesOccurences_BY_Round"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Define DNAsOccurences_BY_Round, which returns the Occurences of DNAs groupped by round\n",
    "# DNAsOccurences_BY_Round = {Round_X:    {DNA_XY:    Occurence_XY}}\n",
    "def DNAsOccurences_BY_Round(DataFolderLocation, cDNA_MinLength, cDNA_MaxLength, QualityScore):\n",
    "    SelectionSummary = CompleteSelectionSummary(DataFolderLocation, StartSequence, StopSequence, cDNA_MinLength, cDNA_MaxLength, QualityScore)\n",
    "    \n",
    "    DNAsOccurences_BY_Round = {}\n",
    "    for Round in SelectionSummary:\n",
    "        DNAsOccurences_IN_Round = {}\n",
    "        for Peptide in SelectionSummary[Round]:\n",
    "            for DNA in SelectionSummary[Round][Peptide]:\n",
    "                DNAsOccurences_IN_Round[DNA] = SelectionSummary[Round][Peptide][DNA]\n",
    "        DNAsOccurences_BY_Round[Round] = DNAsOccurences_IN_Round\n",
    "\n",
    "    return DNAsOccurences_BY_Round"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Define TotalReads_BY_Round, which returns number of reads by round\n",
    "# TotalReads_BY_Round = {Round_X:    TotalReads_X}\n",
    "def TotalReads_BY_Round(DataFolderLocation, cDNA_MinLength, cDNA_MaxLength, QualityScore):\n",
    "    SelectionSummary = CompleteSelectionSummary(DataFolderLocation, StartSequence, StopSequence, cDNA_MinLength, cDNA_MaxLength, QualityScore)\n",
    "    Peptides_BY_Round = PeptidesOccurences_BY_Round(DataFolderLocation, cDNA_MinLength, cDNA_MaxLength, QualityScore)\n",
    "    \n",
    "    TotalReads_BY_Round = {}\n",
    "    for Round in SelectionSummary:\n",
    "        TotalReads_BY_Round[Round] = sum(Peptides_BY_Round[Round].values())\n",
    "        \n",
    "    return TotalReads_BY_Round"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Define BaseRoundSortedPeptidesList, which returns list of peptides in base round sorted by their occurence\n",
    "# BaseRoundSortedPeptidesList = [Peptide_1, ..., Peptide_N]; Occurence(Peptide_1) > ... > Occurence(Peptide_N)\n",
    "def BaseRoundSortedPeptidesList(DataFolderLocation, BaseSelectionRoundNumber):\n",
    "    Peptides_BY_Round = PeptidesOccurences_BY_Round(DataFolderLocation, cDNA_MinLength, cDNA_MaxLength, QualityScore)  \n",
    "            \n",
    "    PeptidesOccurencesInBaseRound = Peptides_BY_Round[BaseSelectionRoundNumber]\n",
    "    BaseRoundSortedPeptidesList = sorted(PeptidesOccurencesInBaseRound, key = PeptidesOccurencesInBaseRound.get, reverse = True)\n",
    "    \n",
    "    return BaseRoundSortedPeptidesList"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Define BaseRoundSortedDNAsList, which returns list of DNAs in base round sorted by their occurence\n",
    "# BaseRoundSortedDNAsList = [DNA_1, ..., DNA_N]; Occurence(DNA_1) > ... > Occurence(DNA_N)\n",
    "def BaseRoundSortedDNAsList(DataFolderLocation, BaseSelectionRoundNumber):\n",
    "    DNAs_BY_Round = DNAsOccurences_BY_Round(DataFolderLocation, cDNA_MinLength, cDNA_MaxLength, QualityScore)  \n",
    "            \n",
    "    DNAsOccurences_IN_BaseRound = DNAs_BY_Round[BaseSelectionRoundNumber]\n",
    "    BaseRoundSortedDNAsList = sorted(DNAsOccurences_IN_BaseRound, key = DNAsOccurences_IN_BaseRound.get, reverse = True)\n",
    "    \n",
    "    return BaseRoundSortedDNAsList"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def PeptidesRank_IN_BaseRound(DataFolderLocation, BaseSelectionRoundNumber):\n",
    "    Peptides_BY_Round = PeptidesOccurences_BY_Round(DataFolderLocation, cDNA_MinLength, cDNA_MaxLength, QualityScore)\n",
    "    BaseRoundSortedPeptides = BaseRoundSortedPeptidesList(DataFolderLocation, BaseSelectionRoundNumber)\n",
    "    \n",
    "    BasePeptideCount = 0\n",
    "    PeptideRank = 1\n",
    "    \n",
    "    PeptidesRank_IN_BaseRound = {}\n",
    "    \n",
    "    for Peptide in BaseRoundSortedPeptides:\n",
    "        PeptideCount = Peptides_BY_Round[BaseSelectionRoundNumber][Peptide]\n",
    "        if PeptideCount < BasePeptideCount:\n",
    "            PeptideRank += 1\n",
    "        \n",
    "        PeptidesRank_IN_BaseRound[Peptide] = PeptideRank\n",
    "        BasePeptideCount = PeptideCount\n",
    "        \n",
    "    return PeptidesRank_IN_BaseRound"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Occurences is a bad word in here\n",
    "\n",
    "\n",
    "# Define DNAClonesOccurences_BY_Round_BY_Peptide, which returns number of clones for each peptide groupped by round.\n",
    "# DNAClonesOccurences_BY_Round_BY_Peptide = {SelectionRound_X:    {PeptideXY:    DNAClonesOccurences}}\n",
    "def DNAClonesOccurences_BY_Round_BY_Peptide(DataFolderLocation, cDNA_MinLength, cDNA_MaxLength, QualityScore):\n",
    "    SelectionSummary = CompleteSelectionSummary(DataFolderLocation, StartSequence, StopSequence, cDNA_MinLength, cDNA_MaxLength, QualityScore)\n",
    "    \n",
    "    DNAClonesOccurences_BY_Round_BY_Peptide = {}\n",
    "    for Round in SelectionSummary:\n",
    "        DNAClonesOccurences_BY_Peptide = {}\n",
    "        for Peptide in SelectionSummary[Round]:\n",
    "            DNAClonesOccurences_BY_Peptide[Peptide] = len(SelectionSummary[Round][Peptide])\n",
    "        DNAClonesOccurences_BY_Round_BY_Peptide[Round] = DNAClonesOccurences_BY_Peptide\n",
    "        \n",
    "    return DNAClonesOccurences_BY_Round_BY_Peptide"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Define PeptidesAppearances_BY_Round, which returns for each peptide in selection a list of rounds in which this peptide appears.\n",
    "# PeptidesAppearances_BY_Round = {Peptide_X:    [Round_A, ..., Round_N]}\n",
    "def PeptidesAppearances_BY_Round(BaseRoundSortedPeptidesList, PeptidesOccurences_BY_Round):\n",
    "    \n",
    "    PeptidesAppearances_BY_Round = {}\n",
    "    \n",
    "    for Peptide in BaseRoundSortedPeptidesList:\n",
    "        PeptidesAppearances_BY_Round[Peptide] = []\n",
    "        for Round in PeptidesOccurences_BY_Round:\n",
    "            if Peptide in PeptidesOccurences_BY_Round[Round]:\n",
    "                PeptidesAppearances_BY_Round[Peptide] += [Round]\n",
    "    return PeptidesAppearances_BY_Round"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Define DNAsAppearances_BY_Round, which returns for each DNA in selection a list of rounds in which this DNA appears.\n",
    "# DNAsAppearances_BY_Round = {DNA_X:    [Round_A, ..., Round_N]}\n",
    "def DNAsAppearances_BY_Round(BaseRoundSortedDNAsList, DNAsOccurences_BY_Round):\n",
    "    \n",
    "    DNAsAppearances_BY_Round = {}\n",
    "    \n",
    "    for DNA in BaseRoundSortedDNAsList:\n",
    "        DNAsAppearances_BY_Round[DNA] = []\n",
    "        for Round in DNAsOccurences_BY_Round:\n",
    "            if DNA in DNAsOccurences_BY_Round[Round]:\n",
    "                DNAsAppearances_BY_Round[DNA] += [Round]\n",
    "    return DNAsAppearances_BY_Round"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Define SelectionSummaryReport, which, for the TopNPeptidesNumber, returns a summary table (.txt) and provides a summary grpaph (.png).\n",
    "def SelectionSummaryReport(DataFolderLocation,\n",
    "                           BaseSelectionRoundNumber,\n",
    "                           TopNPeptidesNumber,\n",
    "                           StartSequence,\n",
    "                           StopSequence,\n",
    "                           cDNA_MinLength,\n",
    "                           cDNA_MaxLength,\n",
    "                           QualityScore,\n",
    "                           FileName):\n",
    "    \n",
    "    today = TodaysDate() \n",
    "    \n",
    "    SelectionSummaryFileNameCSV = str(today) + 'SelectionSummary' + FileName + '.csv'\n",
    "    SelectionSummaryReportFile = open(SelectionSummaryFileNameCSV, 'w')\n",
    "    \n",
    "    SelectionSummary = CompleteSelectionSummary(DataFolderLocation, StartSequence, StopSequence, cDNA_MinLength, cDNA_MaxLength, QualityScore)\n",
    "    SortedRoundsList = sorted(SelectionSummary.keys())\n",
    "    \n",
    "    Peptides_BY_Round = PeptidesOccurences_BY_Round(DataFolderLocation, cDNA_MinLength, cDNA_MaxLength, QualityScore)\n",
    "    TotalPeptides_BY_Round = TotalReads_BY_Round(DataFolderLocation, cDNA_MinLength, cDNA_MaxLength, QualityScore)\n",
    "    \n",
    "    BaseRoundSortedPeptides = BaseRoundSortedPeptidesList(DataFolderLocation, BaseSelectionRoundNumber)\n",
    "    BaseRoundTopSortedPeptides = BaseRoundSortedPeptides[0 : (TopNPeptidesNumber)]\n",
    "\n",
    "    BaseRoundPeptidesRank = PeptidesRank_IN_BaseRound(DataFolderLocation, BaseSelectionRoundNumber)\n",
    "    #print (BaseRoundSortedPeptides)\n",
    "    \n",
    "        \n",
    "    SelectionSummaryReportFile.write('peptide sequence' + ',' +\n",
    "                                     'rank (#)' + ',' +\n",
    "                                     'cDNA mutants' + ',')\n",
    "    for Round in SortedRoundsList:\n",
    "        SelectionSummaryReportFile.write('C' +\n",
    "                                         str(Round) +\n",
    "                                         ' count (#) [frequency(%)]' + ',')\n",
    "    SelectionSummaryReportFile.write('\\n')\n",
    "    \n",
    "    for Peptide in BaseRoundTopSortedPeptides:\n",
    "        BaseRoundPeptideFraction = float((Peptides_BY_Round[Round].get(Peptide, 0)))/float(TotalPeptides_BY_Round[BaseSelectionRoundNumber])\n",
    "        PeptideRank = BaseRoundPeptidesRank[Peptide]\n",
    "        FormatedPeptide = HammingDistanceBasedFormating(BaseRoundTopSortedPeptides[0], Peptide)\n",
    "        PeptidecDNAMutants = len(SelectionSummary[BaseSelectionRoundNumber][Peptide])\n",
    "        SelectionSummaryReportFile.write(FormatedPeptide + ',' +\n",
    "                                         str(PeptideRank) + ',' +\n",
    "                                         str(PeptidecDNAMutants) + ',')\n",
    "            \n",
    "        for Round in SortedRoundsList:\n",
    "            PeptideFraction = float((Peptides_BY_Round[Round].get(Peptide, 0)))/float(TotalPeptides_BY_Round[Round])\n",
    "\n",
    "            BaseFraction = PeptideFraction\n",
    "            \n",
    "            SelectionSummaryReportFile.write(str(Peptides_BY_Round[Round].get(Peptide, 0)) +\n",
    "                                             ' [' + '{:.1%}'.format(PeptideFraction) + ']' + ',')\n",
    "            \n",
    "            BaseFraction = PeptideFraction\n",
    "        SelectionSummaryReportFile.write('\\n')\n",
    "        \n",
    "    SelectionSummaryReportFile.write('total count (#)' + ',' + ',')\n",
    "    for Round in SortedRoundsList:\n",
    "        SelectionSummaryReportFile.write(str(TotalPeptides_BY_Round[Round]) + ',')\n",
    "    SelectionSummaryReportFile.write('\\n\\n\\n')\n",
    "            \n",
    "    SelectionSummaryReportFile.close()\n",
    "    \n",
    "#-------------------------------------------------------------------------------\n",
    "   \n",
    "    # Create a figure of size 8x6 inches, 500 dots per inch\n",
    "    plt.figure(figsize = (8, 6),\n",
    "               dpi = 500)\n",
    "    # Create 'ggplot' style\n",
    "    plt.style.use('fivethirtyeight')\n",
    "    # Create a new subplot from a grid of 1x1\n",
    "    Graph = plt.subplot(1, 1, 1)\n",
    "    \n",
    "    Xs = []\n",
    "    Ys = []\n",
    "\n",
    "    # Map colours onto lines  \n",
    "    cNorm  = matplotlib.colors.Normalize(vmin = 0,\n",
    "                                         vmax = TopNPeptidesNumber - 1)\n",
    "    scalarMap = matplotlib.cm.ScalarMappable(norm = cNorm,\n",
    "                                             cmap = 'gist_rainbow')\n",
    "    \n",
    "    PeptideLabels = []\n",
    "    \n",
    "    for Peptide in BaseRoundTopSortedPeptides:\n",
    "    #for Peptide in Top24PeptidesKDs:\n",
    "        PeptidesFractions_BY_Round = []\n",
    "        for Round in SortedRoundsList:\n",
    "            PeptidesFractions_BY_Round += [float((Peptides_BY_Round[Round].get(Peptide, 0)))/float(TotalPeptides_BY_Round[Round])]\n",
    "        \n",
    "        x = SortedRoundsList\n",
    "        y = PeptidesFractions_BY_Round\n",
    "        Xs += x\n",
    "        Ys += y\n",
    "        \n",
    "        PeptideRank = BaseRoundPeptidesRank[Peptide]\n",
    "        #PeptideColour = scalarMap.to_rgba(PeptideRank)\n",
    "        PeptideColour = scalarMap.to_rgba(BaseRoundTopSortedPeptides.index(Peptide))\n",
    "        FormatedPeptide = HammingDistanceBasedFormating(BaseRoundTopSortedPeptides[0], Peptide)\n",
    "        \n",
    "        PeptideLabel =  FormatedPeptide + ' (' + str(PeptideRank) +')'\n",
    "        \n",
    "        #Set PeptideLabel\n",
    "        PeptideLabels += [PeptideLabel]\n",
    "        \n",
    "        plt.plot(x, y,\n",
    "                 'o-',\n",
    "                 c = PeptideColour,\n",
    "                 lw = 2.0,\n",
    "                 ms = 4.0,\n",
    "                 mew = 0.1,\n",
    "                 mec = '#191919')\n",
    "\n",
    "    XMin = min(Xs) - 0.05*(max(Xs) - min(Xs))\n",
    "    XMax = max(Xs) + 0.05*(max(Xs) - min(Xs))\n",
    "    YMin = min(Ys) - 0.05*(max(Ys) - min(Ys))\n",
    "    YMax = max(Ys) + 0.05*(max(Ys) - min(Ys))\n",
    "    \n",
    "    plt.axis([XMin, XMax, YMin, YMax])\n",
    "    \n",
    "    plt.xticks(fontsize = 10)\n",
    "    plt.yticks(fontsize = 10)\n",
    "    \n",
    "    plt.xlabel('Selection Cycle (#)',\n",
    "               fontsize = 10)\n",
    "    plt.ylabel('Peptide Fraction (%)',\n",
    "               fontsize = 10)\n",
    "    \n",
    "    legend = plt.legend(PeptideLabels,\n",
    "                        title = 'cyclic-peptide random region',\n",
    "                        loc = 'upper center',\n",
    "                        bbox_to_anchor = (0.5, -0.10),\n",
    "                        fancybox = True,\n",
    "                        shadow = False,\n",
    "                        fontsize = 10,\n",
    "                        ncol = 3)\n",
    "    \n",
    "    Graph.get_legend().get_title().set_size('small')\n",
    "    \n",
    "    SelectionSummaryFileNamePNG = str(today) + 'SelectionSummary' + FileName + '.png'\n",
    "    \n",
    "    plt.savefig(SelectionSummaryFileNamePNG,\n",
    "                bbox_extra_artists = [legend],\n",
    "                bbox_inches = 'tight',\n",
    "                dpi = 300)\n",
    "    plt.show()\n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def DNAMutantsAnalysis(DataFolderLocation,\n",
    "                       BaseSelectionRoundNumber,\n",
    "                       TopNPeptidesNumber,\n",
    "                       StartSequence,\n",
    "                       StopSequence,\n",
    "                       FileName):\n",
    "    \n",
    "    today = TodaysDate() \n",
    "    \n",
    "    DNAMutantsAnalysisFileNameCSV =  str(today) + 'DNAsMutantsAnalysis' + FileName + '.csv'\n",
    "    DNAMutantsAnalysisFile = open(DNAMutantsAnalysisFileNameCSV, 'w')\n",
    "    \n",
    "    SelectionSummary = CompleteSelectionSummary(DataFolderLocation, StartSequence, StopSequence, cDNA_MinLength, cDNA_MaxLength, QualityScore)\n",
    "    SortedRoundsList = sorted(SelectionSummary.keys())\n",
    "    \n",
    "    Peptides_BY_Round = PeptidesOccurences_BY_Round(DataFolderLocation, cDNA_MinLength, cDNA_MaxLength, QualityScore)\n",
    "    TotalPeptides_BY_Round = TotalReads_BY_Round(DataFolderLocation, cDNA_MinLength, cDNA_MaxLength, QualityScore)\n",
    "    \n",
    "    BaseRoundSortedPeptides = BaseRoundSortedPeptidesList(DataFolderLocation, BaseSelectionRoundNumber)\n",
    "    BaseRoundTopSortedPeptides = BaseRoundSortedPeptides[0 : (TopNPeptidesNumber)]\n",
    "    \n",
    "    DNAClones_BY_Round_BY_Peptide = DNAClonesOccurences_BY_Round_BY_Peptide(DataFolderLocation, cDNA_MinLength, cDNA_MaxLength, QualityScore)\n",
    "    \n",
    "    DNAMutantsAnalysisFile.write('peptide sequence' + ',')\n",
    "    for Round in SortedRoundsList:\n",
    "        DNAMutantsAnalysisFile.write('round # ' + str(Round) + ' DNA clones (#)' + ',')\n",
    "    DNAMutantsAnalysisFile.write('\\n')\n",
    "    \n",
    "    for Peptide in BaseRoundTopSortedPeptides:\n",
    "        DNAMutantsAnalysisFile.write(Peptide + ',')\n",
    "        for Round in SortedRoundsList:\n",
    "            DNAMutantsAnalysisFile.write(str(DNAClones_BY_Round_BY_Peptide[Round].get(Peptide, 0)) + ',')\n",
    "        DNAMutantsAnalysisFile.write('\\n')\n",
    "        \n",
    "#    DNAClonesSummaryReportFile.write('total #' + ',')\n",
    "#    for Round in SortedRoundsList:\n",
    "#        DNAClonesSummaryReportFile.write(str(TotalPeptides_BY_Round[Round]) + ',')\n",
    "#    DNAClonesSummaryReportFile.write('\\n\\n\\n')\n",
    "    \n",
    "#    DNAClonesSummaryReportFile.write('peptide sequence' + ',')\n",
    "#    for Round in SortedRoundsList:\n",
    "#        DNAClonesSummaryReportFile.write('round # ' + str(Round) + ' fraction (%)' + ',')\n",
    "#    DNAClonesSummaryReportFile.write('\\n')\n",
    "    \n",
    "#    for Peptide in BaseRoundTopSortedPeptides:\n",
    "#        DNAClonesSummaryReportFile.write(Peptide + ',')\n",
    "#        for Round in SortedRoundsList:\n",
    "#            PeptideFraction = float((Peptides_BY_Round[Round].get(Peptide, 0)))/float(TotalPeptides_BY_Round[Round])\n",
    "#            DNAClonesSummaryReportFile.write('{:.3%}'.format(PeptideFraction) + ',')\n",
    "#        DNAClonesSummaryReportFile.write('\\n')\n",
    "            \n",
    "    DNAMutantsAnalysisFile.close()\n",
    "    \n",
    "#-------------------------------------------------------------------------------        \n",
    "#    plt.style.use('fivethirtyeight') # just to create 'ggplot' style\n",
    "    \n",
    "#    Xs = []\n",
    "#    Ys = []    \n",
    "#    for Peptide in BaseRoundTopSortedPeptides:\n",
    "#        SortedDNAClones_BY_Peptide = []\n",
    "#        for Round in SortedRoundsList:\n",
    "#            SortedDNAClones_BY_Peptide += [DNAClones_BY_Round_BY_Peptide[Round].get(Peptide, 0)]\n",
    "        \n",
    "#        x = SortedRoundsList\n",
    "#        y = SortedDNAClones_BY_Peptide\n",
    "#        Xs += x\n",
    "#        Ys += y\n",
    "#        plt.plot(x, y,\n",
    "#                'o-',\n",
    "#                lw = 2.0,\n",
    "#                ms = 4.0,\n",
    "#                mew = 0.1,\n",
    "#                mec = '#191919')\n",
    "                \n",
    "#    XMin = min(Xs) - 0.05*(max(Xs) - min(Xs))\n",
    "#    XMax = max(Xs) + 0.05*(max(Xs) - min(Xs))\n",
    "#    YMin = min(Ys) - 0.05*(max(Ys) - min(Ys))\n",
    "#    YMax = max(Ys) + 0.05*(max(Ys) - min(Ys))\n",
    "    \n",
    "#    plt.axis([XMin, XMax, YMin, YMax])\n",
    "    \n",
    "#    plt.xlabel('Selection Round #', fontsize=14)\n",
    "#    plt.ylabel('DNA Clones #', fontsize=14)\n",
    "    \n",
    "#    legend = plt.legend(BaseRoundTopSortedPeptides, loc='upper center', bbox_to_anchor=(0.5, -0.15),\n",
    "#                        fancybox=True, shadow=False, ncol=2)\n",
    "    \n",
    "#    DNAMutantsAnalysisFileNamePNG = str(today) + 'DNAsMutantsAnalysis' + FileName + '.png'\n",
    "    \n",
    "#    plt.savefig(DNAMutantsAnalysisFileNamePNG, bbox_extra_artists=[legend],bbox_inches='tight', dpi = 300)\n",
    "#    plt.show()\n",
    "#    plt.close()\n",
    "#-------------------------------------------------------------------------------    \n",
    "    # Create a figure of size 8x6 inches, 500 dots per inch\n",
    "    plt.figure(figsize = (8, 6),\n",
    "               dpi = 500)\n",
    "    # Create 'ggplot' style\n",
    "    plt.style.use('fivethirtyeight')\n",
    "    # Create a new subplot from a grid of 1x1\n",
    "    Graph = plt.subplot(1, 1, 1)\n",
    "    \n",
    "#    PeptideDNAClonesNumber_IN_BaseRound = []\n",
    "#    PeptideOccurence_IN_BaseRound = []\n",
    "\n",
    "    # Map colours onto lines\n",
    "    cNorm  = matplotlib.colors.Normalize(vmin = 0,\n",
    "                                         vmax = len(BaseRoundSortedPeptides) - 1)\n",
    "    scalarMap = matplotlib.cm.ScalarMappable(norm = cNorm,\n",
    "                                             cmap = 'gist_rainbow')\n",
    "\n",
    "    RoundIndex = BaseSelectionRoundNumber\n",
    "\n",
    "    Xs = []\n",
    "    Ys = []        \n",
    "    for Peptide in DNAClones_BY_Round_BY_Peptide[RoundIndex]:\n",
    "        PeptideDNAClonesNumber_IN_BaseRound = math.log(DNAClones_BY_Round_BY_Peptide[RoundIndex].get(Peptide, 0), 2)\n",
    "        PeptideOccurence_IN_BaseRound = math.log(Peptides_BY_Round[RoundIndex].get(Peptide, 0), 2)\n",
    "        \n",
    "        PeptideColour = scalarMap.to_rgba(BaseRoundSortedPeptides.index(Peptide))\n",
    "    \n",
    "        x = PeptideDNAClonesNumber_IN_BaseRound\n",
    "        y = PeptideOccurence_IN_BaseRound\n",
    "        Xs += [x]\n",
    "        Ys += [y]\n",
    "        \n",
    "        plt.plot(x, y,\n",
    "                'o',\n",
    "                c = PeptideColour,\n",
    "                ms = 5.0,\n",
    "                mew = 0.1,\n",
    "                mec = '#191919')\n",
    "\n",
    "    XMin = min(Xs) - 0.05*(max(Xs) - min(Xs))\n",
    "    XMax = max(Xs) + 0.05*(max(Xs) - min(Xs))\n",
    "    YMin = min(Ys) - 0.05*(max(Ys) - min(Ys))\n",
    "    YMax = max(Ys) + 0.05*(max(Ys) - min(Ys))\n",
    "    \n",
    "    plt.axis([XMin, XMax, YMin, YMax])\n",
    "        \n",
    "    XLabel = 'log$2$ (DNA Clones #)' #$_$ makes subscript possible\n",
    "    plt.xlabel(XLabel, fontsize = 14)\n",
    "    YLabel = 'log$2$ (Peptide Occurence)'\n",
    "    plt.ylabel(YLabel, fontsize = 14)\n",
    "    \n",
    "    legend = plt.legend(BaseRoundSortedPeptides,\n",
    "                        loc = 'upper center',\n",
    "                        bbox_to_anchor = (0.5, -0.15),\n",
    "                        fancybox = True,\n",
    "                        shadow = False,\n",
    "                        ncol = 4)\n",
    "    \n",
    "    DNAClonesAnalysisFileNamePNG = str(today) + 'DNAsMutantsAnalysisRegression' + 'R' + str(Round) + FileName + '.png'\n",
    "    \n",
    "    plt.savefig(DNAClonesAnalysisFileNamePNG, bbox_extra_artists=[legend], bbox_inches='tight', dpi = 300)\n",
    "    plt.show()\n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def PeptidesRelatednessAnalysis(DataFolderLocation,\n",
    "                                BaseSelectionRoundNumber,\n",
    "                                TopNPeptidesNumber,\n",
    "                                StartSequence,\n",
    "                                StopSequence,\n",
    "                                cDNA_MinLength,\n",
    "                                cDNA_MaxLength,\n",
    "                                QualityScore,\n",
    "                                FileName):\n",
    "    \n",
    "    # to extract TodaysDate\n",
    "    today = TodaysDate()\n",
    "    \n",
    "    # to collect DNAs-based summary information By_Round\n",
    "    DNAs_BY_Round = DNAsOccurences_BY_Round(DataFolderLocation, cDNA_MinLength, cDNA_MaxLength, QualityScore)\n",
    "    TotalDNAs_BY_Round = TotalReads_BY_Round(DataFolderLocation, cDNA_MinLength, cDNA_MaxLength, QualityScore)\n",
    "    BaseRoundSortedDNAs = BaseRoundSortedDNAsList(DataFolderLocation, BaseSelectionRoundNumber)\n",
    "    DNAsAppearances = DNAsAppearances_BY_Round(BaseRoundSortedDNAs, DNAs_BY_Round)\n",
    "    \n",
    "    # to collect Peptides-based summary information By_Round\n",
    "    Peptides_BY_Round = PeptidesOccurences_BY_Round(DataFolderLocation, cDNA_MinLength, cDNA_MaxLength, QualityScore)\n",
    "    TotalPeptides_BY_Round = TotalReads_BY_Round(DataFolderLocation, cDNA_MinLength, cDNA_MaxLength, QualityScore)\n",
    "    BaseRoundSortedPeptides = BaseRoundSortedPeptidesList(DataFolderLocation, BaseSelectionRoundNumber)\n",
    "    PeptidesAppearances = PeptidesAppearances_BY_Round(BaseRoundSortedPeptides, Peptides_BY_Round)\n",
    "    \n",
    "    SelectionSummary = CompleteSelectionSummary(DataFolderLocation, StartSequence, StopSequence, cDNA_MinLength, cDNA_MaxLength, QualityScore)\n",
    "    SortedRoundsList = sorted(SelectionSummary.keys())\n",
    "    \n",
    "    # to create a disjoint graph (Forest), based on DNAs in the BaseRound (joint subgraphs are the Trees and the unique DNA sequences are the Leaves)\n",
    "    BaseRoundDNAsForest = nx.Graph()\n",
    "    # to add nodes (Leaves, unique DNA sequences) to the BaseRoundDNAsForest disjoint graph\n",
    "    BaseRoundDNAsForest.add_nodes_from(BaseRoundSortedDNAs)\n",
    "    # to add edges (Twigs, DNA-to-DNA connections based on the hamming distance between unique DNA sequences) to the BaseRoundDNAsForest so that disjoint graphs (stand alone Trees) can be identified\n",
    "    UsedNodes = []\n",
    "    for DNA1 in BaseRoundSortedDNAs:\n",
    "        UsedNodes += [DNA1]\n",
    "        for DNA2 in BaseRoundSortedDNAs:\n",
    "            if DNA2 not in UsedNodes and HammingDistance(DNA1, DNA2) == 1:\n",
    "                BaseRoundDNAsForest.add_edge(DNA1,DNA2,\n",
    "                                                MutationsNumber = 1)\n",
    "    # to extract individual joint subgraphs (stand alone Trees) from the disjoint graph (Forest)\n",
    "    BaseRoundDNAsTrees = list(nx.connected_component_subgraphs(BaseRoundDNAsForest, copy = True))\n",
    "    \n",
    "    # to create a PeptideSummaryDNAPerspectiveCSV file\n",
    "    PeptidesSummaryFileNameCSV =  str(today) + 'PeptideFamiliesSummary' + FileName + '.csv'\n",
    "    PeptidesSummaryFile = open(PeptidesSummaryFileNameCSV, 'w')\n",
    "    \n",
    "    # to convert list of DNAs Trees into a list of Peptides Trees Leaves\n",
    "    PeptidesTreesLeaves = []\n",
    "    for DNAsTree in BaseRoundDNAsTrees:\n",
    "        PeptideLeaves = []\n",
    "        for DNA in DNAsTree:\n",
    "            Peptide = Translation(DNA)\n",
    "            if Peptide not in PeptideLeaves:\n",
    "                PeptideLeaves += [Peptide]\n",
    "        PeptidesTreesLeaves += [PeptideLeaves]\n",
    "    # to sort the resulting list of lists from the largest to smallest\n",
    "    PeptidesTreesLeaves.sort(key = len, reverse = True)\n",
    "    \n",
    "    # to fix the coordinates of the origin of the graph\n",
    "    Positions = {}\n",
    "    X_0_Coordinate = 1\n",
    "    Y_0_Coordinate = 0\n",
    "    Y_X0_Coordinate = 0\n",
    "    \n",
    "    TreesXCoordinates = []\n",
    "    \n",
    "    \n",
    "    PeptideGraphFigure = plt.figure()\n",
    "    PeptideGraph = PeptideGraphFigure.add_subplot(1, 1, 1)\n",
    "    \n",
    "    #to introduce PeptideFamilyCounter\n",
    "    MultiplePeptideFamilyCounter = 0\n",
    "    #to introduce SinglePeptideFamilyCounter\n",
    "    SinglePeptideFamilyCounter = 0\n",
    "    #to introduce PeptideFamilySize\n",
    "    PeptideTreeSize = []\n",
    "    \n",
    "    \n",
    "    # to create a tree for each set of Peptides Trees Leaves \n",
    "    for PeptideLeaves in PeptidesTreesLeaves:\n",
    "        \n",
    "        PeptideTree = nx.Graph()\n",
    "        # to convert each Peptide (Leave) into a node of a Peptide graph (PeptideLeave on a PeptideTree)\n",
    "        \n",
    "        for Peptide in PeptideLeaves:\n",
    "            PeptideTree.add_node(Peptide,\n",
    "                                        Occurence = Peptides_BY_Round[BaseSelectionRoundNumber][Peptide],\n",
    "                                        FirstAppearance = min(PeptidesAppearances[Peptide]))\n",
    "        # to join the Peptide nodes of a graph by edges (Twigs)\n",
    "        for Peptide1 in PeptideLeaves:\n",
    "            for Peptide2 in PeptideLeaves:\n",
    "                if HammingDistance(Peptide1, Peptide2) == 1:\n",
    "                    PeptideTree.add_edge(Peptide1,Peptide2)\n",
    "        \n",
    "        # to identify the RootPeptide of a PeptideTree graph\n",
    "        TreePeptidesOccurences = nx.get_node_attributes(PeptideTree, 'Occurence')              \n",
    "        RootPeptide = max(TreePeptidesOccurences, key=TreePeptidesOccurences.get)\n",
    "        \n",
    "        # to create a dictionary holder for Peptide and their properties (Predecessor and Occurrence)\n",
    "        TreePeptides = {}\n",
    "        TreePeptides[RootPeptide] = [0, '', 0, PeptideTree.node[RootPeptide]['Occurence'], PeptideTree.node[RootPeptide]['FirstAppearance']]\n",
    "        TreePeptidesList = PeptideTree.nodes()\n",
    "        TreePeptidesList.remove(RootPeptide)\n",
    "\n",
    "        for Peptide in TreePeptidesList:\n",
    "            PeptidePredecessor = nx.shortest_path(PeptideTree, source = Peptide, target = RootPeptide, weight = None)[1]\n",
    "            # PredecessorOccurence can be used to sort the Peptides, but does not seem to be useful\n",
    "            PredecessorOccurence = PeptideTree.node[PeptidePredecessor]['Occurence']\n",
    "            PeptideOccurence = PeptideTree.node[Peptide]['Occurence']\n",
    "\n",
    "            TreePeptides[Peptide] = [PeptidePredecessor, PredecessorOccurence, PeptideOccurence]\n",
    "            \n",
    "        \n",
    "        # to sort Peptides in a PeptideTree by their distance to the RootPeptide\n",
    "        Peptides_BY_DistanceToTheRoot = {}\n",
    "        for Peptide in PeptideTree.nodes():\n",
    "            DistanceToTheRoot = nx.shortest_path_length(PeptideTree, source = Peptide, target = RootPeptide, weight = None)\n",
    "            if DistanceToTheRoot not in Peptides_BY_DistanceToTheRoot:\n",
    "                Peptides_BY_DistanceToTheRoot[DistanceToTheRoot] = [Peptide]\n",
    "            else:\n",
    "                Peptides_BY_DistanceToTheRoot[DistanceToTheRoot] += [Peptide]\n",
    "        \n",
    "        # to identify the largest group of equidistanced peptides \n",
    "        MaxPeptidesNumber = max(map(lambda k: len(Peptides_BY_DistanceToTheRoot[k]), Peptides_BY_DistanceToTheRoot))\n",
    "\n",
    "        SortedPeptides_BY_DistanceToTheRoot = {}\n",
    "        # to sort Peptides by their distance to the RootPeptide\n",
    "        for DistanceToTheRoot in Peptides_BY_DistanceToTheRoot:\n",
    "\n",
    "            EquidistantPeptides = Peptides_BY_DistanceToTheRoot[DistanceToTheRoot]\n",
    "\n",
    "            EquidistantPeptides = sorted(EquidistantPeptides, key = lambda Peptide: (TreePeptides[Peptide][2]), reverse = True)\n",
    "            # PredecessorOccurence can be used to sort the Peptides, but does not seem to be useful\n",
    "            # EquidistantPeptides = sorted(EquidistantPeptides, key = lambda Peptide: (TreePeptides[Peptide][1]), reverse = True)\n",
    "            EquidistantPeptides = sorted(EquidistantPeptides, key = lambda Peptide: (TreePeptides[Peptide][0]), reverse = False)\n",
    "\n",
    "            AdditionalElements = MaxPeptidesNumber - len(EquidistantPeptides)\n",
    "            SortedPeptides_BY_DistanceToTheRoot[DistanceToTheRoot] = EquidistantPeptides + AdditionalElements * ['']\n",
    "\n",
    "            if len(PeptideTree.nodes()) > 1:\n",
    "                for Peptide in EquidistantPeptides:\n",
    "                    XCoordinate = X_0_Coordinate + DistanceToTheRoot\n",
    "                    YCoordinate = Y_0_Coordinate - EquidistantPeptides.index(Peptide)\n",
    "                    Positions[Peptide] = (XCoordinate, YCoordinate)\n",
    "                    \n",
    "                                    \n",
    "            elif len(PeptideTree.nodes()) == 1:\n",
    "                for Peptide in EquidistantPeptides:\n",
    "                    XCoordinate = 0\n",
    "                    YCoordinate = Y_X0_Coordinate\n",
    "                    Positions[Peptide] = (XCoordinate, YCoordinate)\n",
    "                    \n",
    "\n",
    "        #BaseRoundPeptidesGraph = nx.Graph()    \n",
    "        #BaseRoundPeptidesGraph.add_nodes_from(BaseRoundSortedPeptides)\n",
    "\n",
    "        Sizes = []\n",
    "        for Peptide in PeptideTree.nodes():\n",
    "            Sizes.append(math.log(Peptides_BY_Round[BaseSelectionRoundNumber][Peptide], 2) + 5)\n",
    "\n",
    "        Colours = []\n",
    "        for Peptide in PeptideTree.nodes():\n",
    "            Colours.append(min(PeptidesAppearances[Peptide]))\n",
    "        \n",
    "        XSpan = max(map(lambda Peptide: Positions[Peptide][0], Positions)) - min(map(lambda Peptide: Positions[Peptide][0], Positions))\n",
    "        YSpan = max(map(lambda Peptide: Positions[Peptide][1], Positions)) - min(map(lambda Peptide: Positions[Peptide][1], Positions))\n",
    "                            \n",
    "        XMin = min(map(lambda Peptide: Positions[Peptide][0], Positions)) - 0.01 * XSpan\n",
    "        XMax = max(map(lambda Peptide: Positions[Peptide][0], Positions)) + 0.01 * XSpan\n",
    "        YMin = min(map(lambda Peptide: Positions[Peptide][1], Positions)) - 0.02 * YSpan\n",
    "        YMax = max(map(lambda Peptide: Positions[Peptide][1], Positions)) + 0.02 * YSpan\n",
    "        \n",
    "        \n",
    "        \n",
    "        NumberOfColours = len(SortedRoundsList)\n",
    "        \n",
    "        ColourMap = plt.get_cmap('Paired', NumberOfColours)\n",
    "        \n",
    "        nx.draw_networkx(PeptideTree,\n",
    "                        pos = Positions,\n",
    "                        node_size = Sizes,\n",
    "                        node_color = Colours,\n",
    "                        cmap = ColourMap,\n",
    "                        linewidths = 0.2,\n",
    "                        width = 0.2,\n",
    "                        with_labels = False,\n",
    "                        #font_size = 6,\n",
    "                        vmin = min(SortedRoundsList),\n",
    "                        vmax = max(SortedRoundsList))\n",
    "                    \n",
    "        if len(PeptideTree.nodes()) > 1:\n",
    "            for DistanceToTheRoot in SortedPeptides_BY_DistanceToTheRoot:\n",
    "                PeptidesSummaryFile.write(str(DistanceToTheRoot) + ' mutations' + ',' + 'frequency' + ',' + 'rank' + ',')\n",
    "            PeptidesSummaryFile.write('\\n')\n",
    "\n",
    "        for i in range(MaxPeptidesNumber):\n",
    "            for MutationsNumber in SortedPeptides_BY_DistanceToTheRoot:                        \n",
    "                Peptide = SortedPeptides_BY_DistanceToTheRoot[MutationsNumber][i]\n",
    "\n",
    "                if Peptide != '':\n",
    "                    FormatedPeptide = HammingDistanceBasedFormating(RootPeptide, Peptide)\n",
    "                    PeptideRank = str(BaseRoundSortedPeptides.index(Peptide) + 1)\n",
    "                    #ClonesNumber = str(len(PeptideTree.neighbors(Peptide)))\n",
    "                    PeptideFraction = ('{:.2%}'.format(float((Peptides_BY_Round[BaseSelectionRoundNumber].get(Peptide, 0)))/float(TotalPeptides_BY_Round[BaseSelectionRoundNumber])))\n",
    "                else:\n",
    "                    FormatedPeptide = ''\n",
    "                    #ClonesNumber = ''\n",
    "                    PeptideRank = ''\n",
    "                    PeptideFraction = ''\n",
    "\n",
    "                PeptidesSummaryFile.write(FormatedPeptide + ',' +\n",
    "                            PeptideFraction + ',' +\n",
    "                            PeptideRank + ',')\n",
    "                            #ClonesNumber + ',')\n",
    "            PeptidesSummaryFile.write('\\n')\n",
    "        \n",
    "        \n",
    "        if len(PeptideTree.nodes()) > 1:\n",
    "            TreesXCoordinates += [X_0_Coordinate]\n",
    "            X_0_Coordinate += max(Peptides_BY_DistanceToTheRoot.keys()) + 1\n",
    "            MultiplePeptideFamilyCounter += 1\n",
    "            PeptideTreeSize += [len(PeptideTree.nodes())]\n",
    "            \n",
    "            \n",
    "\n",
    "        if len(PeptideTree.nodes()) == 1:\n",
    "            Y_X0_Coordinate -= 1\n",
    "            SinglePeptideFamilyCounter += 1\n",
    "\n",
    "        PeptidesSummaryFile.write('\\n')\n",
    "                    \n",
    "    PeptidesSummaryFile.close()\n",
    "    \n",
    "    #plt.axis('off')\n",
    "    plt.axis([XMin, XMax, YMin, YMax])\n",
    "\n",
    "    PeptideLegendColour = PeptideGraphFigure.add_subplot(1, 1, 1)\n",
    "        \n",
    "    ColourMap = plt.get_cmap('Paired', NumberOfColours)\n",
    "    PeptideLegendColours = SortedRoundsList\n",
    "    \n",
    "    LegendDotsX = XMax - 0.3 * XMax\n",
    "    YIncrement = - 0.03 * YMin\n",
    "    #print (YIncrement)\n",
    "    \n",
    "    LegendColourDotsX = np.array([LegendDotsX] * NumberOfColours)\n",
    "    #print (LegendColourDotsX)\n",
    "    FirstYColours = YMin + 12 * YIncrement\n",
    "    #print (FirstYColours)\n",
    "    LastYColours = YMin + (12 + NumberOfColours) * YIncrement\n",
    "    #print (LastYColours)\n",
    "    LegendColourDotsY = np.linspace(FirstYColours, LastYColours, NumberOfColours, endpoint = False)\n",
    "    #print (LegendColourDotsY)\n",
    "    \n",
    "    PeptideLegendColour.scatter(x = LegendColourDotsX,\n",
    "                                y = LegendColourDotsY,\n",
    "                                s = 15,\n",
    "                                c = PeptideLegendColours,\n",
    "                                cmap = ColourMap,\n",
    "                                linewidths = 0.2)\n",
    "\n",
    "    ColourLabels = ['{0}'.format(i) for i in range(1,NumberOfColours + 1)]\n",
    "\n",
    "    for label, x, y in zip(ColourLabels, LegendColourDotsX, LegendColourDotsY):\n",
    "        plt.annotate(label, xy = (x, y), xytext = (5, 0),\n",
    "                     textcoords = 'offset points',\n",
    "                     fontsize = 5,\n",
    "                     ha = 'left', va = 'center')\n",
    "    plt.text(x = LegendDotsX, y = (max(LegendColourDotsY) + YIncrement),\n",
    "             s = 'first-appearance round #',\n",
    "             fontsize = 5)\n",
    "    #plt.axis('off')\n",
    "\n",
    "    PeptideLegendSize = PeptideGraphFigure.add_subplot(1, 1, 1)\n",
    "\n",
    "    Size = []\n",
    "    for i in [1, 10, 100, 1000, 10000]:\n",
    "        Size.append(math.log(i, 2) + 5)\n",
    "\n",
    "    LegendSizeDotsX = np.array([LegendDotsX] * 5)\n",
    "    FirstYSizes = YMin + 5 * YIncrement\n",
    "    LastYSizez = YMin + 10 * YIncrement\n",
    "    LegendSizeDotsY = np.linspace(FirstYSizes, LastYSizez, 5, endpoint = False)\n",
    "    PeptideLegendSize.scatter(x = LegendSizeDotsX,\n",
    "                              y = LegendSizeDotsY,\n",
    "                              s = Size,\n",
    "                              c = 'w',\n",
    "                              linewidths = 0.2)\n",
    "\n",
    "    SizeLabels = ['{0}'.format(i) for i in [1, 10, 100, 1000, 10000]]\n",
    "\n",
    "    for label, x, y in zip(SizeLabels, LegendSizeDotsX, LegendSizeDotsY):\n",
    "        plt.annotate(label, xy = (x, y), xytext = (5, 0),\n",
    "                     textcoords = 'offset points',\n",
    "                     fontsize = 5,\n",
    "                     ha = 'left', va = 'center')\n",
    "    plt.text(x = LegendDotsX, y = (max(LegendSizeDotsY) - 0.03 * YMin),\n",
    "             s = 'frequency in the last round',\n",
    "             fontsize = 5)\n",
    "    \n",
    "    \n",
    "    \n",
    "    #FamilySizeLabels = ['{0}'.format(i) for i in PeptideTreeSize]\n",
    "\n",
    "    #for label, x, y in zip(FamilySizeLabels, LegendSizeDotsX, LegendSizeDotsY):\n",
    "    #    plt.annotate(label, xy = (x, y), xytext = (5, 0),\n",
    "    #                 textcoords = 'offset points',\n",
    "    #                 fontsize = 5,\n",
    "    #                 ha = 'left', va = 'center')\n",
    "    for i in range(len(PeptideTreeSize)):\n",
    "        plt.text(x = TreesXCoordinates[i], y = YIncrement,\n",
    "                 s = PeptideTreeSize[i],\n",
    "                 fontsize = 5)\n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    plt.text(x = LegendDotsX, y = YMin + 3 * YIncrement,\n",
    "             s = ('total # unique peptide sequence ' + str(len(BaseRoundSortedPeptides))),\n",
    "             fontsize = 5)\n",
    "    plt.text(x = LegendDotsX, y = YMin + 2 * YIncrement,\n",
    "             s = 'single-member peptide family # ' + str(SinglePeptideFamilyCounter),\n",
    "             fontsize = 5)\n",
    "    plt.text(x = LegendDotsX, y = YMin + 1 * YIncrement,\n",
    "             s = 'multiple-member peptide family # ' + str(MultiplePeptideFamilyCounter),\n",
    "             fontsize = 5)\n",
    "    \n",
    "    plt.axis('off')\n",
    "\n",
    "    PeptidesSummaryFileNamePNG = str(today) + 'PeptideFamiliesSummary' + FileName + '.png'\n",
    "    plt.savefig(PeptidesSummaryFileNamePNG, dpi = 500)\n",
    "    \n",
    "    \n",
    "    fig = plt.gcf()\n",
    "    SizeInches = fig.get_size_inches()*fig.dpi\n",
    "    SizeDots = fig.get_size_inches()\n",
    "    \n",
    "    #print (SizeInches)\n",
    "    #print (SizeDots)\n",
    "    \n",
    "    \n",
    "    #print (XMin)\n",
    "    #print (XMax)\n",
    "    #print (YMin)\n",
    "    #print (YMax)\n",
    "    \n",
    "    #print (PeptideTreeSize)\n",
    "    #print (len(PeptideTreeSize))\n",
    "    #print (TreesXCoordinates)\n",
    "    #print (len(TreesXCoordinates))\n",
    "    \n",
    "    plt.show()\n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Set the parameters for the analysis\n",
    "\n",
    "DataFolderLocation = '/Users/nikita/Documents/GitHub/mRNADisplayResults_Analysis/SampleInput_RawNGSData'\n",
    "\n",
    "BaseSelectionRoundNumber = 6\n",
    "TopNPeptidesNumber = 24\n",
    "\n",
    "FileName = 'PHD2_NNK'\n",
    "\n",
    "StartSequence = 'TAATACGACTCACTATAGGGTTAACTTTAAGAAGGAGATATACATATG'    # NNK - T7g10M.F48 \n",
    "StopSequence = 'TGCGGCAGCGGCAGCGGCAGCTAGGACGGGGGGCGGAAA' #NNK - CGS3an13.R39\n",
    "#StartSequence = 'TAATACGACTCACTATAGGGTTGAACTTTAAGTAGGAGATATATCCATG'   #NNU - T7-CH-F49\n",
    "#StopSequence = 'TGTGGGTCTGGGTCTGGGTCTTAGGACGGGGGGCGGAAA'  #NNU - CGS3-CH-R39\n",
    "#StartSequence = 'ATG' # Met codon\n",
    "#StopSequence = 'TGCGGCAGC'# Akane seams to have trimmed siquences\n",
    "#StopSequence = 'TAG' # amber stop codon\n",
    "\n",
    "QualityScore = 29 # threshold QualityScore must be between 0 and 93\n",
    "cDNA_MinLength = 24\n",
    "cDNA_MaxLength = 240"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#SelectionSummaryReport(DataFolderLocation,\n",
    "#                       BaseSelectionRoundNumber,\n",
    "#                       TopNPeptidesNumber,\n",
    "#                       StartSequence,\n",
    "#                       StopSequence,\n",
    "#                       cDNA_MinLength,\n",
    "#                       cDNA_MaxLength,\n",
    "#                       QualityScore,\n",
    "#                       FileName)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#DNAMutantsAnalysis(DataFolderLocation,\n",
    "#                   BaseSelectionRoundNumber,\n",
    "#                   TopNPeptidesNumber,\n",
    "#                   StartSequence,\n",
    "#                   StopSequence,\n",
    "#                   FileName)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "PeptidesMutantsSummaryReport(DataFolderLocation,\n",
    "                             BaseSelectionRoundNumber,\n",
    "                             StartSequence,\n",
    "                             StopSequence,\n",
    "                             FileName)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#PeptidesRelatednessAnalysis(DataFolderLocation,\n",
    "#                            BaseSelectionRoundNumber,\n",
    "#                            TopNPeptidesNumber,\n",
    "#                            StartSequence,\n",
    "#                            StopSequence,\n",
    "#                            cDNA_MinLength,\n",
    "#                            cDNA_MaxLength,\n",
    "#                            QualityScore,\n",
    "#                            FileName)"
   ]
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
