{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import math\n",
    "import datetime\n",
    "import networkx as nx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def TodaysDate():\n",
    "        \n",
    "    today = datetime.date.today()\n",
    "    TodaysDate = today.strftime('%d%b%Y')\n",
    "    \n",
    "    return TodaysDate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def DNACodingSequence(DNASequence, QualityScoreSequence, StartSequence, StopSequence):\n",
    "#utilises ONLY ONE StopSequence, returns ONLY ONE CodingSequence\n",
    "    \n",
    "    QualityScoreString = \"\"\"!\"#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~\"\"\"\n",
    "    ThresholdQualityScore = 29 # ThresholdQualityScore must be between 0 and 93\n",
    "    ThresholdQualityString = QualityScoreString[ThresholdQualityScore:]\n",
    "    \n",
    "    MinLength = 24\n",
    "    MaxLength = 240\n",
    "            \n",
    "    StartIndex = DNASequence.find(StartSequence) + len(StartSequence)\n",
    "    StopIndex = DNASequence.rfind(StopSequence)\n",
    "    CodingSequence =  DNASequence[StartIndex:StopIndex]\n",
    "    if MinLength <= len(CodingSequence) and len(CodingSequence) <= MaxLength and len(CodingSequence)%3 == 0:\n",
    "        for Character in QualityScoreSequence[StartIndex:StopIndex]:\n",
    "            if Character not in ThresholdQualityString:\n",
    "                return None\n",
    "        return str(CodingSequence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def Translation(CodingSequence):\n",
    "#translates DNA sequence\n",
    "\n",
    "    TranslationCode = {'AAA':'K','AAC':'N','AAG':'K','AAU':'N',\n",
    "                    'ACA':'T','ACC':'T','ACG':'T','ACU':'T',\n",
    "                    'AGA':'R','AGC':'S','AGG':'R','AGU':'S',\n",
    "                    'AUA':'I','AUC':'I','AUG':'M','AUU':'I',\n",
    "                    \n",
    "                    'CAA':'Q','CAC':'H','CAG':'Q','CAU':'H',\n",
    "                    'CCA':'P','CCC':'P','CCG':'P','CCU':'P',\n",
    "                    'CGA':'R','CGC':'R','CGG':'R','CGU':'R',\n",
    "                    'CUA':'L','CUC':'L','CUG':'L','CUU':'L',\n",
    "                    \n",
    "                    'GAA':'E','GAC':'D','GAG':'E','GAU':'D',\n",
    "                    'GCA':'A','GCC':'A','GCG':'A','GCU':'A',\n",
    "                    'GGA':'G','GGC':'G','GGG':'G','GGU':'G',\n",
    "                    'GUA':'V','GUC':'V','GUG':'V','GUU':'V',\n",
    "                    \n",
    "                    'UAA':'#','UAC':'Y','UAG':'*','UAU':'Y',\n",
    "                    'UCA':'S','UCC':'S','UCG':'S','UCU':'S',\n",
    "                    'UGA':'&','UGC':'C','UGG':'W','UGU':'C',\n",
    "                    'UUA':'L','UUC':'F','UUG':'L','UUU':'F'}\n",
    "    # UAA (ochre) — #\n",
    "    # UAG (amber) — *\n",
    "    # UGA (opal) — &\n",
    "                    \n",
    "    TranscriptionCode = {'A':'A','C':'C','G':'G','T':'U','U':'T'}\n",
    "      \n",
    "    RNASequence = ''\n",
    "    for Nucleotide in CodingSequence:\n",
    "        RNASequence += TranscriptionCode.get(Nucleotide,'X')\n",
    "    #converts DNA to RNA\n",
    "    #print RNASequence\n",
    "        \n",
    "    Peptide = ''\n",
    "    while len(RNASequence) != 0:\n",
    "        Peptide += TranslationCode.get(RNASequence[0:3],'Do not fuck with me!')\n",
    "        RNASequence = RNASequence[3:]\n",
    "    return Peptide"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def SingleSelectionRoundSummary(fastqFileLocation):\n",
    "#returns a list of lists with peptide-sequences and their frequencies, sorted by frequency in descending order\n",
    "    \n",
    "    RawDataFile = open(fastqFileLocation, 'r')\n",
    "    Lines = RawDataFile.readlines()\n",
    "    RawDataFile.close\n",
    "    \n",
    "    #StartSequence = 'ATG' # Met codon\n",
    "    #StopSequence = 'TAG' # amber stop codon\n",
    "    \n",
    "    StartSequence = 'TAATACGACTCACTATAGGGTTAACTTTAAGAAGGAGATATACATATG'    # NNK - T7g10M.F48 \n",
    "    StopSequence = 'TGCGGCAGCGGCAGCGGCAGCTAGGACGGGGGGCGGAAA' #NNK - CGS3an13.R39 \n",
    "    #StartSequence = 'TAATACGACTCACTATAGGGTTGAACTTTAAGTAGGAGATATATCCATG'   #NNU - T7-CH-F49\n",
    "    #StopSequence = 'TGTGGGTCTGGGTCTGGGTCTTAGGACGGGGGGCGGAAA'  #NNU - CGS3-CH-R39\n",
    "    \n",
    "    SingleSelectionRoundSummary = {}\n",
    "    #creates empty SingleSelectionRoundSummary dictionary to store the results from a single round of selection\n",
    "    #SingleSelectionRoundSummary = {PeptideY:    {CodingSequence_YZ:    Occurence_YZ}}\n",
    "        \n",
    "    #populates SingleSelectionRoundSummary dictionary with the results from a single round of selection\n",
    "    for i,Line in enumerate(Lines):\n",
    "        if StartSequence in Line and StopSequence in Line:\n",
    "            CodingSequence = DNACodingSequence(Line, Lines[i + 2], StartSequence, StopSequence)\n",
    "            if CodingSequence != None:\n",
    "                PeptideSequence = Translation(CodingSequence)\n",
    "                if PeptideSequence not in SingleSelectionRoundSummary:\n",
    "                    SingleSelectionRoundSummary[str(PeptideSequence)] = {str(CodingSequence) : 1}\n",
    "                else:\n",
    "                    if CodingSequence not in SingleSelectionRoundSummary[str(PeptideSequence)]:\n",
    "                        SingleSelectionRoundSummary[str(PeptideSequence)][str(CodingSequence)] = 1\n",
    "                    else:\n",
    "                        SingleSelectionRoundSummary[str(PeptideSequence)][str(CodingSequence)] += 1\n",
    "\n",
    "    return SingleSelectionRoundSummary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def HammingDistance(Sequence1, Sequence2):\n",
    "    \n",
    "    if len(Sequence1) < len(Sequence2):\n",
    "        Sequence1 = Sequence1 + (len(Sequence2) - len(Sequence1)) * '%'\n",
    "    elif len(Sequence1) > len(Sequence2):\n",
    "        Sequence2 = Sequence2 + (len(Sequence1) - len(Sequence2)) * '%'\n",
    "    \n",
    "    HammingDistance = 0\n",
    "    for i in range(len(Sequence1)):\n",
    "        if Sequence1[i] == Sequence2[i]:\n",
    "            HammingDistance = HammingDistance\n",
    "        else:\n",
    "            HammingDistance = HammingDistance + 1\n",
    "            \n",
    "    return HammingDistance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def HammingDistanceBasedFormating(Sequence1, Sequence2):\n",
    "    \n",
    "    if len(Sequence1) < len(Sequence2):\n",
    "        Sequence1 = Sequence1 + (len(Sequence2) - len(Sequence1)) * '-'\n",
    "    elif len(Sequence1) > len(Sequence2):\n",
    "        Sequence2 = Sequence2 + (len(Sequence1) - len(Sequence2)) * '-'\n",
    "    \n",
    "    HammingDistance = 0\n",
    "    FormatedSequence2 = ''\n",
    "    for i in range(len(Sequence1)):\n",
    "        if Sequence1[i] == Sequence2[i]:\n",
    "            FormatedSequence2 += Sequence2[i].lower()\n",
    "            HammingDistance = HammingDistance\n",
    "        else:\n",
    "            FormatedSequence2 += Sequence2[i]\n",
    "            HammingDistance = HammingDistance + 1            \n",
    "    return FormatedSequence2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def CompleteSelectionSummary(fastqDataFolderLocation):\n",
    "# returns a SelectionSummary dictionary with the following structure\n",
    "# SelectionSummary = {SelectionRound_X:    {PeptideXY:    {CodingDNA_XYZ:    Occurence_XYZ}}}\n",
    "\n",
    "    CompleteSelectionSummary = {}\n",
    "    # creates empty SelectionSummary dictionary to store the results from all the rounds of selection\n",
    "\n",
    "    import os           \n",
    "    for file in os.listdir(fastqDataFolderLocation):\n",
    "        \n",
    "        FileLocation = os.path.join(fastqDataFolderLocation, file)\n",
    "          \n",
    "        if file.endswith('.fastq'): # this conditional is necessary; without it some shit appears in the beginning of the file list\n",
    "            RoundNumberFirstDigit = file[file.find('.')-2]\n",
    "            RoundNumberSecondDigit = file[file.find('.')-1]\n",
    "            if RoundNumberFirstDigit == '0':\n",
    "                RoundNumber = int(RoundNumberSecondDigit)\n",
    "                #print RoundNumber\n",
    "            elif RoundNumberFirstDigit != '0':\n",
    "                RoundNumber = int(file[file.find('.')-2 : file.find('.')])\n",
    "                #print RoundNumber\n",
    "        #(1.A) extracts the round number from the file name (file name should have two digit number before full stop — '00.') \n",
    "                \n",
    "            SelectionRoundSummary = SingleSelectionRoundSummary(FileLocation)\n",
    "            #(1.B) extracts single round results \n",
    "                    \n",
    "            CompleteSelectionSummary[RoundNumber] = SelectionRoundSummary\n",
    "            #(1.C) populate ConcatenatedResultsList\n",
    "            #print ConcatenatedResultsList\n",
    "            \n",
    "    return CompleteSelectionSummary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def PeptidesOccurences_BY_Round(fastqDataFolderLocation):\n",
    "    SelectionSummary = CompleteSelectionSummary(fastqDataFolderLocation)\n",
    "    \n",
    "    PeptidesOccurences_BY_Round = {}\n",
    "    for Round in SelectionSummary:\n",
    "        PeptidesOccurences_IN_Round = {}\n",
    "        for Peptide in SelectionSummary[Round]:\n",
    "            PeptidesOccurences_IN_Round[Peptide] = sum(SelectionSummary[Round][Peptide].values())\n",
    "        PeptidesOccurences_BY_Round[Round] = PeptidesOccurences_IN_Round\n",
    "        \n",
    "    return PeptidesOccurences_BY_Round"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def DNAsOccurences_BY_Round(fastqDataFolderLocation):\n",
    "    SelectionSummary = CompleteSelectionSummary(fastqDataFolderLocation)\n",
    "    \n",
    "    DNAsOccurences_BY_Round = {}\n",
    "    for Round in SelectionSummary:\n",
    "        DNAsOccurences_IN_Round = {}\n",
    "        for Peptide in SelectionSummary[Round]:\n",
    "            for DNA in SelectionSummary[Round][Peptide]:\n",
    "                DNAsOccurences_IN_Round[DNA] = SelectionSummary[Round][Peptide][DNA]\n",
    "        DNAsOccurences_BY_Round[Round] = DNAsOccurences_IN_Round\n",
    "\n",
    "    return DNAsOccurences_BY_Round"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def TotalReads_BY_Round(fastqDataFolderLocation):\n",
    "    SelectionSummary = CompleteSelectionSummary(fastqDataFolderLocation)\n",
    "    Peptides_BY_Round = PeptidesOccurences_BY_Round(fastqDataFolderLocation)\n",
    "    \n",
    "    TotalReads_BY_Round = {}\n",
    "    for Round in SelectionSummary:\n",
    "        TotalReads_BY_Round[Round] = sum(Peptides_BY_Round[Round].values())\n",
    "        \n",
    "    return TotalReads_BY_Round"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def BaseRoundSortedPeptidesList(fastqDataFolderLocation, BaseRoundIndex):\n",
    "    Peptides_BY_Round = PeptidesOccurences_BY_Round(fastqDataFolderLocation)  \n",
    "            \n",
    "    PeptidesOccurencesInBaseRound = Peptides_BY_Round[BaseRoundIndex]\n",
    "    BaseRoundSortedPeptidesList = sorted(PeptidesOccurencesInBaseRound, key = PeptidesOccurencesInBaseRound.get, reverse = True)\n",
    "    \n",
    "    return BaseRoundSortedPeptidesList"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def BaseRoundSortedDNAsList(fastqDataFolderLocation, BaseRoundIndex):\n",
    "    DNAs_BY_Round = DNAsOccurences_BY_Round(fastqDataFolderLocation)  \n",
    "            \n",
    "    DNAsOccurences_IN_BaseRound = DNAs_BY_Round[BaseRoundIndex]\n",
    "    BaseRoundSortedDNAsList = sorted(DNAsOccurences_IN_BaseRound, key = DNAsOccurences_IN_BaseRound.get, reverse = True)\n",
    "    \n",
    "    return BaseRoundSortedDNAsList"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def DNAClonesOccurences_BY_Round_BY_Peptide(fastqDataFolderLocation):\n",
    "    SelectionSummary = CompleteSelectionSummary(fastqDataFolderLocation)\n",
    "    \n",
    "    DNAClonesOccurences_BY_Round_BY_Peptide = {}\n",
    "    for Round in SelectionSummary:\n",
    "        DNAClonesOccurences_BY_Peptide = {}\n",
    "        for Peptide in SelectionSummary[Round]:\n",
    "            DNAClonesOccurences_BY_Peptide[Peptide] = len(SelectionSummary[Round][Peptide])\n",
    "        DNAClonesOccurences_BY_Round_BY_Peptide[Round] = DNAClonesOccurences_BY_Peptide\n",
    "        \n",
    "    return DNAClonesOccurences_BY_Round_BY_Peptide"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def PeptidesAppearances_BY_Round(BaseRoundSortedPeptidesList, PeptidesOccurences_BY_Round):\n",
    "    \n",
    "    PeptidesAppearances_BY_Round = {}\n",
    "    \n",
    "    for Peptide in BaseRoundSortedPeptidesList:\n",
    "        PeptidesAppearances_BY_Round[Peptide] = []\n",
    "        for Round in PeptidesOccurences_BY_Round:\n",
    "            if Peptide in PeptidesOccurences_BY_Round[Round]:\n",
    "                PeptidesAppearances_BY_Round[Peptide] += [Round]\n",
    "    return PeptidesAppearances_BY_Round"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def DNAsAppearances_BY_Round(BaseRoundSortedDNAsList, DNAsOccurences_BY_Round):\n",
    "    \n",
    "    DNAsAppearances_BY_Round = {}\n",
    "    \n",
    "    for DNA in BaseRoundSortedDNAsList:\n",
    "        DNAsAppearances_BY_Round[DNA] = []\n",
    "        for Round in DNAsOccurences_BY_Round:\n",
    "            if DNA in DNAsOccurences_BY_Round[Round]:\n",
    "                DNAsAppearances_BY_Round[DNA] += [Round]\n",
    "    return DNAsAppearances_BY_Round"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def SelectionSummaryReport(fastqDataFolderLocation, BaseRoundIndex, NumberOfTopPeptides, SelectionSummaryReportFileName):\n",
    "    \n",
    "    today = TodaysDate() \n",
    "    \n",
    "    SelectionSummaryFileNameCSV = str(today) + 'SelectionSummary' + SelectionSummaryReportFileName + '.csv'\n",
    "    SelectionSummaryReportFile = open(SelectionSummaryFileNameCSV, 'w')\n",
    "    \n",
    "    SelectionSummary = CompleteSelectionSummary(fastqDataFolderLocation)\n",
    "    SortedRoundsList = sorted(SelectionSummary.keys())\n",
    "    \n",
    "    Peptides_BY_Round = PeptidesOccurences_BY_Round(fastqDataFolderLocation)\n",
    "    TotalPeptides_BY_Round = TotalReads_BY_Round(fastqDataFolderLocation)\n",
    "    \n",
    "    BaseRoundSortedPeptides = BaseRoundSortedPeptidesList(fastqDataFolderLocation, BaseRoundIndex)\n",
    "    BaseRoundTopSortedPeptides = BaseRoundSortedPeptides[0 : (NumberOfTopPeptides)] \n",
    "    \n",
    "    SelectionSummaryReportFile.write('peptide sequence' + ',')\n",
    "    for Round in SortedRoundsList:\n",
    "        SelectionSummaryReportFile.write('round # ' + str(Round) + ' occurence (#)' + ',')\n",
    "    SelectionSummaryReportFile.write('\\n')\n",
    "    \n",
    "    for Peptide in BaseRoundTopSortedPeptides:\n",
    "        SelectionSummaryReportFile.write(Peptide + ',')\n",
    "        for Round in SortedRoundsList:\n",
    "            SelectionSummaryReportFile.write(str(Peptides_BY_Round[Round].get(Peptide, 0)) + ',')\n",
    "        SelectionSummaryReportFile.write('\\n')\n",
    "        \n",
    "    SelectionSummaryReportFile.write('total #' + ',')\n",
    "    for Round in SortedRoundsList:\n",
    "        SelectionSummaryReportFile.write(str(TotalPeptides_BY_Round[Round]) + ',')\n",
    "    SelectionSummaryReportFile.write('\\n\\n\\n')\n",
    "    \n",
    "    SelectionSummaryReportFile.write('peptide sequence' + ',')\n",
    "    for Round in SortedRoundsList:\n",
    "        SelectionSummaryReportFile.write('round # ' + str(Round) + ' fraction (%)' + ',')\n",
    "    SelectionSummaryReportFile.write('\\n')\n",
    "    \n",
    "    for Peptide in BaseRoundTopSortedPeptides:\n",
    "        SelectionSummaryReportFile.write(Peptide + ',')\n",
    "        for Round in SortedRoundsList:\n",
    "            PeptideFraction = float((Peptides_BY_Round[Round].get(Peptide, 0)))/float(TotalPeptides_BY_Round[Round])\n",
    "            SelectionSummaryReportFile.write('{:.3%}'.format(PeptideFraction) + ',')\n",
    "        SelectionSummaryReportFile.write('\\n')\n",
    "            \n",
    "    SelectionSummaryReportFile.close()\n",
    "    \n",
    "#-------------------------------------------------------------------------------\n",
    "   \n",
    "    plt.style.use('fivethirtyeight') # just to create 'ggplot' style\n",
    "    \n",
    "    Xs = []\n",
    "    Ys = []\n",
    "    for Peptide in BaseRoundTopSortedPeptides:\n",
    "        PeptidesFractions_BY_Round = []\n",
    "        for Round in SortedRoundsList:\n",
    "            PeptidesFractions_BY_Round += [float((Peptides_BY_Round[Round].get(Peptide, 0)))/float(TotalPeptides_BY_Round[Round])]\n",
    "        \n",
    "        x = SortedRoundsList\n",
    "        y = PeptidesFractions_BY_Round\n",
    "        Xs += x\n",
    "        Ys += y\n",
    "        \n",
    "        plt.plot(x, y,\n",
    "                    'o-',\n",
    "                    lw = 2.0,\n",
    "                    ms = 4.0,\n",
    "                    mew = 0.1,\n",
    "                    mec = '#191919')\n",
    "    \n",
    "    XMin = min(Xs) - 0.05*(max(Xs) - min(Xs))\n",
    "    XMax = max(Xs) + 0.05*(max(Xs) - min(Xs))\n",
    "    YMin = min(Ys) - 0.05*(max(Ys) - min(Ys))\n",
    "    YMax = max(Ys) + 0.05*(max(Ys) - min(Ys))\n",
    "    \n",
    "    plt.axis([XMin, XMax, YMin, YMax])\n",
    "    \n",
    "    plt.xlabel('Selection Round #', fontsize=14)\n",
    "    plt.ylabel('Peptide Fraction', fontsize=14)\n",
    "    \n",
    "    legend = plt.legend(BaseRoundTopSortedPeptides, loc='upper center', bbox_to_anchor=(0.5, -0.15),\n",
    "                        fancybox=True, shadow=False, ncol=2)\n",
    "    \n",
    "    SelectionSummaryFileNamePNG = str(today) + 'SelectionSummary' + SelectionSummaryReportFileName + '.png'\n",
    "    \n",
    "    plt.savefig(SelectionSummaryFileNamePNG, bbox_extra_artists=[legend],bbox_inches='tight', dpi = 300)\n",
    "    plt.show()\n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def DNAMutantsAnalysis(fastqDataFolderLocation, BaseRoundIndex, NumberOfTopPeptides, SelectionSummaryReportFileName):\n",
    "    \n",
    "    today = TodaysDate() \n",
    "    \n",
    "    DNAMutantsAnalysisFileNameCSV =  str(today) + 'DNAsMutantsAnalysis' + SelectionSummaryReportFileName + '.csv'\n",
    "    DNAMutantsAnalysisFile = open(DNAMutantsAnalysisFileNameCSV, 'w')\n",
    "    \n",
    "    SelectionSummary = CompleteSelectionSummary(fastqDataFolderLocation)\n",
    "    SortedRoundsList = sorted(SelectionSummary.keys())\n",
    "    \n",
    "    Peptides_BY_Round = PeptidesOccurences_BY_Round(fastqDataFolderLocation)\n",
    "    TotalPeptides_BY_Round = TotalReads_BY_Round(fastqDataFolderLocation)\n",
    "    \n",
    "    BaseRoundSortedPeptides = BaseRoundSortedPeptidesList(fastqDataFolderLocation, BaseRoundIndex)\n",
    "    BaseRoundTopSortedPeptides = BaseRoundSortedPeptides[0 : (NumberOfTopPeptides)]\n",
    "    \n",
    "    DNAClones_BY_Round_BY_Peptide = DNAClonesOccurences_BY_Round_BY_Peptide(fastqDataFolderLocation)\n",
    "    \n",
    "    DNAMutantsAnalysisFile.write('peptide sequence' + ',')\n",
    "    for Round in SortedRoundsList:\n",
    "        DNAMutantsAnalysisFile.write('round # ' + str(Round) + ' DNA clones (#)' + ',')\n",
    "    DNAMutantsAnalysisFile.write('\\n')\n",
    "    \n",
    "    for Peptide in BaseRoundTopSortedPeptides:\n",
    "        DNAMutantsAnalysisFile.write(Peptide + ',')\n",
    "        for Round in SortedRoundsList:\n",
    "            DNAMutantsAnalysisFile.write(str(DNAClones_BY_Round_BY_Peptide[Round].get(Peptide, 0)) + ',')\n",
    "        DNAMutantsAnalysisFile.write('\\n')\n",
    "            \n",
    "    DNAMutantsAnalysisFile.close()\n",
    "    \n",
    "#-------------------------------------------------------------------------------        \n",
    "    plt.style.use('fivethirtyeight') # just to create 'ggplot' style\n",
    "    \n",
    "    Xs = []\n",
    "    Ys = []    \n",
    "    for Peptide in BaseRoundTopSortedPeptides:\n",
    "        SortedDNAClones_BY_Peptide = []\n",
    "        for Round in SortedRoundsList:\n",
    "            SortedDNAClones_BY_Peptide += [DNAClones_BY_Round_BY_Peptide[Round].get(Peptide, 0)]\n",
    "        \n",
    "        x = SortedRoundsList\n",
    "        y = SortedDNAClones_BY_Peptide\n",
    "        Xs += x\n",
    "        Ys += y\n",
    "        plt.plot(x, y,\n",
    "                'o-',\n",
    "                lw = 2.0,\n",
    "                ms = 4.0,\n",
    "                mew = 0.1,\n",
    "                mec = '#191919')\n",
    "                \n",
    "    XMin = min(Xs) - 0.05*(max(Xs) - min(Xs))\n",
    "    XMax = max(Xs) + 0.05*(max(Xs) - min(Xs))\n",
    "    YMin = min(Ys) - 0.05*(max(Ys) - min(Ys))\n",
    "    YMax = max(Ys) + 0.05*(max(Ys) - min(Ys))\n",
    "    \n",
    "    plt.axis([XMin, XMax, YMin, YMax])\n",
    "    \n",
    "    plt.xlabel('Selection Round #', fontsize=14)\n",
    "    plt.ylabel('DNA Clones #', fontsize=14)\n",
    "    \n",
    "    legend = plt.legend(BaseRoundTopSortedPeptides, loc='upper center', bbox_to_anchor=(0.5, -0.15),\n",
    "                        fancybox=True, shadow=False, ncol=2)\n",
    "    \n",
    "    DNAMutantsAnalysisFileNamePNG = str(today) + 'DNAsMutantsAnalysis' + SelectionSummaryReportFileName + '.png'\n",
    "    \n",
    "    plt.savefig(DNAMutantsAnalysisFileNamePNG, bbox_extra_artists=[legend],bbox_inches='tight', dpi = 300)\n",
    "    plt.show()\n",
    "    plt.close()\n",
    "#-------------------------------------------------------------------------------    \n",
    "    plt.style.use('fivethirtyeight')\n",
    "    \n",
    "#    PeptideDNAClonesNumber_IN_BaseRound = []\n",
    "#    PeptideOccurence_IN_BaseRound = []\n",
    "\n",
    "    RoundIndex = BaseRoundIndex\n",
    "\n",
    "    Xs = []\n",
    "    Ys = []        \n",
    "    for Peptide in DNAClones_BY_Round_BY_Peptide[RoundIndex]:\n",
    "        PeptideDNAClonesNumber_IN_BaseRound = math.log(DNAClones_BY_Round_BY_Peptide[RoundIndex].get(Peptide, 0), 2)\n",
    "        PeptideOccurence_IN_BaseRound = math.log(Peptides_BY_Round[RoundIndex].get(Peptide, 0), 2)\n",
    "    \n",
    "        x = PeptideDNAClonesNumber_IN_BaseRound\n",
    "        y = PeptideOccurence_IN_BaseRound\n",
    "        Xs += [x]\n",
    "        Ys += [y]\n",
    "        \n",
    "        plt.plot(x, y,\n",
    "                'o',\n",
    "                ms = 5.0,\n",
    "                mew = 0.1,\n",
    "                mec = '#191919')\n",
    "\n",
    "    XMin = min(Xs) - 0.05*(max(Xs) - min(Xs))\n",
    "    XMax = max(Xs) + 0.05*(max(Xs) - min(Xs))\n",
    "    YMin = min(Ys) - 0.05*(max(Ys) - min(Ys))\n",
    "    YMax = max(Ys) + 0.05*(max(Ys) - min(Ys))\n",
    "    \n",
    "    plt.axis([XMin, XMax, YMin, YMax])\n",
    "        \n",
    "    XLabel = 'log$2$ (DNA Clones #)' #$_$ makes subscript possible\n",
    "    plt.xlabel(XLabel, fontsize = 14)\n",
    "    YLabel = 'log$2$ (Peptide Occurence)'\n",
    "    plt.ylabel(YLabel, fontsize = 14)\n",
    "    \n",
    "#    legend = plt.legend(BaseRoundTopSortedPeptides, loc='upper center', bbox_to_anchor=(0.5, -0.15),\n",
    "#                        fancybox=True, shadow=False, ncol=2)\n",
    "    \n",
    "    DNAClonesAnalysisFileNamePNG = str(today) + 'DNAsMutantsAnalysisRegression' + 'R' + str(Round) + SelectionSummaryReportFileName + '.png'\n",
    "    \n",
    "    plt.savefig(DNAClonesAnalysisFileNamePNG, bbox_extra_artists=[legend], bbox_inches='tight', dpi = 300)\n",
    "    plt.show()\n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "def PeptidesMutantsSummaryReport(fastqDataFolderLocation, BaseRoundIndex, PeptidesClonesSummaryReportFileName):\n",
    "    \n",
    "    today = TodaysDate()\n",
    "    \n",
    "    Peptides_BY_Round = PeptidesOccurences_BY_Round(fastqDataFolderLocation)\n",
    "    TotalPeptides_BY_Round = TotalReads_BY_Round(fastqDataFolderLocation)\n",
    "\n",
    "    BaseRoundSortedPeptides = BaseRoundSortedPeptidesList(fastqDataFolderLocation, BaseRoundIndex)\n",
    "    PeptidesAppearances = PeptidesAppearances_BY_Round(BaseRoundSortedPeptides, Peptides_BY_Round)\n",
    "    \n",
    "    BaseRoundPeptidesForest = nx.Graph()\n",
    "    for Peptide in BaseRoundSortedPeptides:\n",
    "        BaseRoundPeptidesForest.add_node(Peptide,\n",
    "                                        Occurence = Peptides_BY_Round[BaseRoundIndex][Peptide],\n",
    "                                        FirstAppearance = min(PeptidesAppearances[Peptide]))\n",
    "    # adds nodes with attributes of Occurence in the BaseRound and FirstAppearance to the BaseRoundPeptidesForest\n",
    "    UsedNodes = []\n",
    "    for Peptide1 in BaseRoundSortedPeptides:\n",
    "        UsedNodes += [Peptide1]\n",
    "        for Peptide2 in BaseRoundSortedPeptides:\n",
    "            if Peptide2 not in UsedNodes and HammingDistance(Peptide1, Peptide2) == 1:\n",
    "                BaseRoundPeptidesForest.add_edge(Peptide1,Peptide2,\n",
    "                                                MutationsNumber = 1)\n",
    "    # adds edges to the BaseRoundPeptidesForest so that it can be made into a directed graph\n",
    "    \n",
    "    BaseRoundPeptidesTrees = list(nx.connected_component_subgraphs(BaseRoundPeptidesForest, copy = True))\n",
    "    \n",
    "    PeptidesClonesSummaryReportFileNameCSV =  str(today) + 'PeptidesMutantsSummary' + PeptidesClonesSummaryReportFileName + '.csv'\n",
    "    PeptidesClonesSummaryReportFile = open(PeptidesClonesSummaryReportFileNameCSV, 'w')\n",
    "    \n",
    "    Positions = {}\n",
    "    X_0_Coordinate = 1\n",
    "    Y_0_Coordinate = 0\n",
    "    Y_X0_Coordinate = 0\n",
    "    \n",
    "    for Tree in BaseRoundPeptidesTrees:\n",
    "                        \n",
    "        RootPeptide = max(nx.get_node_attributes(Tree, 'Occurence'),\n",
    "                        key = lambda Peptide: Tree[Peptide])\n",
    "        \n",
    "        TreePeptides = {}\n",
    "        TreePeptides[RootPeptide] = [0, '', 0, Tree.node[RootPeptide]['Occurence'], Tree.node[RootPeptide]['FirstAppearance']]\n",
    "        TreePeptidesList = Tree.nodes()\n",
    "        TreePeptidesList.remove(RootPeptide)\n",
    "        \n",
    "        for Peptide in TreePeptidesList:\n",
    "            PeptidePredecessor = nx.shortest_path(Tree, source = Peptide, target = RootPeptide, weight = None)[1]\n",
    "            PredecessorOccurence = Tree.node[PeptidePredecessor]['Occurence']\n",
    "            PeptideOccurence = Tree.node[Peptide]['Occurence']\n",
    "            \n",
    "            TreePeptides[Peptide] = [PeptidePredecessor, PredecessorOccurence, PeptideOccurence]\n",
    "        \n",
    "        Peptides_BY_DistanceToTheRoot = {}\n",
    "        for Peptide in Tree.nodes():\n",
    "            DistanceToTheRoot = nx.shortest_path_length(Tree, source = Peptide, target = RootPeptide, weight = None)\n",
    "            if DistanceToTheRoot not in Peptides_BY_DistanceToTheRoot:\n",
    "                Peptides_BY_DistanceToTheRoot[DistanceToTheRoot] = [Peptide]\n",
    "            else:\n",
    "                Peptides_BY_DistanceToTheRoot[DistanceToTheRoot] += [Peptide]\n",
    "                \n",
    "        MaxPeptideMutantsNumber = max(map(lambda k: len(Peptides_BY_DistanceToTheRoot[k]), Peptides_BY_DistanceToTheRoot))\n",
    "        \n",
    "        SortedPeptides_BY_DistanceToTheRoot = {}\n",
    "        \n",
    "        for DistanceToTheRoot in Peptides_BY_DistanceToTheRoot:\n",
    "            \n",
    "            EquidistantPeptides = Peptides_BY_DistanceToTheRoot[DistanceToTheRoot]\n",
    "            \n",
    "            EquidistantPeptides = sorted(EquidistantPeptides, key = lambda Peptide: (TreePeptides[Peptide][2]), reverse = True)\n",
    "#            EquidistantPeptides = sorted(EquidistantPeptides, key = lambda Peptide: (TreePeptides[Peptide][1]), reverse = True)\n",
    "            EquidistantPeptides = sorted(EquidistantPeptides, key = lambda Peptide: (TreePeptides[Peptide][0]), reverse = False)\n",
    "            \n",
    "            AdditionalElements = MaxPeptideMutantsNumber - len(EquidistantPeptides)\n",
    "            SortedPeptides_BY_DistanceToTheRoot[DistanceToTheRoot] = EquidistantPeptides + AdditionalElements * ['']\n",
    "            \n",
    "            if len(Tree.nodes()) > 1:\n",
    "                for Peptide in EquidistantPeptides:\n",
    "                    XCoordinate = X_0_Coordinate + DistanceToTheRoot\n",
    "                    YCoordinate = Y_0_Coordinate - EquidistantPeptides.index(Peptide)\n",
    "                    Positions[Peptide] = (XCoordinate, YCoordinate)\n",
    "            elif len(Tree.nodes()) == 1:\n",
    "                for Peptide in EquidistantPeptides:\n",
    "                    XCoordinate = 0\n",
    "                    YCoordinate = Y_X0_Coordinate\n",
    "                    Positions[Peptide] = (XCoordinate, YCoordinate)\n",
    "        \n",
    "                            \n",
    "        if len(Tree.nodes()) > 1:\n",
    "            X_0_Coordinate += max(Peptides_BY_DistanceToTheRoot.keys()) + 1\n",
    "        \n",
    "        if len(Tree.nodes()) == 1:\n",
    "            Y_X0_Coordinate -= 1\n",
    "            \n",
    "            \n",
    "#        print SortedPeptides_BY_DistanceToTheRoot\n",
    "        \n",
    "\n",
    "#-------------------------------------------------------------------------------\n",
    "\n",
    "        if len(Tree.nodes()) > 1:\n",
    "            for DistanceToTheRoot in SortedPeptides_BY_DistanceToTheRoot:\n",
    "                PeptidesClonesSummaryReportFile.write(str(DistanceToTheRoot) + ' mutations' + ',' + 'frequency' + ',' + 'clones #' + ',' )\n",
    "            PeptidesClonesSummaryReportFile.write('\\n')\n",
    "        \n",
    "        for i in range(MaxPeptideMutantsNumber):\n",
    "            for MutationsNumber in SortedPeptides_BY_DistanceToTheRoot:                        \n",
    "                Peptide = SortedPeptides_BY_DistanceToTheRoot[MutationsNumber][i]\n",
    "                \n",
    "                if Peptide != '':\n",
    "                    FormatedPeptide = HammingDistanceBasedFormating(RootPeptide, Peptide)\n",
    "                    ClonesNumber = str(len(Tree.neighbors(Peptide)))\n",
    "                    PeptideFraction = ('{:.2%}'.format(float((Peptides_BY_Round[BaseRoundIndex].get(Peptide, 0)))/float(TotalPeptides_BY_Round[BaseRoundIndex])))\n",
    "                else:\n",
    "                    FormatedPeptide = ''\n",
    "                    ClonesNumber = ''\n",
    "                    PeptideFraction = ''\n",
    "                                                        \n",
    "                PeptidesClonesSummaryReportFile.write(FormatedPeptide + ',' +\n",
    "                            PeptideFraction + ',' +\n",
    "                            ClonesNumber + ',')\n",
    "            PeptidesClonesSummaryReportFile.write('\\n')\n",
    "        PeptidesClonesSummaryReportFile.write('\\n')\n",
    "                    \n",
    "    PeptidesClonesSummaryReportFile.close()\n",
    "        \n",
    "    BaseRoundPeptidesGraph = nx.Graph()    \n",
    "    BaseRoundPeptidesGraph.add_nodes_from(BaseRoundSortedPeptides)\n",
    "     \n",
    "    Sizes = []\n",
    "    for Peptide in BaseRoundPeptidesGraph.nodes():\n",
    "        Sizes.append(5 * math.log(Peptides_BY_Round[BaseRoundIndex][Peptide], 10) + 2)\n",
    "    Colours = []\n",
    "    for Peptide in BaseRoundPeptidesGraph.nodes():\n",
    "        Colours.append(min(PeptidesAppearances[Peptide]))\n",
    "    \n",
    "#    print(BaseRoundPeptidesGraph.nodes())\n",
    "    i = 0\n",
    "    for Peptide1 in BaseRoundSortedPeptides:\n",
    "        for Peptide2 in BaseRoundSortedPeptides:\n",
    "            if HammingDistance(Peptide1, Peptide2) == 1:\n",
    "                BaseRoundPeptidesGraph.add_edge(Peptide1,Peptide2)\n",
    "        i += 1\n",
    "\n",
    "    XMin = min(map(lambda Peptide: Positions[Peptide][0], Positions)) - 1\n",
    "    XMax = max(map(lambda Peptide: Positions[Peptide][0], Positions)) + 1\n",
    "    YMin = min(map(lambda Peptide: Positions[Peptide][1], Positions)) - 1\n",
    "    YMax = max(map(lambda Peptide: Positions[Peptide][1], Positions)) + 1\n",
    "    \n",
    "    nx.draw_networkx(BaseRoundPeptidesGraph,\n",
    "                    pos = Positions,\n",
    "                    node_size = Sizes,\n",
    "                    node_color = Colours,\n",
    "                    cmap = 'Paired',\n",
    "                    linewidths = 0.2,\n",
    "                    width = 0.2,\n",
    "                    with_labels = False,\n",
    "                    font_size = 6)\n",
    "\n",
    "    plt.axis('off')\n",
    "    plt.axis([XMin, XMax, YMin, YMax])\n",
    "    \n",
    "    PeptidesMutantsSummaryReportFileNamePNG = str(today) + 'PeptidesMutantsSummary' + PeptidesClonesSummaryReportFileName + '.png'\n",
    "    plt.savefig(PeptidesMutantsSummaryReportFileNamePNG, dpi = 500)\n",
    "    \n",
    "    plt.show()\n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def DNAsMutantsSummaryReport(fastqDataFolderLocation, BaseRoundIndex, DNAsClonesSummaryReportFileName):\n",
    "    \n",
    "    today = TodaysDate()\n",
    "    \n",
    "    DNAs_BY_Round = DNAsOccurences_BY_Round(fastqDataFolderLocation)\n",
    "    TotalDNAs_BY_Round = TotalReads_BY_Round(fastqDataFolderLocation)\n",
    "\n",
    "    BaseRoundSortedDNAs = BaseRoundSortedDNAsList(fastqDataFolderLocation, BaseRoundIndex)\n",
    "    DNAsAppearances = DNAsAppearances_BY_Round(BaseRoundSortedDNAs, DNAs_BY_Round)\n",
    "    \n",
    "    BaseRoundSortedPeptides = BaseRoundSortedPeptidesList(fastqDataFolderLocation, BaseRoundIndex)\n",
    "    \n",
    "    BaseRoundDNAsForest = nx.Graph()\n",
    "    for DNA in BaseRoundSortedDNAs:\n",
    "        BaseRoundDNAsForest.add_node(DNA,\n",
    "                                        Occurence = DNAs_BY_Round[BaseRoundIndex][DNA],\n",
    "                                        FirstAppearance = min(DNAsAppearances[DNA]))\n",
    "    # adds nodes with attributes of Occurence in the BaseRound and FirstAppearance to the BaseRoundDNAsForest\n",
    "    UsedNodes = []\n",
    "    for DNA1 in BaseRoundSortedDNAs:\n",
    "        UsedNodes += [DNA1]\n",
    "        for DNA2 in BaseRoundSortedDNAs:\n",
    "            if DNA2 not in UsedNodes and HammingDistance(DNA1, DNA2) == 1:\n",
    "                BaseRoundDNAsForest.add_edge(DNA1,DNA2,\n",
    "                                                MutationsNumber = 1)\n",
    "    # adds edges to the BaseRoundDNAsForest so that it can be made into a directed graph\n",
    "    \n",
    "    BaseRoundDNAsTrees = list(nx.connected_component_subgraphs(BaseRoundDNAsForest, copy = True))\n",
    "    \n",
    "    DNAsClonesSummaryReportFileNameCSV =  str(today) + 'DNAsMutantsSummary' + DNAsClonesSummaryReportFileName + '.csv'\n",
    "    DNAsClonesSummaryReportFile = open(DNAsClonesSummaryReportFileNameCSV, 'w')\n",
    "    \n",
    "    Positions = {}\n",
    "    X_0_Coordinate = 1\n",
    "    Y_0_Coordinate = 0\n",
    "    Y_X0_Coordinate = 0\n",
    "    \n",
    "    for Tree in BaseRoundDNAsTrees:\n",
    "                        \n",
    "        RootDNA = max(nx.get_node_attributes(Tree, 'Occurence'),\n",
    "                        key = lambda DNA: Tree[DNA])\n",
    "        \n",
    "        TreeDNAs = {}\n",
    "        TreeDNAs[RootDNA] = [0, '', 0, Tree.node[RootDNA]['Occurence'], Tree.node[RootDNA]['FirstAppearance']]\n",
    "        TreeDNAsList = Tree.nodes()\n",
    "        TreeDNAsList.remove(RootDNA)\n",
    "        \n",
    "        for DNA in TreeDNAsList:\n",
    "            DNAPredecessor = nx.shortest_path(Tree, source = DNA, target = RootDNA, weight = None)[1]\n",
    "            PredecessorOccurence = Tree.node[DNAPredecessor]['Occurence']\n",
    "            DNAOccurence = Tree.node[DNA]['Occurence']\n",
    "            \n",
    "            TreeDNAs[DNA] = [DNAPredecessor, PredecessorOccurence, DNAOccurence]\n",
    "        \n",
    "        DNAs_BY_DistanceToTheRoot = {}\n",
    "        for DNA in Tree.nodes():\n",
    "            DistanceToTheRoot = nx.shortest_path_length(Tree, source = DNA, target = RootDNA, weight = None)\n",
    "            if DistanceToTheRoot not in DNAs_BY_DistanceToTheRoot:\n",
    "                DNAs_BY_DistanceToTheRoot[DistanceToTheRoot] = [DNA]\n",
    "            else:\n",
    "                DNAs_BY_DistanceToTheRoot[DistanceToTheRoot] += [DNA]\n",
    "                \n",
    "        MaxDNAMutantsNumber = max(map(lambda k: len(DNAs_BY_DistanceToTheRoot[k]), DNAs_BY_DistanceToTheRoot))\n",
    "        \n",
    "        SortedDNAs_BY_DistanceToTheRoot = {}\n",
    "        \n",
    "        for DistanceToTheRoot in DNAs_BY_DistanceToTheRoot:\n",
    "            \n",
    "            EquidistantDNAs = DNAs_BY_DistanceToTheRoot[DistanceToTheRoot]\n",
    "            \n",
    "            EquidistantDNAs = sorted(EquidistantDNAs, key = lambda DNA: (TreeDNAs[DNA][2]), reverse = True)\n",
    "#            EquidistantDNAs = sorted(EquidistantDNAs, key = lambda DNA: (TreeDNAs[DNA][1]), reverse = True)\n",
    "            EquidistantDNAs = sorted(EquidistantDNAs, key = lambda DNA: (TreeDNAs[DNA][0]), reverse = False)\n",
    "            \n",
    "            AdditionalElements = MaxDNAMutantsNumber - len(EquidistantDNAs)\n",
    "            SortedDNAs_BY_DistanceToTheRoot[DistanceToTheRoot] = EquidistantDNAs + AdditionalElements * ['']\n",
    "            \n",
    "            if len(Tree.nodes()) > 1:\n",
    "                for DNA in EquidistantDNAs:\n",
    "                    XCoordinate = X_0_Coordinate + DistanceToTheRoot\n",
    "                    YCoordinate = Y_0_Coordinate - EquidistantDNAs.index(DNA)\n",
    "                    Positions[DNA] = (XCoordinate, YCoordinate)\n",
    "            elif len(Tree.nodes()) == 1:\n",
    "                for DNA in EquidistantDNAs:\n",
    "                    XCoordinate = 0\n",
    "                    YCoordinate = Y_X0_Coordinate\n",
    "                    Positions[DNA] = (XCoordinate, YCoordinate)\n",
    "        \n",
    "                            \n",
    "        if len(Tree.nodes()) > 1:\n",
    "            X_0_Coordinate += max(DNAs_BY_DistanceToTheRoot.keys()) + 1\n",
    "        \n",
    "        if len(Tree.nodes()) == 1:\n",
    "            Y_X0_Coordinate -= 1\n",
    "            \n",
    "            \n",
    "#        print SortedDNAs_BY_DistanceToTheRoot\n",
    "        \n",
    "\n",
    "#-------------------------------------------------------------------------------\n",
    "\n",
    "        if len(Tree.nodes()) > 1:\n",
    "            for DistanceToTheRoot in SortedDNAs_BY_DistanceToTheRoot:\n",
    "                DNAsClonesSummaryReportFile.write(str(DistanceToTheRoot) + ' mutations' + ',' + 'frequency' + ',' + 'clones #' + ',' )\n",
    "            DNAsClonesSummaryReportFile.write('\\n')\n",
    "        \n",
    "        for i in range(MaxDNAMutantsNumber):\n",
    "            for MutationsNumber in SortedDNAs_BY_DistanceToTheRoot:                        \n",
    "                DNA = SortedDNAs_BY_DistanceToTheRoot[MutationsNumber][i]\n",
    "                \n",
    "                if DNA != '':\n",
    "                    FormatedDNA = HammingDistanceBasedFormating(RootDNA, DNA)\n",
    "                    ClonesNumber = str(len(Tree.neighbors(DNA)))\n",
    "                    DNAFraction = ('{:.2%}'.format(float((DNAs_BY_Round[BaseRoundIndex].get(DNA, 0)))/float(TotalDNAs_BY_Round[BaseRoundIndex])))\n",
    "                else:\n",
    "                    FormatedDNA = ''\n",
    "                    ClonesNumber = ''\n",
    "                    DNAFraction = ''\n",
    "                                                        \n",
    "                DNAsClonesSummaryReportFile.write(FormatedDNA + ',' +\n",
    "                            DNAFraction + ',' +\n",
    "                            ClonesNumber + ',')\n",
    "            DNAsClonesSummaryReportFile.write('\\n')\n",
    "        DNAsClonesSummaryReportFile.write('\\n')\n",
    "                    \n",
    "    DNAsClonesSummaryReportFile.close()\n",
    "        \n",
    "    BaseRoundDNAsGraph = nx.Graph()    \n",
    "    BaseRoundDNAsGraph.add_nodes_from(BaseRoundSortedDNAs)\n",
    "     \n",
    "    Sizes = []\n",
    "    for DNA in BaseRoundDNAsGraph.nodes():\n",
    "        Sizes.append(5 * math.log(DNAs_BY_Round[BaseRoundIndex][DNA], 10) + 2)\n",
    "#    Colours = []\n",
    "#    for DNA in BaseRoundDNAsGraph.nodes():\n",
    "#        Colours.append(min(DNAsAppearances[DNA]))\n",
    "    \n",
    "    Colours = []\n",
    "    for DNA in BaseRoundDNAsGraph.nodes():\n",
    "        Peptide = Translation(DNA)\n",
    "        Colours.append(BaseRoundSortedPeptides.index(Peptide))\n",
    "    \n",
    "#    print(BaseRoundDNAsGraph.nodes())\n",
    "    i = 0\n",
    "    for DNA1 in BaseRoundSortedDNAs:\n",
    "        for DNA2 in BaseRoundSortedDNAs:\n",
    "            if HammingDistance(DNA1, DNA2) == 1:\n",
    "                BaseRoundDNAsGraph.add_edge(DNA1,DNA2)\n",
    "        i += 1\n",
    "\n",
    "    XMin = min(map(lambda DNA: Positions[DNA][0], Positions)) - 1\n",
    "    XMax = max(map(lambda DNA: Positions[DNA][0], Positions)) + 1\n",
    "    YMin = min(map(lambda DNA: Positions[DNA][1], Positions)) - 5\n",
    "    YMax = max(map(lambda DNA: Positions[DNA][1], Positions)) + 5\n",
    "    \n",
    "    nx.draw_networkx(BaseRoundDNAsGraph,\n",
    "                    pos = Positions,\n",
    "                    node_size = Sizes,\n",
    "                    node_color = Colours,\n",
    "                    cmap = 'Paired',\n",
    "                    linewidths = 0.1,\n",
    "                    width = 0.0,\n",
    "                    with_labels = False,\n",
    "                    font_size = 6)\n",
    "\n",
    "    plt.axis('off')\n",
    "    plt.axis([XMin, XMax, YMin, YMax])\n",
    "    \n",
    "    DNAsMutantsSummaryReportFileNamePNG = str(today) + 'DNAsMutantsSummary' + DNAsClonesSummaryReportFileName + '.png'\n",
    "    plt.savefig(DNAsMutantsSummaryReportFileNamePNG, dpi = 500)\n",
    "    \n",
    "    plt.show()\n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def PeptidesCommonFragments(fastqDataFolderLocation, BaseRoundIndex, PeptidesClonesSummaryReportFileName):\n",
    "    \n",
    "    today = TodaysDate()\n",
    "    \n",
    "    BaseRoundSortedPeptides = BaseRoundSortedPeptidesList(fastqDataFolderLocation, BaseRoundIndex)\n",
    "    \n",
    "    MaxFragmentLength = len(max(BaseRoundSortedPeptides, key = len))\n",
    "\n",
    "    FragmentsList = {}    \n",
    "    for FragmentLength in range(3, MaxFragmentLength + 1):\n",
    "        for Peptide in BaseRoundSortedPeptides:\n",
    "            for i in range(len(Peptide) - FragmentLength):\n",
    "                Fragment = Peptide[i:i+FragmentLength]\n",
    "                if Fragment not in FragmentsList:\n",
    "                    FragmentsList[Fragment] = 1\n",
    "                else:\n",
    "                    FragmentsList[Fragment] += 1\n",
    "\n",
    "    ForestFragmentsList = {}\n",
    "    for key, value in FragmentsList.items():\n",
    "        if value > 1:\n",
    "            ForestFragmentsList[key] = value\n",
    "    \n",
    "    BaseRoundPeptidesForest = nx.Graph()\n",
    "    for Peptide in BaseRoundSortedPeptides:\n",
    "        BaseRoundPeptidesForest.add_node(Peptide)\n",
    "    # adds nodes with attributes of Occurence in the BaseRound and FirstAppearance to the BaseRoundPeptidesForest\n",
    "    UsedNodes = []\n",
    "    \n",
    "    for Peptide1 in BaseRoundSortedPeptides:\n",
    "        UsedNodes += [Peptide1]\n",
    "        for Peptide2 in BaseRoundSortedPeptides:\n",
    "            if Peptide2 not in UsedNodes and HammingDistance(Peptide1, Peptide2) == 1:\n",
    "                BaseRoundPeptidesForest.add_edge(Peptide1,Peptide2,\n",
    "                                                MutationsNumber = 1)\n",
    "    # adds edges to the BaseRoundPeptidesForest so that it can be made into a directed graph\n",
    "    \n",
    "    BaseRoundPeptidesTrees = list(nx.connected_component_subgraphs(BaseRoundPeptidesForest, copy = True))\n",
    "    \n",
    "    PeptidesClonesSummaryReportFileNameCSV =  str(today) + 'PeptidesMutantsSummary' + PeptidesClonesSummaryReportFileName + '.csv'\n",
    "    PeptidesClonesSummaryReportFile = open(PeptidesClonesSummaryReportFileNameCSV, 'w')\n",
    "    \n",
    "    BetweenTreesCommonFragments = {}\n",
    "    \n",
    "    for Tree in BaseRoundPeptidesTrees:\n",
    "        MaxFragmentLength = len(max(map(len, Tree)))\n",
    "        \n",
    "        for FragmentLength in range(3, MaxFragmentLength + 1):\n",
    "\n",
    "            FragmentsList = {}\n",
    "            for Peptide in Tree:\n",
    "                for i in range(len(Peptide) - FragmentLength):\n",
    "                    Fragment = Peptide[i:i+FragmentLength]\n",
    "                    if Fragment not in FragmentsList:\n",
    "                        FragmentsList[Fragment] = 1\n",
    "                    else:\n",
    "                        FragmentsList[Fragment] += 1\n",
    "        \n",
    "        \n",
    "        \n",
    "        \n",
    "        \n",
    "                        \n",
    "        RootPeptide = max(nx.get_node_attributes(Tree, 'Occurence'),\n",
    "                        key = lambda Peptide: Tree[Peptide])\n",
    "        \n",
    "        TreePeptides = {}\n",
    "        TreePeptides[RootPeptide] = [0, '', 0, Tree.node[RootPeptide]['Occurence'], Tree.node[RootPeptide]['FirstAppearance']]\n",
    "        TreePeptidesList = Tree.nodes()\n",
    "        TreePeptidesList.remove(RootPeptide)\n",
    "        \n",
    "        for Peptide in TreePeptidesList:\n",
    "            PeptidePredecessor = nx.shortest_path(Tree, source = Peptide, target = RootPeptide, weight = None)[1]\n",
    "            PredecessorOccurence = Tree.node[PeptidePredecessor]['Occurence']\n",
    "            PeptideOccurence = Tree.node[Peptide]['Occurence']\n",
    "            \n",
    "            TreePeptides[Peptide] = [PeptidePredecessor, PredecessorOccurence, PeptideOccurence]\n",
    "        \n",
    "        Peptides_BY_DistanceToTheRoot = {}\n",
    "        for Peptide in Tree.nodes():\n",
    "            DistanceToTheRoot = nx.shortest_path_length(Tree, source = Peptide, target = RootPeptide, weight = None)\n",
    "            if DistanceToTheRoot not in Peptides_BY_DistanceToTheRoot:\n",
    "                Peptides_BY_DistanceToTheRoot[DistanceToTheRoot] = [Peptide]\n",
    "            else:\n",
    "                Peptides_BY_DistanceToTheRoot[DistanceToTheRoot] += [Peptide]\n",
    "                \n",
    "        MaxPeptideMutantsNumber = max(map(lambda k: len(Peptides_BY_DistanceToTheRoot[k]), Peptides_BY_DistanceToTheRoot))\n",
    "        \n",
    "        SortedPeptides_BY_DistanceToTheRoot = {}\n",
    "        \n",
    "        for DistanceToTheRoot in Peptides_BY_DistanceToTheRoot:\n",
    "            \n",
    "            EquidistantPeptides = Peptides_BY_DistanceToTheRoot[DistanceToTheRoot]\n",
    "            \n",
    "            EquidistantPeptides = sorted(EquidistantPeptides, key = lambda Peptide: (TreePeptides[Peptide][2]), reverse = True)\n",
    "#            EquidistantPeptides = sorted(EquidistantPeptides, key = lambda Peptide: (TreePeptides[Peptide][1]), reverse = True)\n",
    "            EquidistantPeptides = sorted(EquidistantPeptides, key = lambda Peptide: (TreePeptides[Peptide][0]), reverse = False)\n",
    "            \n",
    "            AdditionalElements = MaxPeptideMutantsNumber - len(EquidistantPeptides)\n",
    "            SortedPeptides_BY_DistanceToTheRoot[DistanceToTheRoot] = EquidistantPeptides + AdditionalElements * ['']\n",
    "            \n",
    "            if len(Tree.nodes()) > 1:\n",
    "                for Peptide in EquidistantPeptides:\n",
    "                    XCoordinate = X_0_Coordinate + DistanceToTheRoot\n",
    "                    YCoordinate = Y_0_Coordinate - EquidistantPeptides.index(Peptide)\n",
    "                    Positions[Peptide] = (XCoordinate, YCoordinate)\n",
    "            elif len(Tree.nodes()) == 1:\n",
    "                for Peptide in EquidistantPeptides:\n",
    "                    XCoordinate = 0\n",
    "                    YCoordinate = Y_X0_Coordinate\n",
    "                    Positions[Peptide] = (XCoordinate, YCoordinate)\n",
    "        \n",
    "                            \n",
    "        if len(Tree.nodes()) > 1:\n",
    "            X_0_Coordinate += max(Peptides_BY_DistanceToTheRoot.keys()) + 1\n",
    "        \n",
    "        if len(Tree.nodes()) == 1:\n",
    "            Y_X0_Coordinate -= 1\n",
    "            \n",
    "            \n",
    "#        print SortedPeptides_BY_DistanceToTheRoot\n",
    "        \n",
    "\n",
    "#-------------------------------------------------------------------------------"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#_____________________________RUNNING THE FUNCTIONS_____________________________#\n",
    "\n",
    "#___DataFolderLocation, BaseSelectionRoundNumber, TopNPeptidesNumber, SummaryFileName___\n",
    "SelectionSummaryReport('/Users/NikitaLoik/SequenceTest', 6, 24, 'Test')\n",
    "\n",
    "#___DataFolderLocation, BaseSelectionRoundNumber, TopNPeptidesNumber, SummaryFileName___\n",
    "#DNAMutantsAnalysis('/Users/NikitaLoik/SequenceTest', 6, 24, 'Test')\n",
    "\n",
    "#___DataFolderLocation, BaseSelectionRoundNumber, TopNPeptidesNumber, SummaryFileName___\n",
    "#PeptidesMutantsSummaryReport('/Users/NikitaLoik/SequenceTest', 6, 'Test')\n",
    "\n",
    "#___DataFolderLocation, BaseSelectionRoundNumber, SummaryFileName___\n",
    "#DNAsMutantsSummaryReport('/Users/NikitaLoik/SequenceTest', 6, 'Test')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
