{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Import relevant libraries\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib\n",
    "from matplotlib import rcParams\n",
    "rcParams['font.family'] = 'monospace'\n",
    "#rcParams['font.sans-serif'] = ['Tahoma']\n",
    "\n",
    "import math\n",
    "\n",
    "import datetime\n",
    "\n",
    "import numpy as np\n",
    "\n",
    "import networkx as nx\n",
    "\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def todaysDate():\n",
    "    '''Return todays data, used in output-file name'''\n",
    "    today = datetime.date.today()\n",
    "    todaysDate = today.strftime('%d%b%Y')\n",
    "    return todaysDate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def DNACodingSequence(DNASequence, QualitySequence, StartSequence, StopSequence, cDNA_MinLength, cDNA_MaxLength, QualityScore):\n",
    "    '''Validate the DNACodingSequence based on (1) its Lenght AND (2) QualityScore of each nucleotide;\n",
    "    ONLY ONE StopSequence is used, ONLY ONE CodingSequence is returned'''\n",
    "    \n",
    "    QualityScoreString = '''!\"#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~'''\n",
    "    ThresholdQualityString = QualityScoreString[QualityScore:]\n",
    "            \n",
    "    StartIndex = DNASequence.find(StartSequence) + len(StartSequence)\n",
    "    StopIndex = DNASequence.rfind(StopSequence)\n",
    "    CodingSequence =  DNASequence[StartIndex:StopIndex]\n",
    "    if cDNA_MinLength <= len(CodingSequence) and len(CodingSequence) <= cDNA_MaxLength and len(CodingSequence)%3 == 0:\n",
    "        for Character in QualitySequence[StartIndex:StopIndex]:\n",
    "            if Character not in ThresholdQualityString:\n",
    "                return None\n",
    "        return str(CodingSequence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Define Translation of RNASequence into Peptide (peptide can be encoded as DNA or as RNA)\n",
    "\n",
    "def Translation(CodingSequence):\n",
    "\n",
    "    # Set RNA TranslationCode\n",
    "    TranslationCode = {\n",
    "                    'AAA':'K','AAC':'N','AAG':'K','AAU':'N',\n",
    "                    'ACA':'T','ACC':'T','ACG':'T','ACU':'T',\n",
    "                    'AGA':'R','AGC':'S','AGG':'R','AGU':'S',\n",
    "                    'AUA':'I','AUC':'I','AUG':'M','AUU':'I',\n",
    "                    \n",
    "                    'CAA':'Q','CAC':'H','CAG':'Q','CAU':'H',\n",
    "                    'CCA':'P','CCC':'P','CCG':'P','CCU':'P',\n",
    "                    'CGA':'R','CGC':'R','CGG':'R','CGU':'R',\n",
    "                    'CUA':'L','CUC':'L','CUG':'L','CUU':'L',\n",
    "                    \n",
    "                    'GAA':'E','GAC':'D','GAG':'E','GAU':'D',\n",
    "                    'GCA':'A','GCC':'A','GCG':'A','GCU':'A',\n",
    "                    'GGA':'G','GGC':'G','GGG':'G','GGU':'G',\n",
    "                    'GUA':'V','GUC':'V','GUG':'V','GUU':'V',\n",
    "                    \n",
    "                    'UAA':'#','UAC':'Y','UAG':'*','UAU':'Y',\n",
    "                    'UCA':'S','UCC':'S','UCG':'S','UCU':'S',\n",
    "                    'UGA':'&','UGC':'C','UGG':'W','UGU':'C',\n",
    "                    'UUA':'L','UUC':'F','UUG':'L','UUU':'F'\n",
    "                        }\n",
    "    # Stop Codons are as follows:\n",
    "    # UAA (ochre) — #\n",
    "    # UAG (amber) — *\n",
    "    # UGA (opal) — &\n",
    "                    \n",
    "    TranscriptionCode = {'A':'A','C':'C','G':'G','T':'U','U':'T'}\n",
    "    \n",
    "    # Convert DNA to RNA\n",
    "    RNASequence = ''\n",
    "    for Nucleotide in CodingSequence:\n",
    "        RNASequence += TranscriptionCode.get(Nucleotide,'X')\n",
    "        \n",
    "    #RNASequence-Test Print\n",
    "    #print (RNASequence)\n",
    "        \n",
    "    Peptide = ''\n",
    "    while len(RNASequence) != 0:\n",
    "        Peptide += TranslationCode.get(RNASequence[0:3],'Do not fuck with me!')\n",
    "        RNASequence = RNASequence[3:]\n",
    "    return Peptide"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Define SingleSelectionRoundSummary function, which returns the Occurence of CodingSequences, grouped by Peptide\n",
    "# SingleSelectionRoundSummary = {PeptideY:    {CodingSequence_YZ:    Occurence_YZ}}\n",
    "\n",
    "def SingleSelectionRoundSummary(fastqFileLocation, StartSequence, StopSequence, cDNA_MinLength, cDNA_MaxLength, QualityScore):\n",
    "    \n",
    "    RawDataFile = open(fastqFileLocation, 'r')\n",
    "    Lines = RawDataFile.readlines()\n",
    "    RawDataFile.close\n",
    "\n",
    "    # Create an empty SingleSelectionRoundSummary dictionary to store the results from a single cycle of selection\n",
    "    # SingleSelectionRoundSummary = {PeptideY:    {CodingSequence_YZ:    Occurence_YZ}}\n",
    "    SingleSelectionRoundSummary = {}\n",
    "   \n",
    "    # Populate SingleSelectionRoundSummary with the results from a single round of selection\n",
    "    # Go through the NGS-data file (.fastq) line by line\n",
    "    for i,Line in enumerate(Lines):\n",
    "        # Check that the line contains a valid sequence\n",
    "        if StartSequence in Line and StopSequence in Line:\n",
    "            CodingSequence = DNACodingSequence(Line, Lines[i + 2], StartSequence, StopSequence, cDNA_MinLength, cDNA_MaxLength, QualityScore)                      \n",
    "            if CodingSequence != None:\n",
    "                # Translated DNA or RNA sequence into PeptideSequence\n",
    "                PeptideSequence = Translation(CodingSequence)\n",
    "                # Add sequence to a SingleSelectionRoundSummary\n",
    "                if PeptideSequence not in SingleSelectionRoundSummary:\n",
    "                    SingleSelectionRoundSummary[str(PeptideSequence)] = {str(CodingSequence) : 1}\n",
    "                else:\n",
    "                    if CodingSequence not in SingleSelectionRoundSummary[str(PeptideSequence)]:\n",
    "                        SingleSelectionRoundSummary[str(PeptideSequence)][str(CodingSequence)] = 1\n",
    "                    else:\n",
    "                        SingleSelectionRoundSummary[str(PeptideSequence)][str(CodingSequence)] += 1\n",
    "\n",
    "    return SingleSelectionRoundSummary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Define HammingDistance, which returns the number of mismatches between two Sequences\n",
    "def HammingDistance(Sequence1, Sequence2):\n",
    "    \n",
    "    if len(Sequence1) < len(Sequence2):\n",
    "        Sequence1 = Sequence1 + (len(Sequence2) - len(Sequence1)) * '%'\n",
    "    elif len(Sequence1) > len(Sequence2):\n",
    "        Sequence2 = Sequence2 + (len(Sequence1) - len(Sequence2)) * '%'\n",
    "    \n",
    "    HammingDistance = 0\n",
    "    for i in range(len(Sequence1)):\n",
    "        if Sequence1[i] == Sequence2[i]:\n",
    "            HammingDistance = HammingDistance\n",
    "        else:\n",
    "            HammingDistance = HammingDistance + 1\n",
    "            \n",
    "    return HammingDistance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Define HammingDistanceBasedFormating, which returns FormatedSequence, such that mismatches are capitalised, and the rest of the FormatedSequence is lowercase \n",
    "def HammingDistanceBasedFormating(Sequence1, Sequence2):\n",
    "    \n",
    "    if len(Sequence1) < len(Sequence2):\n",
    "        Sequence1 = Sequence1 + (len(Sequence2) - len(Sequence1)) * '-'\n",
    "    elif len(Sequence1) > len(Sequence2):\n",
    "        Sequence2 = Sequence2 + (len(Sequence1) - len(Sequence2)) * '-'\n",
    "    \n",
    "    HammingDistance = 0\n",
    "    FormatedSequence2 = ''\n",
    "    for i in range(len(Sequence1)):\n",
    "        if Sequence1[i] == Sequence2[i]:\n",
    "            FormatedSequence2 += Sequence2[i].lower()\n",
    "            HammingDistance = HammingDistance\n",
    "        else:\n",
    "            FormatedSequence2 += Sequence2[i]\n",
    "            HammingDistance = HammingDistance + 1            \n",
    "    return FormatedSequence2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Define CompleteSelectionSummary, which returns the Occurence of CodingSequences, grouped by Peptide, grouped by SelectionRound\n",
    "# SelectionSummary = {SelectionRound_X:    {PeptideXY:    {CodingDNA_XYZ:    Occurence_XYZ}}}\n",
    "def CompleteSelectionSummary(DataFolderLocation,\n",
    "                             StartSequence,\n",
    "                             StopSequence,\n",
    "                             cDNA_MinLength,\n",
    "                             cDNA_MaxLength,\n",
    "                             QualityScore):\n",
    "    # Create an empty SelectionSummary dictionary to store the results from all the rounds of selection\n",
    "    CompleteSelectionSummary = {}\n",
    "    \n",
    "    for file in os.listdir(DataFolderLocation):\n",
    "        \n",
    "        FileLocation = os.path.join(DataFolderLocation, file)\n",
    "        \n",
    "        #(1.A) Extract the round number from the file name (file name should have two digit number before full stop — '00.')    \n",
    "        if file.endswith('.fastq'): # this conditional is necessary; without it some shit appears in the beginning of the file list\n",
    "            RoundNumberFirstDigit = file[file.find('.')-2]\n",
    "            RoundNumberSecondDigit = file[file.find('.')-1]\n",
    "            if RoundNumberFirstDigit == '0':\n",
    "                RoundNumber = int(RoundNumberSecondDigit)\n",
    "                # RoundNumber_Test Print\n",
    "                #print RoundNumber\n",
    "            elif RoundNumberFirstDigit != '0':\n",
    "                RoundNumber = int(file[file.find('.')-2 : file.find('.')])\n",
    "                # RoundNumber_Test Print\n",
    "                #print RoundNumber\n",
    "            \n",
    "            #(1.B) Extract single round results\n",
    "            SelectionRoundSummary = SingleSelectionRoundSummary(FileLocation, StartSequence, StopSequence, cDNA_MinLength, cDNA_MaxLength, QualityScore)\n",
    "             \n",
    "            #(1.C) Populate ConcatenatedResultsList\n",
    "            CompleteSelectionSummary[RoundNumber] = SelectionRoundSummary\n",
    "            \n",
    "            # ConcatenatedResultsList-Test Print\n",
    "            #print ConcatenatedResultsList\n",
    "            \n",
    "    return CompleteSelectionSummary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Define PeptidesOccurences_BY_Round, which returns the Occurences of Peptides groupped by round\n",
    "# PeptidesOccurences_BY_Round = {Round_X:    {PeptideXY:    Occurence_XY}}\n",
    "def PeptidesOccurences_BY_Round(DataFolderLocation, cDNA_MinLength, cDNA_MaxLength, QualityScore):\n",
    "    SelectionSummary = CompleteSelectionSummary(DataFolderLocation, StartSequence, StopSequence, cDNA_MinLength, cDNA_MaxLength, QualityScore)\n",
    "    \n",
    "    PeptidesOccurences_BY_Round = {}\n",
    "    for Round in SelectionSummary:\n",
    "        PeptidesOccurences_IN_Round = {}\n",
    "        for Peptide in SelectionSummary[Round]:\n",
    "            PeptidesOccurences_IN_Round[Peptide] = sum(SelectionSummary[Round][Peptide].values())\n",
    "        PeptidesOccurences_BY_Round[Round] = PeptidesOccurences_IN_Round\n",
    "        \n",
    "    return PeptidesOccurences_BY_Round"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Define DNAsOccurences_BY_Round, which returns the Occurences of DNAs groupped by round\n",
    "# DNAsOccurences_BY_Round = {Round_X:    {DNA_XY:    Occurence_XY}}\n",
    "def DNAsOccurences_BY_Round(DataFolderLocation, cDNA_MinLength, cDNA_MaxLength, QualityScore):\n",
    "    SelectionSummary = CompleteSelectionSummary(DataFolderLocation, StartSequence, StopSequence, cDNA_MinLength, cDNA_MaxLength, QualityScore)\n",
    "    \n",
    "    DNAsOccurences_BY_Round = {}\n",
    "    for Round in SelectionSummary:\n",
    "        DNAsOccurences_IN_Round = {}\n",
    "        for Peptide in SelectionSummary[Round]:\n",
    "            for DNA in SelectionSummary[Round][Peptide]:\n",
    "                DNAsOccurences_IN_Round[DNA] = SelectionSummary[Round][Peptide][DNA]\n",
    "        DNAsOccurences_BY_Round[Round] = DNAsOccurences_IN_Round\n",
    "\n",
    "    return DNAsOccurences_BY_Round"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Define TotalReads_BY_Round, which returns number of reads by round\n",
    "# TotalReads_BY_Round = {Round_X:    TotalReads_X}\n",
    "def TotalReads_BY_Round(DataFolderLocation, cDNA_MinLength, cDNA_MaxLength, QualityScore):\n",
    "    SelectionSummary = CompleteSelectionSummary(DataFolderLocation, StartSequence, StopSequence, cDNA_MinLength, cDNA_MaxLength, QualityScore)\n",
    "    Peptides_BY_Round = PeptidesOccurences_BY_Round(DataFolderLocation, cDNA_MinLength, cDNA_MaxLength, QualityScore)\n",
    "    \n",
    "    TotalReads_BY_Round = {}\n",
    "    for Round in SelectionSummary:\n",
    "        TotalReads_BY_Round[Round] = sum(Peptides_BY_Round[Round].values())\n",
    "        \n",
    "    return TotalReads_BY_Round"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Define BaseRoundSortedPeptidesList, which returns list of peptides in base round sorted by their occurence\n",
    "# BaseRoundSortedPeptidesList = [Peptide_1, ..., Peptide_N]; Occurence(Peptide_1) > ... > Occurence(Peptide_N)\n",
    "def BaseRoundSortedPeptidesList(DataFolderLocation, BaseSelectionRoundNumber):\n",
    "    Peptides_BY_Round = PeptidesOccurences_BY_Round(DataFolderLocation, cDNA_MinLength, cDNA_MaxLength, QualityScore)  \n",
    "            \n",
    "    PeptidesOccurencesInBaseRound = Peptides_BY_Round[BaseSelectionRoundNumber]\n",
    "    BaseRoundSortedPeptidesList = sorted(PeptidesOccurencesInBaseRound, key = PeptidesOccurencesInBaseRound.get, reverse = True)\n",
    "    \n",
    "    return BaseRoundSortedPeptidesList"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Define BaseRoundSortedDNAsList, which returns list of DNAs in base round sorted by their occurence\n",
    "# BaseRoundSortedDNAsList = [DNA_1, ..., DNA_N]; Occurence(DNA_1) > ... > Occurence(DNA_N)\n",
    "def BaseRoundSortedDNAsList(DataFolderLocation, BaseSelectionRoundNumber):\n",
    "    DNAs_BY_Round = DNAsOccurences_BY_Round(DataFolderLocation, cDNA_MinLength, cDNA_MaxLength, QualityScore)  \n",
    "            \n",
    "    DNAsOccurences_IN_BaseRound = DNAs_BY_Round[BaseSelectionRoundNumber]\n",
    "    BaseRoundSortedDNAsList = sorted(DNAsOccurences_IN_BaseRound, key = DNAsOccurences_IN_BaseRound.get, reverse = True)\n",
    "    \n",
    "    return BaseRoundSortedDNAsList"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def PeptidesRank_IN_BaseRound(DataFolderLocation, BaseSelectionRoundNumber):\n",
    "    Peptides_BY_Round = PeptidesOccurences_BY_Round(DataFolderLocation, cDNA_MinLength, cDNA_MaxLength, QualityScore)\n",
    "    BaseRoundSortedPeptides = BaseRoundSortedPeptidesList(DataFolderLocation, BaseSelectionRoundNumber)\n",
    "    \n",
    "    BasePeptideCount = 0\n",
    "    PeptideRank = 1\n",
    "    \n",
    "    PeptidesRank_IN_BaseRound = {}\n",
    "    \n",
    "    for Peptide in BaseRoundSortedPeptides:\n",
    "        PeptideCount = Peptides_BY_Round[BaseSelectionRoundNumber][Peptide]\n",
    "        if PeptideCount < BasePeptideCount:\n",
    "            PeptideRank += 1\n",
    "        \n",
    "        PeptidesRank_IN_BaseRound[Peptide] = PeptideRank\n",
    "        BasePeptideCount = PeptideCount\n",
    "        \n",
    "    return PeptidesRank_IN_BaseRound"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Occurences is a bad word in here\n",
    "\n",
    "\n",
    "# Define DNAClonesOccurences_BY_Round_BY_Peptide, which returns number of clones for each peptide groupped by round.\n",
    "# DNAClonesOccurences_BY_Round_BY_Peptide = {SelectionRound_X:    {PeptideXY:    DNAClonesOccurences}}\n",
    "def DNAClonesOccurences_BY_Round_BY_Peptide(DataFolderLocation, cDNA_MinLength, cDNA_MaxLength, QualityScore):\n",
    "    SelectionSummary = CompleteSelectionSummary(DataFolderLocation, StartSequence, StopSequence, cDNA_MinLength, cDNA_MaxLength, QualityScore)\n",
    "    \n",
    "    DNAClonesOccurences_BY_Round_BY_Peptide = {}\n",
    "    for Round in SelectionSummary:\n",
    "        DNAClonesOccurences_BY_Peptide = {}\n",
    "        for Peptide in SelectionSummary[Round]:\n",
    "            DNAClonesOccurences_BY_Peptide[Peptide] = len(SelectionSummary[Round][Peptide])\n",
    "        DNAClonesOccurences_BY_Round_BY_Peptide[Round] = DNAClonesOccurences_BY_Peptide\n",
    "        \n",
    "    return DNAClonesOccurences_BY_Round_BY_Peptide"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Define PeptidesAppearances_BY_Round, which returns for each peptide in selection a list of rounds in which this peptide appears.\n",
    "# PeptidesAppearances_BY_Round = {Peptide_X:    [Round_A, ..., Round_N]}\n",
    "def PeptidesAppearances_BY_Round(BaseRoundSortedPeptidesList, PeptidesOccurences_BY_Round):\n",
    "    \n",
    "    PeptidesAppearances_BY_Round = {}\n",
    "    \n",
    "    for Peptide in BaseRoundSortedPeptidesList:\n",
    "        PeptidesAppearances_BY_Round[Peptide] = []\n",
    "        for Round in PeptidesOccurences_BY_Round:\n",
    "            if Peptide in PeptidesOccurences_BY_Round[Round]:\n",
    "                PeptidesAppearances_BY_Round[Peptide] += [Round]\n",
    "    return PeptidesAppearances_BY_Round"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Define DNAsAppearances_BY_Round, which returns for each DNA in selection a list of rounds in which this DNA appears.\n",
    "# DNAsAppearances_BY_Round = {DNA_X:    [Round_A, ..., Round_N]}\n",
    "def DNAsAppearances_BY_Round(BaseRoundSortedDNAsList, DNAsOccurences_BY_Round):\n",
    "    \n",
    "    DNAsAppearances_BY_Round = {}\n",
    "    \n",
    "    for DNA in BaseRoundSortedDNAsList:\n",
    "        DNAsAppearances_BY_Round[DNA] = []\n",
    "        for Round in DNAsOccurences_BY_Round:\n",
    "            if DNA in DNAsOccurences_BY_Round[Round]:\n",
    "                DNAsAppearances_BY_Round[DNA] += [Round]\n",
    "    return DNAsAppearances_BY_Round"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Define SelectionSummaryReport, which, for the TopNPeptidesNumber, returns a summary table (.txt) and provides a summary grpaph (.png).\n",
    "def SelectionSummaryReport(DataFolderLocation,\n",
    "                           BaseSelectionRoundNumber,\n",
    "                           TopNPeptidesNumber,\n",
    "                           StartSequence,\n",
    "                           StopSequence,\n",
    "                           cDNA_MinLength,\n",
    "                           cDNA_MaxLength,\n",
    "                           QualityScore,\n",
    "                           FileName):\n",
    "    \n",
    "    today = todaysDate() \n",
    "    \n",
    "    SelectionSummaryFileNameCSV = str(today) + 'SelectionSummary' + FileName + '.csv'\n",
    "    SelectionSummaryReportFile = open(SelectionSummaryFileNameCSV, 'w')\n",
    "    \n",
    "    SelectionSummary = CompleteSelectionSummary(DataFolderLocation, StartSequence, StopSequence, cDNA_MinLength, cDNA_MaxLength, QualityScore)\n",
    "    SortedRoundsList = sorted(SelectionSummary.keys())\n",
    "    \n",
    "    Peptides_BY_Round = PeptidesOccurences_BY_Round(DataFolderLocation, cDNA_MinLength, cDNA_MaxLength, QualityScore)\n",
    "    TotalPeptides_BY_Round = TotalReads_BY_Round(DataFolderLocation, cDNA_MinLength, cDNA_MaxLength, QualityScore)\n",
    "    \n",
    "    BaseRoundSortedPeptides = BaseRoundSortedPeptidesList(DataFolderLocation, BaseSelectionRoundNumber)\n",
    "    BaseRoundTopSortedPeptides = BaseRoundSortedPeptides[0 : (TopNPeptidesNumber)]\n",
    "\n",
    "    BaseRoundPeptidesRank = PeptidesRank_IN_BaseRound(DataFolderLocation, BaseSelectionRoundNumber)\n",
    "    #print (BaseRoundSortedPeptides)\n",
    "    \n",
    "        \n",
    "    SelectionSummaryReportFile.write('peptide sequence' + ',' +\n",
    "                                     'rank (#)' + ',' +\n",
    "                                     'cDNA mutants' + ',')\n",
    "    for Round in SortedRoundsList:\n",
    "        SelectionSummaryReportFile.write('C' +\n",
    "                                         str(Round) +\n",
    "                                         ' count (#) [frequency(%)]' + ',')\n",
    "    SelectionSummaryReportFile.write('\\n')\n",
    "    \n",
    "    for Peptide in BaseRoundTopSortedPeptides:\n",
    "        BaseRoundPeptideFraction = float((Peptides_BY_Round[Round].get(Peptide, 0)))/float(TotalPeptides_BY_Round[BaseSelectionRoundNumber])\n",
    "        PeptideRank = BaseRoundPeptidesRank[Peptide]\n",
    "        FormatedPeptide = HammingDistanceBasedFormating(BaseRoundTopSortedPeptides[0], Peptide)\n",
    "        PeptidecDNAMutants = len(SelectionSummary[BaseSelectionRoundNumber][Peptide])\n",
    "        SelectionSummaryReportFile.write(FormatedPeptide + ',' +\n",
    "                                         str(PeptideRank) + ',' +\n",
    "                                         str(PeptidecDNAMutants) + ',')\n",
    "            \n",
    "        for Round in SortedRoundsList:\n",
    "            PeptideFraction = float((Peptides_BY_Round[Round].get(Peptide, 0)))/float(TotalPeptides_BY_Round[Round])\n",
    "\n",
    "            BaseFraction = PeptideFraction\n",
    "            \n",
    "            SelectionSummaryReportFile.write(str(Peptides_BY_Round[Round].get(Peptide, 0)) +\n",
    "                                             ' [' + '{:.1%}'.format(PeptideFraction) + ']' + ',')\n",
    "            \n",
    "            BaseFraction = PeptideFraction\n",
    "        SelectionSummaryReportFile.write('\\n')\n",
    "        \n",
    "    SelectionSummaryReportFile.write('total count (#)' + ',' + ',')\n",
    "    for Round in SortedRoundsList:\n",
    "        SelectionSummaryReportFile.write(str(TotalPeptides_BY_Round[Round]) + ',')\n",
    "    SelectionSummaryReportFile.write('\\n\\n\\n')\n",
    "            \n",
    "    SelectionSummaryReportFile.close()\n",
    "    \n",
    "#-------------------------------------------------------------------------------\n",
    "   \n",
    "    # Create a figure of size 8x6 inches, 500 dots per inch\n",
    "    plt.figure(figsize = (8, 6),\n",
    "               dpi = 500)\n",
    "    # Create 'ggplot' style\n",
    "    plt.style.use('fivethirtyeight')\n",
    "    # Create a new subplot from a grid of 1x1\n",
    "    Graph = plt.subplot(1, 1, 1)\n",
    "    \n",
    "    Xs = []\n",
    "    Ys = []\n",
    "\n",
    "    # Map colours onto lines  \n",
    "    cNorm  = matplotlib.colors.Normalize(vmin = 0,\n",
    "                                         vmax = TopNPeptidesNumber - 1)\n",
    "    scalarMap = matplotlib.cm.ScalarMappable(norm = cNorm,\n",
    "                                             cmap = 'gist_rainbow')\n",
    "    \n",
    "    PeptideLabels = []\n",
    "    \n",
    "    for Peptide in BaseRoundTopSortedPeptides:\n",
    "    #for Peptide in Top24PeptidesKDs:\n",
    "        PeptidesFractions_BY_Round = []\n",
    "        for Round in SortedRoundsList:\n",
    "            PeptidesFractions_BY_Round += [float((Peptides_BY_Round[Round].get(Peptide, 0)))/float(TotalPeptides_BY_Round[Round])]\n",
    "        \n",
    "        x = SortedRoundsList\n",
    "        y = PeptidesFractions_BY_Round\n",
    "        Xs += x\n",
    "        Ys += y\n",
    "        \n",
    "        PeptideRank = BaseRoundPeptidesRank[Peptide]\n",
    "        #PeptideColour = scalarMap.to_rgba(PeptideRank)\n",
    "        PeptideColour = scalarMap.to_rgba(BaseRoundTopSortedPeptides.index(Peptide))\n",
    "        FormatedPeptide = HammingDistanceBasedFormating(BaseRoundTopSortedPeptides[0], Peptide)\n",
    "        \n",
    "        PeptideLabel =  FormatedPeptide + ' (' + str(PeptideRank) +')'\n",
    "        \n",
    "        #Set PeptideLabel\n",
    "        PeptideLabels += [PeptideLabel]\n",
    "        \n",
    "        plt.plot(x, y,\n",
    "                 'o-',\n",
    "                 c = PeptideColour,\n",
    "                 lw = 2.0,\n",
    "                 ms = 4.0,\n",
    "                 mew = 0.1,\n",
    "                 mec = '#191919')\n",
    "\n",
    "    XMin = min(Xs) - 0.05*(max(Xs) - min(Xs))\n",
    "    XMax = max(Xs) + 0.05*(max(Xs) - min(Xs))\n",
    "    YMin = min(Ys) - 0.05*(max(Ys) - min(Ys))\n",
    "    YMax = max(Ys) + 0.05*(max(Ys) - min(Ys))\n",
    "    \n",
    "    plt.axis([XMin, XMax, YMin, YMax])\n",
    "    \n",
    "    plt.xticks(fontsize = 10)\n",
    "    plt.yticks(fontsize = 10)\n",
    "    \n",
    "    plt.xlabel(\"Selection Cycle (#)\",\n",
    "               fontsize = 10)\n",
    "    plt.ylabel(\"Peptide Fraction\",\n",
    "               fontsize = 10)\n",
    "    \n",
    "    legend = plt.legend(PeptideLabels,\n",
    "                        title = 'cyclic-peptide random region',\n",
    "                        loc = 'upper center',\n",
    "                        bbox_to_anchor = (0.5, -0.10),\n",
    "                        fancybox = True,\n",
    "                        shadow = False,\n",
    "                        fontsize = 10,\n",
    "                        ncol = 3)\n",
    "    \n",
    "    Graph.get_legend().get_title().set_size('small')\n",
    "    \n",
    "    SelectionSummaryFileNamePNG = str(today) + 'SelectionSummary' + FileName + '.png'\n",
    "    \n",
    "    plt.savefig(SelectionSummaryFileNamePNG,\n",
    "                bbox_extra_artists = [legend],\n",
    "                bbox_inches = 'tight',\n",
    "                dpi = 300)\n",
    "    plt.show()\n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def DNAMutantsAnalysis(DataFolderLocation,\n",
    "                       BaseSelectionRoundNumber,\n",
    "                       TopNPeptidesNumber,\n",
    "                       StartSequence,\n",
    "                       StopSequence,\n",
    "                       FileName):\n",
    "    \n",
    "    today = todaysDate() \n",
    "    \n",
    "    DNAMutantsAnalysisFileNameCSV =  str(today) + 'DNAsMutantsAnalysis' + FileName + '.csv'\n",
    "    DNAMutantsAnalysisFile = open(DNAMutantsAnalysisFileNameCSV, 'w')\n",
    "    \n",
    "    SelectionSummary = CompleteSelectionSummary(DataFolderLocation, StartSequence, StopSequence, cDNA_MinLength, cDNA_MaxLength, QualityScore)\n",
    "    SortedRoundsList = sorted(SelectionSummary.keys())\n",
    "    \n",
    "    Peptides_BY_Round = PeptidesOccurences_BY_Round(DataFolderLocation, cDNA_MinLength, cDNA_MaxLength, QualityScore)\n",
    "    TotalPeptides_BY_Round = TotalReads_BY_Round(DataFolderLocation, cDNA_MinLength, cDNA_MaxLength, QualityScore)\n",
    "    \n",
    "    BaseRoundSortedPeptides = BaseRoundSortedPeptidesList(DataFolderLocation, BaseSelectionRoundNumber)\n",
    "    BaseRoundTopSortedPeptides = BaseRoundSortedPeptides[0 : (TopNPeptidesNumber)]\n",
    "    \n",
    "    DNAClones_BY_Round_BY_Peptide = DNAClonesOccurences_BY_Round_BY_Peptide(DataFolderLocation, cDNA_MinLength, cDNA_MaxLength, QualityScore)\n",
    "    \n",
    "    DNAMutantsAnalysisFile.write('peptide sequence' + ',')\n",
    "    for Round in SortedRoundsList:\n",
    "        DNAMutantsAnalysisFile.write('round # ' + str(Round) + ' DNA clones (#)' + ',')\n",
    "    DNAMutantsAnalysisFile.write('\\n')\n",
    "    \n",
    "    for Peptide in BaseRoundTopSortedPeptides:\n",
    "        DNAMutantsAnalysisFile.write(Peptide + ',')\n",
    "        for Round in SortedRoundsList:\n",
    "            DNAMutantsAnalysisFile.write(str(DNAClones_BY_Round_BY_Peptide[Round].get(Peptide, 0)) + ',')\n",
    "        DNAMutantsAnalysisFile.write('\\n')\n",
    "    DNAMutantsAnalysisFile.close()\n",
    "    \n",
    "#-------------------------------------------------------------------------------        \n",
    "\n",
    "    # Create a figure of size 8x6 inches, 500 dots per inch\n",
    "    plt.figure(figsize = (8, 6),\n",
    "               dpi = 500)\n",
    "    # Create 'ggplot' style\n",
    "    plt.style.use('fivethirtyeight')\n",
    "    # Create a new subplot from a grid of 1x1\n",
    "    Graph = plt.subplot(1, 1, 1)\n",
    "    \n",
    "#    PeptideDNAClonesNumber_IN_BaseRound = []\n",
    "#    PeptideOccurence_IN_BaseRound = []\n",
    "\n",
    "    # Map colours onto lines\n",
    "    cNorm  = matplotlib.colors.Normalize(vmin = 0,\n",
    "                                         vmax = len(BaseRoundSortedPeptides) - 1)\n",
    "    scalarMap = matplotlib.cm.ScalarMappable(norm = cNorm,\n",
    "                                             cmap = 'gist_rainbow')\n",
    "\n",
    "    RoundIndex = BaseSelectionRoundNumber\n",
    "\n",
    "    Xs = []\n",
    "    Ys = []        \n",
    "    for Peptide in DNAClones_BY_Round_BY_Peptide[RoundIndex]:\n",
    "        PeptideDNAClonesNumber_IN_BaseRound = math.log(DNAClones_BY_Round_BY_Peptide[RoundIndex].get(Peptide, 0), 2)\n",
    "        PeptideOccurence_IN_BaseRound = math.log(Peptides_BY_Round[RoundIndex].get(Peptide, 0), 2)\n",
    "        \n",
    "        PeptideColour = scalarMap.to_rgba(BaseRoundSortedPeptides.index(Peptide))\n",
    "    \n",
    "        x = PeptideDNAClonesNumber_IN_BaseRound\n",
    "        y = PeptideOccurence_IN_BaseRound\n",
    "        Xs += [x]\n",
    "        Ys += [y]\n",
    "        \n",
    "        plt.plot(x, y,\n",
    "                'o',\n",
    "                c = PeptideColour,\n",
    "                ms = 5.0,\n",
    "                mew = 0.1,\n",
    "                mec = '#191919')\n",
    "\n",
    "    XMin = min(Xs) - 0.05*(max(Xs) - min(Xs))\n",
    "    XMax = max(Xs) + 0.05*(max(Xs) - min(Xs))\n",
    "    YMin = min(Ys) - 0.05*(max(Ys) - min(Ys))\n",
    "    YMax = max(Ys) + 0.05*(max(Ys) - min(Ys))\n",
    "    \n",
    "    plt.axis([XMin, XMax, YMin, YMax])\n",
    "        \n",
    "    XLabel = 'log$2$ (DNA Clones #)' #$_$ makes subscript possible\n",
    "    plt.xlabel(XLabel, fontsize = 14)\n",
    "    YLabel = 'log$2$ (Peptide Occurence)'\n",
    "    plt.ylabel(YLabel, fontsize = 14)\n",
    "    \n",
    "    legend = plt.legend(BaseRoundSortedPeptides,\n",
    "                        loc = 'upper center',\n",
    "                        bbox_to_anchor = (0.5, -0.15),\n",
    "                        fancybox = True,\n",
    "                        shadow = False,\n",
    "                        ncol = 4)\n",
    "    \n",
    "    DNAClonesAnalysisFileNamePNG = str(today) + 'DNAsMutantsAnalysisRegression' + 'R' + str(Round) + FileName + '.png'\n",
    "    \n",
    "    plt.savefig(DNAClonesAnalysisFileNamePNG, bbox_extra_artists=[legend], bbox_inches='tight', dpi = 300)\n",
    "    plt.show()\n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def PeptidesRelatednessAnalysis(DataFolderLocation,\n",
    "                                BaseSelectionRoundNumber,\n",
    "                                TopNPeptidesNumber,\n",
    "                                StartSequence,\n",
    "                                StopSequence,\n",
    "                                cDNA_MinLength,\n",
    "                                cDNA_MaxLength,\n",
    "                                QualityScore,\n",
    "                                FileName):\n",
    "    \n",
    "    # to extract todaysDate\n",
    "    today = todaysDate()\n",
    "    \n",
    "    # to collect DNAs-based summary information By_Round\n",
    "    DNAs_BY_Round = DNAsOccurences_BY_Round(DataFolderLocation, cDNA_MinLength, cDNA_MaxLength, QualityScore)\n",
    "    TotalDNAs_BY_Round = TotalReads_BY_Round(DataFolderLocation, cDNA_MinLength, cDNA_MaxLength, QualityScore)\n",
    "    BaseRoundSortedDNAs = BaseRoundSortedDNAsList(DataFolderLocation, BaseSelectionRoundNumber)\n",
    "    DNAsAppearances = DNAsAppearances_BY_Round(BaseRoundSortedDNAs, DNAs_BY_Round)\n",
    "    \n",
    "    # to collect Peptides-based summary information By_Round\n",
    "    Peptides_BY_Round = PeptidesOccurences_BY_Round(DataFolderLocation, cDNA_MinLength, cDNA_MaxLength, QualityScore)\n",
    "    TotalPeptides_BY_Round = TotalReads_BY_Round(DataFolderLocation, cDNA_MinLength, cDNA_MaxLength, QualityScore)\n",
    "    BaseRoundSortedPeptides = BaseRoundSortedPeptidesList(DataFolderLocation, BaseSelectionRoundNumber)\n",
    "    PeptidesAppearances = PeptidesAppearances_BY_Round(BaseRoundSortedPeptides, Peptides_BY_Round)\n",
    "    \n",
    "    SelectionSummary = CompleteSelectionSummary(DataFolderLocation, StartSequence, StopSequence, cDNA_MinLength, cDNA_MaxLength, QualityScore)\n",
    "    SortedRoundsList = sorted(SelectionSummary.keys())\n",
    "    \n",
    "    # to create a disjoint graph (Forest), based on DNAs in the BaseRound (joint subgraphs are the Trees and the unique DNA sequences are the Leaves)\n",
    "    BaseRoundDNAsForest = nx.Graph()\n",
    "    # to add nodes (Leaves, unique DNA sequences) to the BaseRoundDNAsForest disjoint graph\n",
    "    BaseRoundDNAsForest.add_nodes_from(BaseRoundSortedDNAs)\n",
    "    # to add edges (Twigs, DNA-to-DNA connections based on the hamming distance between unique DNA sequences) to the BaseRoundDNAsForest so that disjoint graphs (stand alone Trees) can be identified\n",
    "    UsedNodes = []\n",
    "    for DNA1 in BaseRoundSortedDNAs:\n",
    "        UsedNodes += [DNA1]\n",
    "        for DNA2 in BaseRoundSortedDNAs:\n",
    "            if DNA2 not in UsedNodes and HammingDistance(DNA1, DNA2) == 1:\n",
    "                BaseRoundDNAsForest.add_edge(DNA1,DNA2,\n",
    "                                                MutationsNumber = 1)\n",
    "    # to extract individual joint subgraphs (stand alone Trees) from the disjoint graph (Forest)\n",
    "    BaseRoundDNAsTrees = list(nx.connected_component_subgraphs(BaseRoundDNAsForest, copy = True))\n",
    "    \n",
    "    # to create a PeptideSummaryDNAPerspectiveCSV file\n",
    "    PeptidesSummaryFileNameCSV =  str(today) + 'PeptideFamiliesSummary' + FileName + '.csv'\n",
    "    PeptidesSummaryFile = open(PeptidesSummaryFileNameCSV, 'w')\n",
    "    \n",
    "    # to convert list of DNAs Trees into a list of Peptides Trees Leaves\n",
    "    PeptidesTreesLeaves = []\n",
    "    for DNAsTree in BaseRoundDNAsTrees:\n",
    "        PeptideLeaves = []\n",
    "        for DNA in DNAsTree:\n",
    "            Peptide = Translation(DNA)\n",
    "            if Peptide not in PeptideLeaves:\n",
    "                PeptideLeaves += [Peptide]\n",
    "        PeptidesTreesLeaves += [PeptideLeaves]\n",
    "    # to sort the resulting list of lists from the largest to smallest\n",
    "    PeptidesTreesLeaves.sort(key = len, reverse = True)\n",
    "    \n",
    "    # to fix the coordinates of the origin of the graph\n",
    "    Positions = {}\n",
    "    X_0_Coordinate = 1\n",
    "    Y_0_Coordinate = 0\n",
    "    Y_X0_Coordinate = 0\n",
    "    \n",
    "    TreesXCoordinates = []\n",
    "    \n",
    "    PeptideGraphFigure = plt.figure()\n",
    "    PeptideGraph = PeptideGraphFigure.add_subplot(1, 1, 1)\n",
    "    \n",
    "    #to introduce PeptideFamilyCounter\n",
    "    MultiplePeptideFamilyCounter = 0\n",
    "    #to introduce SinglePeptideFamilyCounter\n",
    "    SinglePeptideFamilyCounter = 0\n",
    "    #to introduce PeptideFamilySize\n",
    "    PeptideTreeSize = []\n",
    "    \n",
    "    \n",
    "    # to create a tree for each set of Peptides Trees Leaves \n",
    "    for PeptideLeaves in PeptidesTreesLeaves:\n",
    "        \n",
    "        PeptideTree = nx.Graph()\n",
    "        # to convert each Peptide (Leave) into a node of a Peptide graph (PeptideLeave on a PeptideTree)\n",
    "        \n",
    "        for Peptide in PeptideLeaves:\n",
    "            PeptideTree.add_node(Peptide,\n",
    "                                        Occurence = Peptides_BY_Round[BaseSelectionRoundNumber][Peptide],\n",
    "                                        FirstAppearance = min(PeptidesAppearances[Peptide]))\n",
    "        # to join the Peptide nodes of a graph by edges (Twigs)\n",
    "        for Peptide1 in PeptideLeaves:\n",
    "            for Peptide2 in PeptideLeaves:\n",
    "                if HammingDistance(Peptide1, Peptide2) == 1:\n",
    "                    PeptideTree.add_edge(Peptide1,Peptide2)\n",
    "        \n",
    "        # to identify the RootPeptide of a PeptideTree graph\n",
    "        TreePeptidesOccurences = nx.get_node_attributes(PeptideTree, 'Occurence')              \n",
    "        RootPeptide = max(TreePeptidesOccurences, key=TreePeptidesOccurences.get)\n",
    "        \n",
    "        # to create a dictionary holder for Peptide and their properties (Predecessor and Occurrence)\n",
    "        TreePeptides = {}\n",
    "        TreePeptides[RootPeptide] = [0, '', 0, PeptideTree.node[RootPeptide]['Occurence'], PeptideTree.node[RootPeptide]['FirstAppearance']]\n",
    "        TreePeptidesList = list(PeptideTree.nodes())\n",
    "        TreePeptidesList.remove(RootPeptide)\n",
    "\n",
    "        for Peptide in TreePeptidesList:\n",
    "            PeptidePredecessor = nx.shortest_path(PeptideTree, source = Peptide, target = RootPeptide, weight = None)[1]\n",
    "            # PredecessorOccurence can be used to sort the Peptides, but does not seem to be useful\n",
    "            PredecessorOccurence = PeptideTree.node[PeptidePredecessor]['Occurence']\n",
    "            PeptideOccurence = PeptideTree.node[Peptide]['Occurence']\n",
    "\n",
    "            TreePeptides[Peptide] = [PeptidePredecessor, PredecessorOccurence, PeptideOccurence]\n",
    "            \n",
    "        \n",
    "        # to sort Peptides in a PeptideTree by their distance to the RootPeptide\n",
    "        Peptides_BY_DistanceToTheRoot = {}\n",
    "        for Peptide in PeptideTree.nodes():\n",
    "            DistanceToTheRoot = nx.shortest_path_length(PeptideTree, source = Peptide, target = RootPeptide, weight = None)\n",
    "            if DistanceToTheRoot not in Peptides_BY_DistanceToTheRoot:\n",
    "                Peptides_BY_DistanceToTheRoot[DistanceToTheRoot] = [Peptide]\n",
    "            else:\n",
    "                Peptides_BY_DistanceToTheRoot[DistanceToTheRoot] += [Peptide]\n",
    "        \n",
    "        # to identify the largest group of equidistanced peptides \n",
    "        MaxPeptidesNumber = max(map(lambda k: len(Peptides_BY_DistanceToTheRoot[k]), Peptides_BY_DistanceToTheRoot))\n",
    "\n",
    "        SortedPeptides_BY_DistanceToTheRoot = {}\n",
    "        # to sort Peptides by their distance to the RootPeptide\n",
    "        for DistanceToTheRoot in Peptides_BY_DistanceToTheRoot:\n",
    "\n",
    "            EquidistantPeptides = Peptides_BY_DistanceToTheRoot[DistanceToTheRoot]\n",
    "\n",
    "            EquidistantPeptides = sorted(EquidistantPeptides, key = lambda Peptide: (TreePeptides[Peptide][2]), reverse = True)\n",
    "            # PredecessorOccurence can be used to sort the Peptides, but does not seem to be useful\n",
    "            # EquidistantPeptides = sorted(EquidistantPeptides, key = lambda Peptide: (TreePeptides[Peptide][1]), reverse = True)\n",
    "            EquidistantPeptides = sorted(EquidistantPeptides, key = lambda Peptide: (TreePeptides[Peptide][0]), reverse = False)\n",
    "\n",
    "            AdditionalElements = MaxPeptidesNumber - len(EquidistantPeptides)\n",
    "            SortedPeptides_BY_DistanceToTheRoot[DistanceToTheRoot] = EquidistantPeptides + AdditionalElements * ['']\n",
    "\n",
    "            if len(PeptideTree.nodes()) > 1:\n",
    "                for Peptide in EquidistantPeptides:\n",
    "                    XCoordinate = X_0_Coordinate + DistanceToTheRoot\n",
    "                    YCoordinate = Y_0_Coordinate - EquidistantPeptides.index(Peptide)\n",
    "                    Positions[Peptide] = (XCoordinate, YCoordinate)\n",
    "                    \n",
    "                                    \n",
    "            elif len(PeptideTree.nodes()) == 1:\n",
    "                for Peptide in EquidistantPeptides:\n",
    "                    XCoordinate = 0\n",
    "                    YCoordinate = Y_X0_Coordinate\n",
    "                    Positions[Peptide] = (XCoordinate, YCoordinate)\n",
    "                    \n",
    "\n",
    "        #BaseRoundPeptidesGraph = nx.Graph()    \n",
    "        #BaseRoundPeptidesGraph.add_nodes_from(BaseRoundSortedPeptides)\n",
    "\n",
    "        Sizes = []\n",
    "        for Peptide in PeptideTree.nodes():\n",
    "            Sizes.append(math.log(Peptides_BY_Round[BaseSelectionRoundNumber][Peptide], 2) + 5)\n",
    "\n",
    "        Colours = []\n",
    "        for Peptide in PeptideTree.nodes():\n",
    "            Colours.append(min(PeptidesAppearances[Peptide]))\n",
    "        \n",
    "        XSpan = max(map(lambda Peptide: Positions[Peptide][0], Positions)) - min(map(lambda Peptide: Positions[Peptide][0], Positions))\n",
    "        YSpan = max(map(lambda Peptide: Positions[Peptide][1], Positions)) - min(map(lambda Peptide: Positions[Peptide][1], Positions))\n",
    "                            \n",
    "        XMin = min(map(lambda Peptide: Positions[Peptide][0], Positions)) - 0.01 * XSpan\n",
    "        XMax = max(map(lambda Peptide: Positions[Peptide][0], Positions)) + 0.01 * XSpan\n",
    "        YMin = min(map(lambda Peptide: Positions[Peptide][1], Positions)) - 0.02 * YSpan\n",
    "        YMax = max(map(lambda Peptide: Positions[Peptide][1], Positions)) + 0.02 * YSpan\n",
    "        \n",
    "        \n",
    "        \n",
    "        NumberOfColours = len(SortedRoundsList)\n",
    "        \n",
    "        ColourMap = plt.get_cmap('Paired', NumberOfColours)\n",
    "        \n",
    "        nx.draw_networkx(PeptideTree,\n",
    "                        pos = Positions,\n",
    "                        node_size = Sizes,\n",
    "                        node_color = Colours,\n",
    "                        cmap = ColourMap,\n",
    "                        linewidths = 0.2,\n",
    "                        width = 0.2,\n",
    "                        with_labels = False,\n",
    "                        #font_size = 6,\n",
    "                        vmin = min(SortedRoundsList),\n",
    "                        vmax = max(SortedRoundsList))\n",
    "                    \n",
    "        if len(PeptideTree.nodes()) > 1:\n",
    "            for DistanceToTheRoot in SortedPeptides_BY_DistanceToTheRoot:\n",
    "                PeptidesSummaryFile.write(str(DistanceToTheRoot) + ' mutations' + ',' + 'frequency' + ',' + 'rank' + ',')\n",
    "            PeptidesSummaryFile.write('\\n')\n",
    "\n",
    "        for i in range(MaxPeptidesNumber):\n",
    "            for MutationsNumber in SortedPeptides_BY_DistanceToTheRoot:                        \n",
    "                Peptide = SortedPeptides_BY_DistanceToTheRoot[MutationsNumber][i]\n",
    "\n",
    "                if Peptide != '':\n",
    "                    FormatedPeptide = HammingDistanceBasedFormating(RootPeptide, Peptide)\n",
    "                    PeptideRank = str(BaseRoundSortedPeptides.index(Peptide) + 1)\n",
    "                    #ClonesNumber = str(len(PeptideTree.neighbors(Peptide)))\n",
    "                    PeptideFraction = ('{:.2%}'.format(float((Peptides_BY_Round[BaseSelectionRoundNumber].get(Peptide, 0)))/float(TotalPeptides_BY_Round[BaseSelectionRoundNumber])))\n",
    "                else:\n",
    "                    FormatedPeptide = ''\n",
    "                    #ClonesNumber = ''\n",
    "                    PeptideRank = ''\n",
    "                    PeptideFraction = ''\n",
    "\n",
    "                PeptidesSummaryFile.write(FormatedPeptide + ',' +\n",
    "                            PeptideFraction + ',' +\n",
    "                            PeptideRank + ',')\n",
    "                            #ClonesNumber + ',')\n",
    "            PeptidesSummaryFile.write('\\n')\n",
    "        \n",
    "        \n",
    "        if len(PeptideTree.nodes()) > 1:\n",
    "            TreesXCoordinates += [X_0_Coordinate]\n",
    "            X_0_Coordinate += max(Peptides_BY_DistanceToTheRoot.keys()) + 1\n",
    "            MultiplePeptideFamilyCounter += 1\n",
    "            PeptideTreeSize += [len(PeptideTree.nodes())]\n",
    "            \n",
    "            \n",
    "\n",
    "        if len(PeptideTree.nodes()) == 1:\n",
    "            Y_X0_Coordinate -= 1\n",
    "            SinglePeptideFamilyCounter += 1\n",
    "\n",
    "        PeptidesSummaryFile.write('\\n')\n",
    "                    \n",
    "    PeptidesSummaryFile.close()\n",
    "    \n",
    "    #plt.axis('off')\n",
    "    plt.axis([XMin, XMax, YMin, YMax])\n",
    "\n",
    "    PeptideLegendColour = PeptideGraphFigure.add_subplot(1, 1, 1)\n",
    "        \n",
    "    ColourMap = plt.get_cmap('Paired', NumberOfColours)\n",
    "    PeptideLegendColours = SortedRoundsList\n",
    "    \n",
    "    LegendDotsX = XMax - 0.3 * XMax\n",
    "    YIncrement = - 0.03 * YMin\n",
    "    #print (YIncrement)\n",
    "    \n",
    "    LegendColourDotsX = np.array([LegendDotsX] * NumberOfColours)\n",
    "    #print (LegendColourDotsX)\n",
    "    FirstYColours = YMin + 12 * YIncrement\n",
    "    #print (FirstYColours)\n",
    "    LastYColours = YMin + (12 + NumberOfColours) * YIncrement\n",
    "    #print (LastYColours)\n",
    "    LegendColourDotsY = np.linspace(FirstYColours, LastYColours, NumberOfColours, endpoint = False)\n",
    "    #print (LegendColourDotsY)\n",
    "    \n",
    "    PeptideLegendColour.scatter(x = LegendColourDotsX,\n",
    "                                y = LegendColourDotsY,\n",
    "                                s = 15,\n",
    "                                c = PeptideLegendColours,\n",
    "                                cmap = ColourMap,\n",
    "                                linewidths = 0.2)\n",
    "\n",
    "    ColourLabels = ['{0}'.format(i) for i in range(NumberOfColours)]\n",
    "\n",
    "    for label, x, y in zip(ColourLabels, LegendColourDotsX, LegendColourDotsY):\n",
    "        plt.annotate(label, xy = (x, y), xytext = (5, 0),\n",
    "                     textcoords = 'offset points',\n",
    "                     fontsize = 5,\n",
    "                     ha = 'left', va = 'center')\n",
    "    plt.text(x = LegendDotsX, y = (max(LegendColourDotsY) + YIncrement),\n",
    "             s = 'first-appearance round #',\n",
    "             fontsize = 5)\n",
    "    #plt.axis('off')\n",
    "\n",
    "    PeptideLegendSize = PeptideGraphFigure.add_subplot(1, 1, 1)\n",
    "\n",
    "    Size = []\n",
    "    for i in [1, 10, 100, 1000, 10000]:\n",
    "        Size.append(math.log(i, 2) + 5)\n",
    "\n",
    "    LegendSizeDotsX = np.array([LegendDotsX] * 5)\n",
    "    FirstYSizes = YMin + 5 * YIncrement\n",
    "    LastYSizez = YMin + 10 * YIncrement\n",
    "    LegendSizeDotsY = np.linspace(FirstYSizes, LastYSizez, 5, endpoint = False)\n",
    "    PeptideLegendSize.scatter(x = LegendSizeDotsX,\n",
    "                              y = LegendSizeDotsY,\n",
    "                              s = Size,\n",
    "                              c = 'w',\n",
    "                              linewidths = 0.2)\n",
    "\n",
    "    SizeLabels = ['{0}'.format(i) for i in [1, 10, 100, 1000, 10000]]\n",
    "\n",
    "    for label, x, y in zip(SizeLabels, LegendSizeDotsX, LegendSizeDotsY):\n",
    "        plt.annotate(label, xy = (x, y), xytext = (5, 0),\n",
    "                     textcoords = 'offset points',\n",
    "                     fontsize = 5,\n",
    "                     ha = 'left', va = 'center')\n",
    "    plt.text(x = LegendDotsX, y = (max(LegendSizeDotsY) - 0.03 * YMin),\n",
    "             s = 'frequency in the last round',\n",
    "             fontsize = 5)\n",
    "    \n",
    "    \n",
    "    \n",
    "    #FamilySizeLabels = ['{0}'.format(i) for i in PeptideTreeSize]\n",
    "\n",
    "    #for label, x, y in zip(FamilySizeLabels, LegendSizeDotsX, LegendSizeDotsY):\n",
    "    #    plt.annotate(label, xy = (x, y), xytext = (5, 0),\n",
    "    #                 textcoords = 'offset points',\n",
    "    #                 fontsize = 5,\n",
    "    #                 ha = 'left', va = 'center')\n",
    "    for i in range(len(PeptideTreeSize)):\n",
    "        plt.text(x = TreesXCoordinates[i], y = YIncrement,\n",
    "                 s = PeptideTreeSize[i],\n",
    "                 fontsize = 5)\n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    plt.text(x = LegendDotsX, y = YMin + 3 * YIncrement,\n",
    "             s = ('total # unique peptide sequence ' + str(len(BaseRoundSortedPeptides))),\n",
    "             fontsize = 5)\n",
    "    plt.text(x = LegendDotsX, y = YMin + 2 * YIncrement,\n",
    "             s = 'single-member peptide family # ' + str(SinglePeptideFamilyCounter),\n",
    "             fontsize = 5)\n",
    "    plt.text(x = LegendDotsX, y = YMin + 1 * YIncrement,\n",
    "             s = 'multiple-member peptide family # ' + str(MultiplePeptideFamilyCounter),\n",
    "             fontsize = 5)\n",
    "    \n",
    "    plt.axis('off')\n",
    "\n",
    "    PeptidesSummaryFileNamePNG = str(today) + 'PeptideFamiliesSummary' + FileName + '.png'\n",
    "    plt.savefig(PeptidesSummaryFileNamePNG, dpi = 500)\n",
    "    \n",
    "    \n",
    "    fig = plt.gcf()\n",
    "    SizeInches = fig.get_size_inches()*fig.dpi\n",
    "    SizeDots = fig.get_size_inches()\n",
    "    \n",
    "    #print (SizeInches)\n",
    "    #print (SizeDots)\n",
    "    \n",
    "    \n",
    "    #print (XMin)\n",
    "    #print (XMax)\n",
    "    #print (YMin)\n",
    "    #print (YMax)\n",
    "    \n",
    "    #print (PeptideTreeSize)\n",
    "    #print (len(PeptideTreeSize))\n",
    "    #print (TreesXCoordinates)\n",
    "    #print (len(TreesXCoordinates))\n",
    "    \n",
    "    plt.show()\n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Set the parameters for the analysis\n",
    "\n",
    "DataFolderLocation = \"SampleInput_RawNGSData\"\n",
    "\n",
    "BaseSelectionRoundNumber = 6\n",
    "TopNPeptidesNumber = 24\n",
    "\n",
    "FileName = \"PHD2_NNK\"\n",
    "\n",
    "StartSequence = 'TAATACGACTCACTATAGGGTTAACTTTAAGAAGGAGATATACATATG'    # NNK - T7g10M.F48 \n",
    "StopSequence = 'TGCGGCAGCGGCAGCGGCAGCTAGGACGGGGGGCGGAAA' #NNK - CGS3an13.R39\n",
    "#StartSequence = 'TAATACGACTCACTATAGGGTTGAACTTTAAGTAGGAGATATATCCATG'   #NNU - T7-CH-F49\n",
    "#StopSequence = 'TGTGGGTCTGGGTCTGGGTCTTAGGACGGGGGGCGGAAA'  #NNU - CGS3-CH-R39\n",
    "#StartSequence = 'ATG' # Met codon\n",
    "#StopSequence = 'TGCGGCAGC'# Akane seams to have trimmed siquences\n",
    "#StopSequence = 'TAG' # amber stop codon\n",
    "\n",
    "QualityScore = 29 # threshold QualityScore must be between 0 and 93\n",
    "cDNA_MinLength = 24\n",
    "cDNA_MaxLength = 240"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# SelectionSummaryReport(DataFolderLocation,\n",
    "#                       BaseSelectionRoundNumber,\n",
    "#                       TopNPeptidesNumber,\n",
    "#                       StartSequence,\n",
    "#                       StopSequence,\n",
    "#                       cDNA_MinLength,\n",
    "#                       cDNA_MaxLength,\n",
    "#                       QualityScore,\n",
    "#                       FileName)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# DNAMutantsAnalysis(DataFolderLocation,\n",
    "#                   BaseSelectionRoundNumber,\n",
    "#                   TopNPeptidesNumber,\n",
    "#                   StartSequence,\n",
    "#                   StopSequence,\n",
    "#                   FileName)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXsAAAD/CAYAAAAQaHZxAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAIABJREFUeJzt3Xl4VNX5wPHvO5N9ZQs7hEWEsAkiIC5Rcd+w1bpRra0a\nbZsu1mqrrW1jq120tbYaf62pWrtQ94Vad0UjgojIJoZ9C4QtCdnXmTm/P84dMkSyTDKQG/J+nidP\nJpk7556b5Z1z3/vec8QYg1JKqaObp6s7oJRS6vDTYK+UUj2ABnullOoBNNgrpVQPoMFeKaV6AA32\nSinVA3TbYC8iF4rI7SKySkS+KyLfPcL7P1FEnhGRO0XkFhEZF8l2nccPisgpkWg3nH2H9uFICtn/\nGSLyExH5coTajfjPsRu1eZmI/FZEztQ2e06bh9Jtg70x5n/AWuAxYDiw6wjv/yNgFVAMJES6XRE5\nH1gRqXbD2XfIsR1RIfudBXwOZHS2zcPxc+wubQIYY54H/gKka5s9p81D6bbB3nE8UGuMuZ0IBIaO\nMMbkAb8Dro1w0xOBydhj7HGMMS9FqKnD8XPsLm0iItHA14F/aps9p81D6bbBXkSSgBrAJyK/APYd\n4f1PAGaJyJXAr4DFkWwXWA7MBz6NRLvh7FtEvu98PutI7Tt0/8AyEfkp9sytU4wx9xPhn2N3adNx\nDyDADG2zR7X5BaLTJSil1NGv247slVJKtZ8Ge6WU6gE02CulVA8Q1dUdCJcv79FBwG1ANPDXqKyb\n1rS4cY4kAucBK8kxGyPSgRyJAi4FisgxCyPSpm13NuAhx7zd1qbZmXlRuflZvgjuW4ATgW3kmKI2\ntvUCyUA5ORG64JMjMcBUYBU5pjYibdp2BwBl5Jj6CLYpETvuwylH4oFe5JgjWpKs3MtVF2izM/M8\nuflZgZae9+U9KsD7wKnOt9YDM6Oybir7wsY2gLwFZAK7gUvIMR+31Pb8uRk3A9OBn8+ZV9BywMuR\np4ArgTrgO+SYx1o5nj9i32zycvOzHmilzZ8Cd2OvyP+KHJPTSpv3A98APgAuby3oz5+b8X1gPHDH\nnHkF+1vZ//8B38T+nC5v8U0sR5KB/wEnAI+TY77TSj+jgfuAauBnuflZh/5Ds7+nN4DTgaXAOeSY\nL/4+7bZe4E/AQOAH5JjCVo7pfuD72Iqei8kx21radP7cjGuB2cCv5swr2NxKmxcAfwbKgKvJMRta\n2far2HLcx8gxz7ay3QhsyV0i8C1yzJJWth0L3Am8Ro55upXtjsVW9owCfk+O+UlLm86fm5GKLR1e\nP2deQct/o0B2Zt5vgAuBv+bmZ+W20qYH+DU2c3DnnHkF/lbavBnIBt7Mzc+6rbX9h/H3fCLwG6AA\n+z/aYkzJzsz7CnARcH9uflZrA8eh2Fr4OuB6ckxFK/08Efg28K858wrebGXfB/095+Zntfj3PH9u\nRjrwC+CDOfMKnmixn61wTRonOzPvL8Ae53NLUrEBOejYpVu2nC8ig5t/vFjAKdhAD/aHeU5r+zfG\n/Bi4gdbq5XPEgw0KAHH7qrn0UPsWkcFpScPTjQl8AxjXapvW2YAX8JTX8aWW2hSRwYGA/yqgL/Al\nYFJLDc6fm+EJGPNT4CZgbhv7P9/5PHBrGV9tad8rd3M59o02vraRq0RkSEvb7qsqnAvcAvyU1u+B\nmIoN9ADTl+7kipbafHMT52IDw2XAVW0c0yXYs79Jq/dwbWs/00Z/4E5snfM32mjzK8BoYNpne7m5\ntTZrG/k+cC5wczv6eQowdUMJ32mtzX3V3AJchz2zbc0FwFggusHX8t+oiAzeXdVwk9PHu+bPzYhp\nqcHszDwP9v9jks/fmNVam6t2V18I/Bi4nabfbUuuAyYZE7gxLWl4ekttJsd6h/gD5i7a9/f8VWe/\n3/rHSk5qra8+f+OPnT7c2Eabl2Hf6C5bV8zXWmuzusH/Q+z/fFt39U/D+Xuuqt/f6s+0tLbxO9i/\nz9vbaLNFrgj2zh/SZUA/4FLn60OpAD4LflFaXV3xnXnzegNTgGNDN/zjR2yp87ESoKIOH9BqymXT\n/rpX99c2LgCeb3EjO0L4GMAXgPnrAJv+mAIMCd20uLqwsbqh/P2A8e/DjrJa81HwwcLt1GL/UKcA\nxzTfsNFf/74xJlDfWL0Ue5fpIV3yn7WyubTuVeAZoK0blD4EqKyn8aElxGLrfacAg0I3mreaj+t9\n9ue/vZzVwMXOdpOB2NBtF215Nr/eV/s08AiwrpV9r8KO6Nm0n6qs/zLAaXMqNl10wC2v89nm/SzY\nX8vHtP0zXQBQ66PoJ+8Sg/3HOg5Ia77h/lrfv7E/o5ZHy9YbQEmdj8LfLqQGOMlpc3jzDfdU8yL2\nb+WpdrS5ut5H4T9WUgScySH+ngE+3smLVQ28AfytjTYXAIUAlQ0sam3D97dWvFnT6J8PPDJnXkFD\nS9s5Z9wvBgL+onpf9euttfnw0t0rqup9/3L6+UEbfX0JKGzw171WXF3Y2NJGVQ0BU1Hve5L2/T2/\nBCyvqOflu9+nxTM6gOqG/U9h/5b+00abrwIf1jSy8F+reb+1DXdUNDxf1+h/C/h7G20uB+Y1+hsW\nFpWve6u1DT/bU/Pf2kb/a0BeG222yDVpnOzMvL8IgUsNnhdy87O+2dJ2vrxHjwV+PO+jj77x8oqV\n/3vh00+fAPYAa4Ax2FSIH1htfsEPSmu5c8kOEvZWk3zdS4fO3YpIwrXHpV0xdWDi8NF94u5t7bTT\nyYVed9Vz3PX0Gh7D/kPvAopo+qevBj6/fOpd1ybH9p0dF52Ym5uftbSVNgX4yvde48cPf8xSA68D\ne7EpiDHBQwdWN/7MO/jzfbG/WlRY/+43X/H9vaUmReR04ANjTMvH0rT/KGDOLa8z8E9L6I29bb8Q\nKMG+iRmgElhnfsGZ2Dfme+RuaoAJwEZgADbgC7DRGFOSnZl3GVCTm5/1Whv77wWcMfpPXLO5jLed\n9jY47aU6WxUZY3ZcMvmHP4mLShzQL2nYbbn5WS0GB+cs7ERgvdxNCfaMsNo5joHOVmXAhm+f+ugE\n7BnLY7n5WS0GPKfdJKCRHFMvIpOAOGArMMLZoh747OWrx/XHngk8PWdeQes3/Nnfv4cc4xeRUdg3\n2S/8Pb989TgPcD2wYM68goI22hyIPQtZ1C2uMajDzjXBnhz5izFcJsLz5JgWg32QiPiAt4HnsP8Y\nBig3xhSISFRqLJP33sabMVH0BXhkKU9kv8pdLTQ3/amvjHkgPto7CvjenHkFD7Vj/69gA/Jr2OCU\nCGwxxux07u7NuHHWQ/NjomIHAgty87Nmt9JcsM0rgKuBl7Gjdi9QYoxZLyIxwMTCH3Dv0BTOq2qg\nIOnXZnwL7RwPbDamhdx3y/sX4AfAFucjCdhgjNkjIinA2Mo7eSophlHYnP0NzuvGAn2Aj40xfhE5\n5vQx187OGHByroinHpiUm5+1pR37vw4bND8FtgEpwKfGmBoRGXLCsItOOSH9on96xBMNfC83P6vN\n31Oz9vtgc75rjTHFItIbGHPjrD89FhMVPxG4Kzc/694w24zDXsPYbozZLiLxwITHLxl9T9+E6HOB\nJ+bMK7g+zDY92LOrA3/PwMQ/nz/yO+m9Ym8A8ufMKzgtnDaVckUaxxmFXSZi0zjO1+3RHzvySwBi\ngH0iciqQWnYHy2Oi2A5Q3QBvb2YYzVINjl5A2a6qhiqfP7CLVlIjzbyBTQn0wgbFCiBW7KyF9caY\npQHj2+5s22agczyP/Z2kYt88BCh32ow3xnw6NIWPAobAzgoOGchFZDRQGm6gBzD2nf91oDc2hVIC\npIrIyUC1MWYpsMEYfNiL48HXrcOmYmaJyMVAn93lm3cFjNmA/Xm2dyqLZ5z9xjk/g01AhogcZ4zZ\nOWPEnLca/XX7/IHGajowlYIxptQYsxBIE5ETscH045io+KXATkJShGG0Wee0Kc7vyRhjPumbEP0B\nTWec4bYZcCaFO/D3bIxZkd4r9i3sGWR7/0aVOsBVI3tsSeMLYYzstwP/ApZhA2o/4D3sBcGkZ6++\ndITft+mviwp9nj9/tGYB8KgzW2ZoOycDixKiPbddMb7v20+s2Lu8Pd0VkVTgBWx1SoHTlzhsHm46\nUHlMvxmDpg0//5K+iUNuz83Palf5n4j8HvvGERzd9gHysRdjYy6Z9MPlUax78IPNSz/dU7X7iWav\n7QcMMcasbM++WunDZUCts/9oYDUwEyge379PwyUZE4/99Xvvv+lsm4BN5QTTDSXAUGDlt099NAZo\nyM3Pqgxj31/HlgRvx1YHlWPTLZOBDeP6n/TV1Pj+fT7a+uJPO3mM0c4x7Xr56nHbfYHoAZc+tWpH\nJ9sU7Ci/7tunPvqZVxqG+03M9hYrkdrf7gTsgOLjl68eNwTYM2deQcspLKUOwR0je6scO7FZeRiv\nicP+EyRhR9jrgTOA9alx/Ze/se303+bvzk6t92Ynje43bQQwVkR6BV8sIkOwI7rjcy8cec2XMvp8\nOH9uxqXt2bExphybxgnuvx921DXTGLME2DtjxMV/6Js45LvA78M4pqXYUX2yc0wbsBVAa4A18dEp\n8wb0mvOdc8bf8v3QFznBK6OzgT6kD3HYM6e9wPHGmEVA+ZRhtzxZ5v/qq5dNueM+EZmOLe9bZoz5\n2BizzBiz1RizcPrwOWf5A41rjTErsjPz+oSx73zs8fuxo/s4IM0Y88GkwWcec/KoK38+ddh5t2dn\n5rXr99QSY0yjMyKvK63xveuVho3z52Z8rZNtGufsp7C0umieLxC9BTtJXqcYY9YAy395xrDfGmM2\nYwc4SoXFTcH+Sux8zleG8Zo4bGBIcB4Pw1a2nBrljYmK8cYJgNcTJdHeuGhsUA4t3RyBDfZeQYKB\npXcY+9/g7N+LvYA6DlgjIplASVJM7+ANQqktvP5Q3nCOJ9juaGy1zOmAv0/ioCrnmOKbve4UZ7tO\nM8Zsx47Q67FVIZtE5CRjzJ5eCQOrRMQ7MGV0lTFmqTHmM2O+WMc8Pf2iHR6JinPOgJr3tbV9b8Ze\nRE3Gnl3EAQERGX/q6Ct3R3mjo5ycfTi/p9b2t7N3vNcvIrHYktZItFnWJ3FwrTPSj1SbDccNTNzv\nvKn3avMFSjXjpmD/NDZt0Fb5W6gomnK8AaARe2q+oKR6x7Tq+rJfllTtrP9898LytXs+3IoNoANE\nZJCT296ELclb9sv3Ct95b2vFbcDjYex/KfYNJNZpeyf2AuDHwBlr9y7+QUVd8U+wN/e0i5Nr34e9\nBhGNPdM5EXgXOKlg94d3bdq37PGFG5/6v+BrRGQGsPRQQbcTCrDHVgSMxC6ocnqjv+E6bI3/Pa29\nODc/a6mInA2cmZuftbMD+07EvknWYH+3FY98cFPSsu2v/qFg94cPE97vqVUicjW2jPTBSLWJrZ/+\nCvZeg0i5D1uXH+m1E1QP4KbpEj7HBvtwLj55saPG4GinBhsoTjbGfPCNE//w675JQ2JjoxOi+ieN\nrN5btSXe2X4MdiS+EqgzxvjvPGXItMkDE4/H5vzbO6f4+9h/vuCZRX/stYMRQH6/xKHPJ8b0Oh47\nUn04jONahy1l7I0tF1wHTDfGvPelybfd1y9p+Jyk2D4DgQedSpidxpiqMNpvkzFmr4hUYcsUtzmf\nF49OuO/hcf3iJ4lIHWS90VobuflZH7X2fCs+xZ5R1GF/pjXA1rioJN/wPhMvSorpHYM9i9veShvt\nNmdewS7glUi0FZSbn1VLa/dsdMCceQUB2r6/QKlDctPI/nrsHa/tLVMz2P7HY0fW0diLhMcA60Vk\nUkJMygiApNje3hF9Jwk2YHmxqwJVY2+IWSkigycMSBiWGOOdgr2ppX0dMKYCm+5IwAamRmytvQBp\naYnDR3k9UUNpuuu2vZY7x+V32hoN7BSRsUN6jU2PjYrvlxzXb4qIDASijDHhjpzbazU2lZOGPcaU\ncf3ix4nITNq4I7kzjDH12DeYvthUTglw3AUTsuv6J6WPTYrrHayLV0q1k5uC/ePYi3PtnffBz8HB\n3o9NO5Rgb0qpr6rf/3R1fblvW+ln9St2vL0FW9kSg72YOhRbw26AoU+tLl62q7LhUez8F+HYgj1D\n6oOtCy9zPtL3Vm1/uL6x9hns6Xc4NmBTFx7nmIqdtr2F+wse31u57aOC3QtfA0Y5F+8OC2NMJfbN\nsRabohpX1RDIwc6l0uo8KhGwBfsziMe+gS58YeXvpuyu3PS7kuqi/yO8dJ9SPZ570jg55kngyTBe\nUU9ToE90PvbTlPrY/aNpP04dkAQr9yBb9/LGmn2cjB0RzsOmWoL170mxXk90YowH7JtBONZiL45G\nOftNxb6ZLBwa9+SVJw1Njk+I9sZBVrsbNMYYEdmBDfR7sSPcBuCzsppdpw5KHRPrN74hRGgpxDZ8\ngj0DGg8sXFRY8aNzj+mdgD2TOmyMMducFJUXezfvdGB5TX3F1xJiUu0dp0qpdnNNsM/OzLsOm8J5\nPDc/qz1BvxybWoil6WKmYHPx1cCUCf05G4g6cyTcM5sZX36aAmxu/2TshcfeInIscMxl4/tMTIqN\nOh17N+z9YXT9UyC4Vmt/7NQJBcAJmSNSzo7xeiZgR6jvhdEmTj8mYi9A78MG/RMmDj7jLBGZmhCT\nWrR02/zDfpOEMabBFpWwFph81qheX8JeMPYBtx7m3e/H/n4HA3vH9p/Vf1S/qVd5PN547Pwv/zjM\n+1fqqOGm0VG4Ofst2OAejQ32gk2leLA55s/3VduLvRtL8f8qn/7YUWIqNre+DTsz4OnAgILi2hX7\naxs/xU5/0G5O3r4Ce5ZR6ux/LFBdWF6/rN4X2EzHLv6tpulahAebdtpUXFW4tLahcldxdeGqDrTZ\nUctw5mkpqmx4t94X+JAjc6FwBTZ9VQXI+r1LBlU3lL1S21iVj50qQynVTq4Z2dNUStfenP1C7KyD\nXmxATMIG8CRslcbgr796w5v9E+Nmby/b3vDprv9GY6cCyMIe9yhscD4WGPTK+v2rvjG1/7re8dEd\nyYHvwJ5lDMKOwHcD+36zkGUnjbp4w5BeY/85J8wGjTF1IlKJvfAbPKaU/6358+Jpwy4cv3bPhwWR\nrRRstS8Bpy/b7lucsPnkUVfs6p0w6MNwj6kD+20UkRLsG2mSIbD008LXVg7rPaHitc8faX2RFaXU\nQdw0sg9qb2riDZpG9sHRb39sCqcPkJTeZ8YNCXHHxR474MLkY/pNn4hdzKQcGzyD5ZVbgJRvnjDg\n+hG94v5K2/OPH8ombOllNfZaQhIw9vwJ2ZcO6TX2V9iFFDpiE/YspcI5pn5fm3nfnElDzrjurLE3\ntjWnd0QZY1YDk84em/X13gmDHgJ+doR2vQP7+93RN3HI5JNHXXHryL5T/pidmXfREdq/UkcFNwX7\ncNM4wWAdhU3jeLCBth82R7653lfbANDorzd1vurgXOGraRqFG2yQLga2+wOmAhtcwlWAHX3GYy8U\nRwE7/IHG3caYGpy5xTtgO/YNZJBzTOv3VW3zB4y/zhCISI15mPb6Ag3FxphqOn5MYXHKShuBQdX1\n5TEB498VMP5S7NmTUqqdXDMRWsgF2idy87P+3p7XOJOh7QaWYGcs3AJsxlauSPbMiRkT0hofWbi9\nzj9v1bbfYi/K+rHTx67Bjpp3AWOiPZ6aCQMGv7+8qPCdjvRfRG7AplyKsDdRNXrwjJiefvG2j7a+\ntKIjbTrtfg17sTaAPZPh2H7D0jeV7Cz1BfwtLnl2uER7o7505jHHb3193ccdPqZwicg07BvzpoTo\nlEv7JQ3fvK10dasLSCilDuamkX1QuO8+HuxIOngsg7Aj7Lo/nv3Z1781bV3sI+dvi79iPDHYVMh2\n7JvCOc7jvkD85cf/8uKZo+58Pjsz74IO9ruUpjOFKCBh7vR7L5s69LyF2Zl53+tgm2ADfbrTpu+p\nr4yZ/uvZ8U88eP7w7E602WF/uSj9upunVnwwf25GWHO0d9IKwBsXnTTx8qk/u/O88d9+Jjsz7/gj\nuH+luj03Bftw0zhgg2vwAm1w6oQS7Gm/qffbpQKTYpAzRtobnbCrCoEd/Q/C3p26PSE6OcnriU6l\nacWhcAVzyzj92OsR73CPx5uInVumoyqxZwteoKakxjc92uuJS42N6pLJsNISYxK8HknCXuA+IpzV\ntrzx0ckSF53Y2yvevjRbBlIp1To3BfvHsetVhrtyejDYx2FTNANwRthvbOSRxdtpfGAxdd99jUXY\nfP4WbK19cCrlABD1zrrHn9m475PfEP4dtEHraXrTqQHS3ln3+GNlNXtvBX7SwTbBppvSsSmi5Lvf\nK8xbs7cm5+/L97Y6EdlhlIVdnPuXR3i/W/fX7CpatPm5h7aVrr49Nz/rv0d4/0p1a24qvTwdu0DH\nabS9UG9QcH4cr/N1LM7NUsCeM0cxo1cc0alxeG4/mc1yN9uBr2Ev6O7E3pX5CRB95fh9J45Pezc+\nNS4/A7I6srpQuYg0Om2nAJXfnlY9Jr3Xk+cmx3o/b2vSsFbaDYjIZuy8PkWnHXtH2tqKpK/0Si2L\nxlYXHVFz5hVsJ7w7nSPCGLPb4/GeeGz/maemxPXzZmfm/Sc3P0sv0irVTq4Y2Wdn5sVgp83tBVzi\nfN0ewYnCvM5HDDbQlwCeGC+zAMb3xwvMwlbiGJrmsQlgzwhKJ/RPOD41LmoWcGEnDmWn06dGoHFI\nSuz5ybHeGdipbjujAXsh2RMTFX91Ulyfib3jBx62icjcakDyyKi05OEnx8cknwGc29X9Uao7cUWw\nz83PagBewt4e/7LzdXsEJ0OLwgZtH3bSrnTAu7GUV3dXYd7YiA97Q1Uwb78Ce/t/P+ycL+uf/7xk\nw4aS2jfo3Dzp5TTNVhn13tbyFyrr/f+k83c/rcNOBsbnuz94trKu5NnCss/DTXd1e7srNi3aXLzi\n2fLaff8Enurq/ijVnbgmjZM7+6Yb99X0+1VaQvG2MCYN82FH9FHYEX4y9oam9UCsMewtrSFQUosZ\n9keGYMszrwb+hF2sYg123pUz1+6fUBuInlG1uXZcbSfuDN2ADcoDgH0+vymuavDXJcd6qzve5IEU\n0T4g0WB2VdXv99b5qiI6f313YIwJnD7m2hqfv8GH/d0rpdrJHcE+R7zAy2kJxRcAr5Ijl5Bj/O14\nZSP2GIJpnAZsamYosHVMH65NiME7vj/eQIAvX/sSD2JTRbuxKZ86nJTPzPQ5UxJje6VglzUMZ83Y\nA4wxZSISj73zVWaPSr2pb0L0CU6f2lxEvQ21ABMGZt6WGp92ae+EQUOAf3ayzW4lOzMvOWPAydc6\nFU4fAf/u6j4p1V24Io2DHV1fgB2dX4AtiWyPOuc1HmzwjscG/wogbX8dGwA2lmJOH8kz2PVtC7Fv\nBoOwd77WAN5dFRt3V9aVbseuPtUZW7Cllv71JbXraxv9BURm0q5SILW4avvSBn9dQVV96ZGY3tht\nqj0e70vYdQ8ist6uUj2FO0b2toLmVWygfxV7MbI9qrA3RXmxAT9Ydx8LFL++kWdH9mL2u1vw37uQ\nCppKL0/G3lDlwY70ty/Z+vL69D4Td6/c+fbSTh5LvXM8/XOX1r9z4sgLV41Jm/5SBCYN2wQMf3fD\nk++fOOLLvTbuW/raka9+7Fq5+VkB4Jqu7odS3ZE7gn2O8b8yd+xDIsYYIw9fNG9de1I4YEe7I2i6\ngzYWO9JPALh8Aj9IicV78nC8KXHM/fHblGGrcGKxI/rx2NLLPhdO/M7M1Li0ftmZeYty87M6c/Fz\nI3YOnh0XTvzenX0SBs3CVgjd0Yk2gwuaVF123B0390kcfP3oftNm0jSPvlJKtcotaRwCeG7wG+9F\nATw3hPGy4DS3wdkvDTadUwkMafTTCNDghzX7GICt9nkfewHVg520LAnY6PM3GGOMD5sC6jBjTJnT\ndhyYCmfZw061GcLf6K+vBQIJMSkdmbBNKdVDuWNkD0R7/I8nx9SlVDbEhVP6uJ6mnL1gR+wBbGqn\n4O78wQ+nxI95ZGPxFt/Tn233YOvg+2EDfADYgy2X7PviqvsX9k0YsnFXxcbnI3A4W4Ehzy6/92+z\nRlx23+QhsxdEoE2Az15Y+bvnTxp1+YvHDTlTJwJTSrWbO0b2ORJ9/jEbbjtleOG55x+z4TZypL3r\nmy7n4JuqwAb8KmBQlfzwuv3+r0bFJX43dkjquHjstMjp2NF/cLGTfkDR+eOzp543/tvfzM7Mmx2h\no9r5pUk/vGz8wFOeAW6MRIPGmPqLJ91y+oSBmfOAeyPRplKqZ3BHsLerRZ3pPD4TuwRee6x2Pnto\nmjbBjw3kxiNRQwASopPllNFXLsKunboGewftNmzQ7wds7xU/IDUhJjkFu+5rZ20AhsVFJx8b5Y3p\nC0yOQJsAJMX2GRPlje6PnVpCKaXaxS3Bfj3wrvP4XWywbI/tNM2PE0XThGgAZYX7P39iT8UG/8qd\nb/v7Jg55ERiHnfveYG/KScDm09M/2PSfd9bsev8l4JHOHoyTt6/O3/jvl/ZUbvkD8PPOthm0vPCN\nn5fX7ssBvh+pNpVSRz/X5Ow7whhTKyKGplSOwY7sA0D6g18dMmZ4nz7eT7dL4LlPnh6KDfBXAC8A\nJ2DrtccB/zsrfWf62H6liWP6LhsGWVsi0L3SC0aXDJg66OkZg5JjXoKshRFok9/O3pe2v/aJ03vH\nR22ArPa+KSqleji3jOyPBYK58tm0P40DB898GY3N2ccCu9OSkzNjo6OZNXq0Z0S/vjcAH2MXKV+A\nvdt2IPYsYvyp6SnjxvSNn4B9M4iI6UMSTxuUHDMduC5SbQJf6x0fdTphzCmhlFJuCfbrgeBygO/Q\n/jQONC3X5wn5GiBl87597++vrubdgrWBB95483NgLHa92W00nQ1UAGmvbSzbtHZfzXIidwv+xrc3\nVSwprWl8DXg0Qm0C5GH7+FAE21RKHeXckcbJMY3kyOXYqYCfI8c0hvHq4MyXwSmOY5zvxd39UsW6\nhOjGwK7ynYGdZRVl2AuztcBUYC92dalJwKuLdgyau6d+WuG2+pP2RuBuV4wx+09JT9lUWNEQ2ych\nOmKLg8/Xy9ZnAAAgAElEQVSZV/ApehepUipM7hjZ50gy8AZ2BPyG83V7BWe+DB6Lwc5+WToweVRW\n/+QxnuOGnhU1a+Slx2Hvbt2NXVJvJ/bu2xXASaeNuWbouAEnnQt8JxKHBHDdcWmXHzcw8RoieIFW\nKaU6wh3B3ubRpzuPpztft1dwqttgwI/BSc00+uu3AVTUFpt6X+372GkLKrEpnG1AGrAZGFNRV7yp\nrrF6BzavHxGltb7VvoDZgq0AUkqpLuOONA4sApYB05zPi8J4bQNNc+IEJ0PrB+wYFPOrJ8b2Tj45\nv6LG/8n2io9F5FpgPvBV7MXcBOxof9v76x/oM3No8nOvbdgfkaoZgD8u3vWHi8f2/vDisX161FTE\nSin3cUuw74w6bG19aEVOJWC+N6Pmx0NSarxXjMdLjgRvQkrE5up92HTPaKDo3jOHf2tgUsyX5s/N\n2DpnXsG8SHTsj+eNuCsxxns5MAT4RSTaVEqpjnBTGmea83ga4aVxarGj+tCRfRIQG+O1KR4R+Ncq\nhmDnwhmMvT4wxnlNFFDh9Yhf5EAZZ0QkxniDyyu65eeslOqh3DKyXwQsxebrlxJeGqeapmAfXKLQ\nD/DbD2Mf7p2Ykbtu71b/v1aVleKUaRpj6kUkOEtmHbDxrne2v33ysOTNL64t/VfEjgpuAB6j8wui\nKKVUp7gj2OeYyre+Nvoe4IfAH87+x6bKMF5dRtOoPjgZmgHMzvo7rq2LGuINxOzxpCU9Nn5f1dYX\nsLX2YFesOg3oDyyYMfLbJ/ZKHHZudmbeu7n5WeG82bRozryCeuwNXEop1aVck16o9cVcU+uLyaz1\nxYRbQ76Pg1M4Udi8fFRiTK/+AKlxaTIoZfQo7FlAcGGUVGwZ5ghjjK9f4tBBCbGpg4FZETgcpZRy\nFdcEe+ydoc86n8MRXMAkeHEWbFD3bilZ+UJR+Qb/qp3vNK4uWrAQO6pf52yT7uxrFsDiLc8t2Vyy\n/AXg4U4cg1JKuZI70jjAnHkFb5Ejb5NjTJgv3UJTzh7sG1gKsD8+OmlESlyap95XQ0JMSkV1Q1mi\nMaZaRILpnjVAnIiknZj+5dj4qOQ+2IVPir6wF6WU6sbcM7LPkYeAIudzONY7n0MXMakHUgamjJ6e\nFNtLRvY9znvSyK+MwaZ5AE4FVhtjGoES4LjxgzInD0o9ZjpwbaePRSmlXMYdwT5HBDsvzkDgK87X\n7bXJ+RxcwESwd9HW7a/ZtbLR30BR+frA1pKVC2ga/adj58IHe3frpN0Vm95q8NW9CkRiWUKllHIV\nd6RxcowhR57DTi/8XJipnL3O59CKHA/AjrKCjQ2+Gv/uys2BDcVLxwANIhIHhE60tgEYU1Pzj8IT\nRvXbMyg5ZovOHqyUOtq4JNiLB3uXaW9gCDniIccE2ngVAMaYOmcBk+CoPoBN1/SaOOj0L8fHJHuH\n9ZngjYtOGvrh5mfWAGdx8BTKfiD+5mkDvpIQ4x2BTQHp9MFKqaOKO9I4di6bi7HVNBdhL5KGw8fB\nKZxooKTeV7MfoKahwuyr3L7PGPMp9u7awSGv3Q54Kxv8O/wBUwQUdOpIlFLKhdwxsodiX8D7SnFt\n3wv7xZf8L8rjLwnz9T4OXsAEILW47P7npqf1Ov79fVX+9ftKS0TkGGAxMAEYBnxijNkrIom3v7nt\n/kvG9d7+5Ip9KyJzSEop5R6uCPbZ7z7qAZMAEu189uTm0K40jiM02EfhLFX4lwsrrhnZu8J77SS8\ng5MZ/MhSPNhZMlcCw0RkkjFmNVD90AUjf54c650wf27G3DnzCl6M9DEqpVRXcksaZxTIOfahnAOM\nDPP1jTTV2hvsLJjVsV5iAeKjwRiGYdNDUdi00TKgRERmA0safIHBYqdK7tP5w1FKKXdxS7DfDLzp\nPH4Te6NUOOo5eGRfD/T9Zb7851+rEv0/eD2q8f8+YSm2zHI8MAXIwM6CKcBxv1+868lNpXXXAo93\n+miUUsplXJHGyc3P8r1w1aQ/GuP1i/j/fOlTq31tv+og9djRPNjgHQ/s2+f70aWLi0d7y0yhJy3p\nnwl7K7eUiMjn2NH/emzQPxPYlp72tW+tq5o0e1NtTMEc+DRSx6aUUm7glpE9UR7f9dHe+vOjPL7r\nO/DyWppq7AVbThmdEpeWDtAvaZicfsxXPSISTPOATf30Bl4BnhyYfMzIKG/MDGzwV0qpo4prgj0d\nnwgNoIqmYwkG/bitJSvf2Vu5LbBm1/uBtOT0Bdg7Z7dhz2jOwE4/3GCMqVu58613quvLHwL+2tkD\nUUopt3FFGgecidDgrQ6+PDj/vTfko+aqcf9MPGkosrSIwPVT/80jH9APOyXyccCj2CUJNwJcMGpN\n/ZDkDVOHpsb2h6yKTh2MUkq5jGuC/fy5GWdj5ynIcwJ/OIILmAT5gcRLjmXmwGRk0gC82Px8MjAV\nWGaMMSLSyxizcf7cjKRpg5MuifJIPPBN4LbOH5FSSrmHm9I4WcDldGximmJsLj5YkRMLBNaVsMEX\ngI92EPjy05QAw40xCw/x+uryOt97xphVwGsd7L9SSrmWa0b2NOXqO5KzD5ZQBuep9wGBm1+dvmby\nwGFnrS3eFVi9a3Ef7Pz1XzBnXoH5/omDfzOmT+zY4b3i3u3A/pVSytVcE+zf2HHrWcAMYMuc8HP3\nO2hK4wh2ZB8zY8Sl5yfH9fX2Sq73jhswu/7ZFfeKiCQBlSISj11sHIDZI1P+LCJTsD+Tv3T6gJRS\nykXclMa5Elstc2UHXruFpvnsg2vR0uCvqwWo99WYpdtfScPOdjkIu/bsOJqWKKTBb0oCxlRg3ziU\nUuqo4qZg/zS2LPLpDry2kKbpEoK19ny8+Y//6+u5379p9wMNW0tX1hhjGrALjVcAUc5KVQBc9+LG\nuzwiQ+fMK3ilsweilFJu45o0zrlDH3gbOyfO2x24RluKrcAJvnl5gaj/Xl1x3oT+Fd6sqXiuf5nk\n4MZOJc5BC6TU+gKBOfMKKlFKqaOQa4I9TdU4EGbO3hhTLyLBWTI92AVMJCGaXgD9EpCMtKZg7zhQ\nqikiY2ha3lAppY46bgr2nanGgaZpjsEel/9nC3jrgjGMyt9G4K/LKPqzfc6IyEBs3j6ojzFmA0op\ndZRyU7DvrODkacHyy9hrJzMiMx3v0BRkxW72Od8XYCh2imNEJHgmoJRSRy03BfsOp3Ec9TQFegEC\nGWlMiI+GzHQ8H3yDxJBtxZgDi5pPAlZ3qMdKKdVNuKkapzMToYFdgSp4B60X4OW1LPtsD+aJ5fij\nvS2uLRtrjKlr4TmllDoquGZk38mJ0MDeIBXjPDZAbHk9JftqMLurD3wP7JuBD0BE4rBnBEopdVRz\nTbCfPzfjOuB64PE58wqe7EATddgaesEeV8PcSZw1qjeezHQARorINmAITW8qk3By90opdTRzUxrn\neiDT+dwRNTTdQQsQu6GEnQEDy3cTAIqA/kC9MabK2cZjjNGLs0qpo55rRvY0rf36RAdfHwzgwTtp\nJet/GfnHDc6YtWHfZv/avcvruFv6Y2/AQkT6AiWd6rFSSnUTrgn2b+y4tQg7e+XOOR1ropKmQO8B\nGk8edeVlfRIHewek1HizM/MGANE0lWiONsZ83OmOK6VUN+CaYE/nSy/LnM+CrZuP9xu/H8BvGg12\nzdkEOFBvr5RSPYabgn1n76AtpansMgrwvbPusecnDzlr7ObiFb5tpStLH5GbhgMfikg6dvI0pZTq\nEVwT7M8d+kDIVx1ZrIq9hMyLA8ipo68+b0ivsd6ByaM82Zl58dibqfwiMtAYs6TzvVZKqe7BNcGe\nzqdx9mJr6aOcz9HxMSm9AZLj+gmQAoiISMtNKKXU0clNwb6zaZxCmhYuiQYalxe+/u6wXuNH7q7c\nFPhs13t7ne0yaGF5QqWUOlq5qc6+s4qwgT6YyokemDJ66ODUMZ7Bqcd64qNT0rEzXSaF1NkrpVSP\n4KZgH0zjdChhD5TTNHulB/AMTj12QlJcHxnd73hvep9J451t/J3vqlJKdS9HTRonZAGT4MyX/sL9\na1YLMnxP5ebAhr1LarBTJcyPQF+VUqpbkaaZfrs/EVmFvZN2N7Dn5mnEnDOa6xYVEvjDYq4C9htj\nFnRtL5VS6shzzch+/tyMs7EpnDxnBsyOaMSWXRqA789kdkYa3gvH4F1USPriHayNUHeVUqpbOZpy\n9tA0XbEAZls5xQAFxZj1JdRCi3PaK6XUUc01I3s6X3oJdgGTeGzePvqaFwe9f/ro46as3rUxUFK7\nsdYcTTkrpZQKg2uC/Rs7bj3wuIMToQHUYue/MYA5Y8zXL+qfMjIqOaGckpo/1XS2j0op1V25JtjT\n+TtowS5gAsH0lHjE+Ux8TNLeFl6jlFJHPTcF+0ikcaqxgV4A74L1T746YdBpo7aUrjQ7ytZWd7qH\nSinVTbkm2EdgIjSwwR6ckf2MEZecOrLvcVH9k0cE9tcURXemf0op1Z0dbdU4FTRNl0BqfFo/gL6J\ngz2DUo5J7XQPlVKIyE9E5Bzn8YXt2P6uw9+rI0dE7jjE924UkRu7oj/t5ZqRPZFJ45Q7nz2AZ/XO\ndz8c3mfS0D2VmwMb9n28u3PdU0qJyGRgGrDIWRficuB/IjICuAt4GzvYSgPeAYYDU0VkijFmRUg7\nM4CTgGJgFHZuK5+zffCxHxgMbAW2hGy/MGRfm5u1swv7/78EuAD4BBgQbMcY83RIH0L7XA+MxK6L\nMdQYc4/zJuUJadMPJNEsborIZcDp2MzC38L8kR4xrhnZz5lX8NaceQVXdOKGKrCrVQVz9iY+JqVP\ncmwfSYrp4/F6ovUCrVKdZIxZBSw3xrxnjNkGbAx5+gNjzFPAZ0AqEGeMedPZfoWIpIjILSKSii2R\nrgbGY/9fHwP6N3s8E7tUafPtQ/cV+n2DDbZpwDnAA87+Q9tpLthOhjHmAeybQrBEOziLbrDN/saY\nB50+hprtfO/ddv4Yu4R7RvY5cuAOWnJMRwN+KfaX4wW8o/tNy+ibOMTTO2EQA1NGBdp4rVIqDCIy\nFTtqPw9YS9Mkg72xKdWRwApgmIgcb4z5FHjQee0xQA0Qi62i+x6wH4gLebwNG3yXA6HbE7Kv0O/X\nGGOMs2TFW8CPRGQJ8EFIO80F2ykQkVuBnUC1iNwM9AJKQ9osEpHrDtHGcoDQswY3cs/cODnyDPaU\n8FlyzBUdaUJEzgZ+jz0N3HfKqKsGjeo75cxdlZvMmLTp0bn5WRrwlXIZEbnLGHNP88cqstwzso9M\nzn6n89kLsLn4090iEthdvjHwZsGjGuiVcqHQ4K6B/vBxTbCfv37cgceduIN2H00ToXlmjbps5oDk\nkd7RaSd4szPzEnLzs/QuWqVUj+SaYE9k7qAtxwZ6AahpKC8HqKorDSREJzd0uodKKdVNuSnYdzqN\nY4xpEBEfTrB/e+0TS6cNu2DyxuKlgX1V23yR6KRSPc38uRmxwD+xJY4rgGvnzCvYH3xeRO7Ezkv1\nujFGpxF3KdeUXkZQA041zrkZN51y/PBzo88ee0NMdmae3kGrVMdkY8+6hwAXArc2e74YOwGhcjE3\nBftI3EELdgGTAECUNyYGIDoqDtx1FqNUd+Jt9vVB/0vGmDzgd8C1R6xHKmxuCoCRqMYBW7MbDch7\nG/61eEzaCaOKyjeYHWUFtZ1sV6me6q/AWcAsYCXwUOiTzh2k04BFR75rqr3cFOwjpR4b7L3jB548\n4tj+J0b1ThisZZdKddCceQUVwLnz52YkAjVz5hUcdHOOMeZ54Pku6Zxqt6MxjVOHMxnagOTRQxJi\nUhiSeqwnOzPvaHxjU+qImTOvoLp5oFfdh5sCYKTSODXYahzPyp1vra5pKBtZVL4hsKroHa3GUaqD\nsjPzLgcygU9z87OeaP68iDwIPGeMWXjEO6faxT3TJUSIiPwFyAD2XjOJ6KsmctH72wjc96GJ6eq+\nKdUdZWfmXQg8h523xg9cn5uf9Y/g8yJyPnZmyY0a7N3LNSP77My864Drgcdz87Oe7ERTlTjVA9+a\nzsSThuGdMQQvORJFjtHRvVLhOx4b6MH+b00B/hHy/ERgEJCCnX5YuZCbcvbXY08Tr+9kO9XY0ktZ\nsIUd28sxb2zCr4FeqQ5bgJ1RFmyadEHok8aY+4H5wKdHuF8qDK4Z2QOPO5+/kA8MUzl29CH3L+6/\n/aVNM/27ytcHrulko0r1VLn5WQuzM/MuAk4DlubmZ73TfBtjzHtHvGMqLG4K9kGdvYhQgc0rejOP\n+erMYb0zoirqSkx2Zp43Nz/L39aLlVJflJuftRhY3NX9UB3npmAfTOMAdCZnX4w9Lqn3VdcB1DVW\nmpS4vlprr1RH5UgCNle/mhxTGfqUiJyGnTdntTHmla7onmqbm4J9JNM4BvC9vfZvS4urCjM2FS/z\nl9XuObrKjpQ6UnJkKPBfbLBfT45cSo5ZE7LFR8CXOfRKUMolXBPszx36QBF2ncidnbyvahe2zt47\ne+z1Jxzbf0b0iL5TvNmZeZKbn6UBX6nwfQ0b6AGOBb4B3BZ80hhT7yzpd1MX9E21k2uCPZGZzx6a\nqgZIiE5JBEiMSRGaVodXSoWntNnX+0O/EJG5wCRg1RHrkQqbm4J9pO6gLXc+e/I3/WfZ2P4zR+ws\nXx8o3L9GA71SHfMoMA6bl1+JXef5AGPMvK7olAqPm4J9RIQuYHLtxF39rj/+Jc+CLV3dK6W6sRwT\nAG7p6m6oznFTsI9UGgfsnPZy0VjSpwzE0zvOrlyllFI9lZuCfaTSOGBnvpSHPmZdVQMj391C4OEI\nNKpUT+TLe1SAu7FpnFXAj6KybjpwR7ozn/104C1jzBduuFLu4JpgP2dewVt0fkQf1ABEl9XRUFoL\npbWdvlFLqZ7sG8DPnMdnYive7g8+aYx5XkSWAbO7oG+qnVwT7OfPzTgbm8rJcwJ/Z9QBsT85hYnn\nj8F75khXzQGkVHczqNnXA0K/EJFo4OvAvUeqQyp8bgqCkVq8BGyw58NC9hWWY97fht49q1THPQWs\ndR5vB/7T7Pl7sPe2zDiSnVLhcc3Insjm7GsBue/D2M0vbpg+vahsg//rEWhUqZ4oKuumTb68R08F\nTgY+icq6aWfo88aYH3dNz1Q43BTsI6ka8Jxx7NdPOCZtWnRx1XZvV3dIqe4sKuumYuDlru6H6jg3\nBftIll7WAcbnr/cDNPobO9mcUj3bf7fePwk7UeHSi0fc/nHocyJyInCrMeaKLumcahc3BftIpnHK\nAM+CDU+uLCz7fMzOsrUB+G0EmlWq5/nv1vunYxcnGQiU/Xfr/VdePOL2N4PPG2M+EhGdKsHl3BTs\nI6kGMCeOuHTsxMFnxOzYX6AXaJXquPOwgR6gl/P1my1vrtzoaK3GKQe8vRIG9or2xtA7YaDeQatU\nx63B3pUOdvrwdaFPisgEYJaInHWkO6baz00j+0incQKLtjy3en9NUfqOsnUBuCsCzSrV81w84vYX\n/rv1/m8Bp2InQns09HljzBrgwq7om2o/NwX7SCoHZEDSyF4DU47x+AM64aVSnXHxiNsfAx7r6n6o\njjta0zh7AM+Y/jNGDE4d4xk3YJaWXirVSf9ZvkPTod2Ym0b2kU7jsHbPh1s84h22s3xtAO6IQLNK\n9Tz/Wb4jDvg3MOs/y3esBK65eurQkuDzInIGMAsoMMa82EXdVG1wTbCP8ERo5QDxsqzq9MHLzHsN\nOhGaUp2QDVzqPB4E/ICDL4LNAj4HxgMa7F3KNcE+khOhGWMaRaTx3jOZOGcsUXPGarBXqhOap3u/\nkBY1xrwkIuOPUH9UBxytOXuAhlV72F9WCyt360RoSnXC/wGvAxXAB8Cfmz2/RER+StNkacqFxBh3\nDHojPMUxIvIcYNJ7Db94Z0WRafQ3xne6k0r1YP9ZviPu6qlD67q6H6pjXJPGeWPHrQcez4lMk3Uz\nR3x54tSh58ZuK12lI3ulOkkDfffmmmBPZCdCA2jw4BEBPHjccfqilFJdxE3BPpKllwB1i7c+v2nb\n/s/G7q3cauChCDWrlFLdj5uCfaRVj+1/0oApQ8+J2V66Sm+hVUr1aG4K9pFO49QNTh3Tt2/iYGn0\n1bqp6kgppY44NwX7SKdxqpdsfXFjZV3x6MKyAr/OZ6+U6slcM+LNzc96Kzc/64rc/KxI3UVbnRjb\nKy4lLk2SY/vqnB5KRYiI/EREzumC/bY5s6aI3BX6uY1tTxaRNkeB7WmrvUSky+Ztcc3IPtJ19kDN\npMGzh48bcJK3T+JgDfZKRYCITAamAYtEZAR22oS3gWhgMLAVO/f9cGAYsN8Yc48TMLeFbDMW2IUd\ncC4BLgA+AaYC9wO3GWPuC9lvOjbN+z8R+XnwtcaYv4Zscw4wVUSm2C/lNux61FXB/Rpjng5ub4z5\n0JnXBxGZAZwEFAMNQBrwjnMcU0VkijFmRci+Qo+9HhgJlAJDQ47XE3KMfiCJLoy5rhnZE/k7aMvW\n7Vm8e0vJSv/a3Yt0EVqlIsAYswpYbox5z/nWB8aYp4CZ2NlmU4FxxpgHsRMSBsuePc22McDfsEH1\nHOABY8yb2BWwfgh82my/24CNwS9DXhu6zZtO34JB+QGgb7P9tiQe+8YwHvjM2TbuEG2GCh57hjHm\nAewbSujxhvazv/Mz6bKBp5uCfR7wLJHL2VfuqdxcUVj6WWBH+Tq9qUqpwyNY6fYB0BtYD6wXkVuw\nSxiuE5Gbnceh2xjTdPv+W8CPRORcY8xy7MLm74buRESmYkfY5zV7bXPDROR4Z5vg/33ofpuLEZGT\ngGOwy5nGOttWYEfroW22dOwFInIrsLPZ8Yb2s0hErmuhz0fE0TxdwszTRl/zjwmDM48tKl/vf3Hl\n/a5JWSnVE4jIHcaYsCojRGQWMNEYE6lBn3K4KQBGuvSyuLRmR1V1fZkpq92jI3uljrBwA73zmsXA\n4sPQnR7PTcE+0qWX5at3vVe0uXjF+BpfhQ/+EaFmlVKq+3FTsI+08mlDz08/If2iuC0lK31d3Rml\njgICujZEd+WmYB/RNI4xpvGU0Vd5vZ5o4qIT9Q9UqY5LA54EjgcKgOuBLcEnReRE4FbsvPeznG3K\n2nqsSxgeWW4K9pFO47Bo8zOfFZVvGLW7YkOdrcJSSnXAXcD5zuMBwC+Ba4NPGmM+EpFVHLw8Ie14\nrMH+CHJT6WXEHZM2vdfUoedETxx4emxX90Wpbqxvs6/7tLShMealcB6rI8dNI/tIV+Mwos/kAQNT\nRnk94nHTcSrV3TyPXVMoGagDDgrWIjIBO6r/s7M8YQFQ3o7H6ghyUxCMeBpn6fZXVlQ3lI8rLCuo\ngXsj1axSPc2LQAlwIrASeCP0SWPMGiA4b03oc++047E6QlxzU9XhMKp37F/OHt3rGyt2V1cv2VHV\n4qmnUkod7Vwzsj8ME6Fx9cR+k2cOS4kZ2y9eJ0JTqnMmA6cBS4GPQp8IpxpHK3C6jmuCPYchZ7+0\nqPqzpFjvCSt211REoj2leqjpwH+xlTjlwFXA68Enw6zG0WDfRdxUjRPpidB4d2vN2r+vGtLw2kaq\nI9WmUj3QudhAD3Y2yBbnstcKHPdy08g+4s4dd1PmiL7HJRaVrXfTm5pS3c0a7Bz10dg7aA+qpAmz\nGkd1ETcF+4incep8NbsbfPWmtrFSp0tQquNeBG4CTgVWYOdoP6AD1TiqC7gp2Ee89PLd9U+8+lHM\nC5fXNFSUwl8i1axSPdHfnQ/VTbkp2EfceRnfnjk4dUzvwv1rjurjVEqptrgpCEY8jZMan9YvLjpR\nUuLTtPRSqc65FXuhdjFwNyGzX7ZVeqnllu7gpmAf8TTOq2sefmBY7wmX7ypfvx1+HalmleppTgR+\nA8QAZ2PXh50ffLIdpZca7F3ATcE+4s4Zd9PIxkB9QkJMSq+u7otS3ZiXg8u0W4wbxpiXRGR888eq\n67kp2Ec8jZOWNOwsjycqNiE6VYO9Uh33IfAz4DxsGuegkXo7Si+VC7gp2Ec8jePxROWu3b3ohqKK\nDZvh55FqVqme6LfOxxe0s/RSdTHXBHtnPpyIjOhDVNX5qn11jVWNEW5XKaW6FdcE++zMvAMToeXm\nZ0Uq6N86ZejZ/Yb1Hu+a41Sqm/Jgp0zYC/hDnzhUNY5W4LiPm6YRCObssyLY5oLS6l2V+yq3bWl7\nU6VUK/4O7MAuXHJQKbMx5iMgtBon40h3TrXNTSPeiOfsgbeWFb6ypah8w2fwRASbVarHmY0dHJ6B\nXbHqkDPJagWOe7kp2B8Ot509Lmvy3sptunCJUp3zGDAXeA2oDH2ihWoc5TJuCvYRL70Eqo0JBGKi\n4ooi1J5SPdUvnI8vaKUaR7mIm4J9xNM4uflZjwzrPX7k3sptK/7NXZFqVimluh3XBPtzhz4Q8lVk\nrtFmZ+Zddm7GzTeW1+5dAvw7Io0qpQ4SrMYxxlzR1X1RLTvaq3GmxEYl9OqfPCIxgm0q1RP1BrKB\noc2fCKnGUS7mmpE9h6ca5zdAMXonn1Kd9VvsAianATqC74bcFOwjLjc/qwb4U1f3Q6mjwBrsDVVf\nqLQJVuOIyFnGmLePeM9Uu7gp2B+OahylVGT8GXgYCDR/olk1jnIpNwX7w5HGUUpFzhcCveo+xBjT\n9lZKKaW6NdeM7OfPzTgwEZozA6ZSSqkIcU2wR3P2Sil12Lgp2GvOXimlDhM33VSllFLqMHHTyF7T\nOEopdZi4aWSfBzyLpnGUcjUROeSsgof6voicLCKHXLtWRG4UkRvD2O9hreUP7X/zfYnIHYdz3+0l\nIpeJyG9F5EwROUNEfiIiXxaRE0XkbhGZ0dJr3TSyV0q5nIicA0wVkSnASOejFCgK+X4McBJQbIz5\nl4iccYh2LgNOB6qBv4V8/y5jzD1O4PUAu5zPr2PP/P8nIpcCw4FhwP6Q7bcBg4GtxpinQ9r8udM/\nH3ZJxcHAVmBsyPdD+78/ZF83AklAlIhc20L7VwBp2GlZpoe039hWP50+BI9xCXAB8Al2Ccgv7MsY\n83GNOEkAAATESURBVLyILMMuJjMQuzLYeGCq0078IX5tgLtG9odjIjSlVAQZY94ElhtjVgAZxpgH\ngMHNvh+PDeKtrVg1G7u84bvNd+F89jiP/wakGWO2ARud58YZYx4EypptPxPYA6Q2a1Owi6/0b7bN\nge+H9r/Zvvo7+5JW2v/M+V5cs23a088DxwicAzzg9OWQ+xKRaODrwD/BrgzmPJUEPAqcTQvcNLLX\nahyluodhInI8UCAitwI7m33/GKAGiHW+HyMiJxljFoW0sRwgdNTqWCciNwO9gFJjjBERRGQqduR9\nHrBeRG5xtlkcsv0H2NHw8mZtGuB72BH7tpBtBoZ8P7T/JmRfRSJynfN8S+33xi7TOLLZNqXt6Gf/\n4DFir1X+SESWtLKve4A6YAawJGRlsArgJ0AhLdA7aJVS3ZKI3GGMOeT1gGbb3WWMuae934+09vbz\nsPfDLcFe76BVSqnDR3P2SqkOa6tCprVqnDD3c1jWFe1MBY5TCXNOG9tc6Hxu18/hUBVKoZU2InKn\niNwiIuPaaqs5zdkrpdqtWeVJLU1VKz+nqarEj71Im+pUoJzhvLalapZc7IXILcB6mipVMkO+3ygi\n38VeC2gI2WYJcBfwtjHmqZA2D3cFzmRgGrDIKXc8CbtQ0qyQPv8t2KYx5sOQn8OPgfuB24wx94W0\necgKJWyFTrDSptj5+YfNNSP7N3bceuBDKeVaBypPmlWthFaVDAYe4Yvx5UCFiYikOCPUVGwg+wx7\n4TG0CiX0+wLkAn35YqXKB6GB3nFYK3CMMauc177HwdVHB/rcrM1QbwI/BD5t9v2WKpQOVNoYY/KA\n3wHXHqLdVrkm2KNpHKW6gwOVJ80qZIxpugC4B/g2dkQNTjUOtsKkN7DeGFNhjHnQGFMe0raEbtPs\n+wb4LlByiG38fFFoBU7o9qHfB6cCp9mxNK/Aad6f5ppXHwFIszYP/ByMMcuxZy3Ng/pyYMEhKpRe\nxam0cUb/vwIWt9KfQ3LNBdrszLwDF2hz87P0Aq1S3ZSIHAecC3xijGke0I5UH7q0Aqc1IjILmOiM\n0o8YN+XslVJHAWPMSmBlF/fhkAG9qwO904fFdGBk3lmaxlFKdVrzapPWqlwOtX0n9tvTq3SeCfn6\nQRE5paU23DSy12ocpVyuWeVMfXC+l2ZVN6Hz54RWuRyokDHG/F2rdDpXpWOM+UhEVjnbnA+saO13\n55qR/blDHzjwoZRyrdAKmdD5Xg5opcoltEImlFbpWOFW6YSaCEwGjm9pA9cEezSNo1R3Ihw8jw00\nVd3AoatcmlfCaJVOyM8h3CodEZkAzBKRs4wx9wPz+eIbRdP2bqnG0ekSlDq6uaQSpsdW6bgm2Cul\nlDp8XHOBVkf2Sil1+Lgm2KNr0Cql1GHjpmCvpZdKKXWYaM5eKaV6ADeVXiqllDpMNNgrpVQPoMFe\nKaV6AA32SinVA2iwV0qpHkCDvVJK9QAa7JVSqgfQYK+UUj2ABnullOoBNNgrpVQPoMFeKaV6AA32\nSinVA2iwV0qpHkCDvVJK9QAa7JVSqgfQYK+UUj2ABnullOoBNNgrpVQPoMFeKaV6AA32SinVA2iw\nV0qpHkCDvVJK9QAa7JVSqgfQYK+UUj2ABnullOoB/h8uF6DesmGc6AAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x1161b4940>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "PeptidesRelatednessAnalysis(DataFolderLocation,\n",
    "                           BaseSelectionRoundNumber,\n",
    "                           TopNPeptidesNumber,\n",
    "                           StartSequence,\n",
    "                           StopSequence,\n",
    "                           cDNA_MinLength,\n",
    "                           cDNA_MaxLength,\n",
    "                           QualityScore,\n",
    "                           FileName)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
